{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/MoumitaRoy13/RNASequence_Analysis_Project/blob/main/RNASEQ.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "9V7XGE0kuiO_",
        "outputId": "6039ecd4-a134-4fba-a63a-bc1c84883ee9"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "The following additional packages will be installed:\n",
            "  default-jre default-jre-headless fonts-dejavu-core fonts-dejavu-extra libapache-pom-java\n",
            "  libargs4j-java libatk-wrapper-java libatk-wrapper-java-jni libcommons-compress-java\n",
            "  libcommons-io-java libcommons-jexl2-java libcommons-lang3-java libcommons-logging-java\n",
            "  libcommons-math3-java libcommons-parent-java libfindbin-libs-perl libfontenc1 libhtsjdk-java\n",
            "  libjbzip2-java libjson-simple-java libngs-java libngs-sdk-dev libngs-sdk2 libsis-base-java\n",
            "  libsis-base-jni libsis-jhdf5-java libsis-jhdf5-jni libsnappy-java libsnappy-jni libxkbfile1\n",
            "  libxtst6 libxxf86dga1 openjdk-11-jre x11-utils\n",
            "Suggested packages:\n",
            "  libxz-java libcommons-io-java-doc libcommons-lang3-java-doc libavalon-framework-java\n",
            "  libcommons-logging-java-doc libexcalibur-logkit-java liblog4j1.2-java picard-tools\n",
            "  libjson-simple-doc mesa-utils\n",
            "The following NEW packages will be installed:\n",
            "  default-jre default-jre-headless fastqc fonts-dejavu-core fonts-dejavu-extra libapache-pom-java\n",
            "  libargs4j-java libatk-wrapper-java libatk-wrapper-java-jni libcommons-compress-java\n",
            "  libcommons-io-java libcommons-jexl2-java libcommons-lang3-java libcommons-logging-java\n",
            "  libcommons-math3-java libcommons-parent-java libfindbin-libs-perl libfontenc1 libhtsjdk-java\n",
            "  libjbzip2-java libjson-simple-java libngs-java libngs-sdk-dev libngs-sdk2 libsis-base-java\n",
            "  libsis-base-jni libsis-jhdf5-java libsis-jhdf5-jni libsnappy-java libsnappy-jni libxkbfile1\n",
            "  libxtst6 libxxf86dga1 openjdk-11-jre x11-utils\n",
            "0 upgraded, 35 newly installed, 0 to remove and 49 not upgraded.\n",
            "Need to get 11.8 MB of archives.\n",
            "After this operation, 24.8 MB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu jammy/main amd64 default-jre-headless amd64 2:1.11-72build2 [3,042 B]\n",
            "Get:2 http://archive.ubuntu.com/ubuntu jammy/main amd64 libxtst6 amd64 2:1.2.3-1build4 [13.4 kB]\n",
            "Get:3 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 openjdk-11-jre amd64 11.0.24+8-1ubuntu3~22.04 [214 kB]\n",
            "Get:4 http://archive.ubuntu.com/ubuntu jammy/main amd64 default-jre amd64 2:1.11-72build2 [896 B]\n",
            "Get:5 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libcommons-math3-java all 3.6.1-3 [2,016 kB]\n",
            "Get:6 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libjbzip2-java all 0.9.1-7 [42.8 kB]\n",
            "Get:7 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libapache-pom-java all 18-1 [4,720 B]\n",
            "Get:8 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libcommons-parent-java all 43-1 [10.8 kB]\n",
            "Get:9 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libcommons-compress-java all 1.21-1 [575 kB]\n",
            "Get:10 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libcommons-logging-java all 1.2-2 [60.3 kB]\n",
            "Get:11 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libcommons-jexl2-java all 2.1.1-5 [251 kB]\n",
            "Get:12 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libjson-simple-java all 2.3.0-1 [61.0 kB]\n",
            "Get:13 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libngs-sdk2 amd64 2.11.2-3 [65.6 kB]\n",
            "Get:14 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libngs-sdk-dev amd64 2.11.2-3 [111 kB]\n",
            "Get:15 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libngs-java amd64 2.11.2-3 [75.8 kB]\n",
            "Get:16 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libsnappy-jni amd64 1.1.8.3-1 [6,756 B]\n",
            "Get:17 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libsnappy-java all 1.1.8.3-1 [93.9 kB]\n",
            "Get:18 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libhtsjdk-java all 2.24.1+dfsg-1 [1,644 kB]\n",
            "Get:19 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libfindbin-libs-perl all 2.190.02-1 [23.2 kB]\n",
            "Get:20 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libargs4j-java all 2.33-1.1 [137 kB]\n",
            "Get:21 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libcommons-io-java all 2.11.0-2 [297 kB]\n",
            "Get:22 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libcommons-lang3-java all 3.11-1 [526 kB]\n",
            "Get:23 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libsis-base-jni amd64 18.09~pre1+git20180928.45fbd31+dfsg-3 [12.7 kB]\n",
            "Get:24 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libsis-base-java all 18.09~pre1+git20180928.45fbd31+dfsg-3 [96.4 kB]\n",
            "Get:25 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libsis-jhdf5-jni amd64 19.04.0+dfsg-4 [96.6 kB]\n",
            "Get:26 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libsis-jhdf5-java all 19.04.0+dfsg-4 [1,490 kB]\n",
            "Get:27 http://archive.ubuntu.com/ubuntu jammy/universe amd64 fastqc all 0.11.9+dfsg-5 [370 kB]\n",
            "Get:28 http://archive.ubuntu.com/ubuntu jammy/main amd64 fonts-dejavu-core all 2.37-2build1 [1,041 kB]\n",
            "Get:29 http://archive.ubuntu.com/ubuntu jammy/main amd64 fonts-dejavu-extra all 2.37-2build1 [2,041 kB]\n",
            "Get:30 http://archive.ubuntu.com/ubuntu jammy/main amd64 libfontenc1 amd64 1:1.1.4-1build3 [14.7 kB]\n",
            "Get:31 http://archive.ubuntu.com/ubuntu jammy/main amd64 libxkbfile1 amd64 1:1.1.0-1build3 [71.8 kB]\n",
            "Get:32 http://archive.ubuntu.com/ubuntu jammy/main amd64 libxxf86dga1 amd64 2:1.1.5-0ubuntu3 [12.6 kB]\n",
            "Get:33 http://archive.ubuntu.com/ubuntu jammy/main amd64 x11-utils amd64 7.7+5build2 [206 kB]\n",
            "Get:34 http://archive.ubuntu.com/ubuntu jammy/main amd64 libatk-wrapper-java all 0.38.0-5build1 [53.1 kB]\n",
            "Get:35 http://archive.ubuntu.com/ubuntu jammy/main amd64 libatk-wrapper-java-jni amd64 0.38.0-5build1 [49.0 kB]\n",
            "Fetched 11.8 MB in 3s (3,549 kB/s)\n",
            "Extracting templates from packages: 100%\n",
            "Selecting previously unselected package default-jre-headless.\n",
            "(Reading database ... 123622 files and directories currently installed.)\n",
            "Preparing to unpack .../00-default-jre-headless_2%3a1.11-72build2_amd64.deb ...\n",
            "Unpacking default-jre-headless (2:1.11-72build2) ...\n",
            "Selecting previously unselected package libxtst6:amd64.\n",
            "Preparing to unpack .../01-libxtst6_2%3a1.2.3-1build4_amd64.deb ...\n",
            "Unpacking libxtst6:amd64 (2:1.2.3-1build4) ...\n",
            "Selecting previously unselected package openjdk-11-jre:amd64.\n",
            "Preparing to unpack .../02-openjdk-11-jre_11.0.24+8-1ubuntu3~22.04_amd64.deb ...\n",
            "Unpacking openjdk-11-jre:amd64 (11.0.24+8-1ubuntu3~22.04) ...\n",
            "Selecting previously unselected package default-jre.\n",
            "Preparing to unpack .../03-default-jre_2%3a1.11-72build2_amd64.deb ...\n",
            "Unpacking default-jre (2:1.11-72build2) ...\n",
            "Selecting previously unselected package libcommons-math3-java.\n",
            "Preparing to unpack .../04-libcommons-math3-java_3.6.1-3_all.deb ...\n",
            "Unpacking libcommons-math3-java (3.6.1-3) ...\n",
            "Selecting previously unselected package libjbzip2-java.\n",
            "Preparing to unpack .../05-libjbzip2-java_0.9.1-7_all.deb ...\n",
            "Unpacking libjbzip2-java (0.9.1-7) ...\n",
            "Selecting previously unselected package libapache-pom-java.\n",
            "Preparing to unpack .../06-libapache-pom-java_18-1_all.deb ...\n",
            "Unpacking libapache-pom-java (18-1) ...\n",
            "Selecting previously unselected package libcommons-parent-java.\n",
            "Preparing to unpack .../07-libcommons-parent-java_43-1_all.deb ...\n",
            "Unpacking libcommons-parent-java (43-1) ...\n",
            "Selecting previously unselected package libcommons-compress-java.\n",
            "Preparing to unpack .../08-libcommons-compress-java_1.21-1_all.deb ...\n",
            "Unpacking libcommons-compress-java (1.21-1) ...\n",
            "Selecting previously unselected package libcommons-logging-java.\n",
            "Preparing to unpack .../09-libcommons-logging-java_1.2-2_all.deb ...\n",
            "Unpacking libcommons-logging-java (1.2-2) ...\n",
            "Selecting previously unselected package libcommons-jexl2-java.\n",
            "Preparing to unpack .../10-libcommons-jexl2-java_2.1.1-5_all.deb ...\n",
            "Unpacking libcommons-jexl2-java (2.1.1-5) ...\n",
            "Selecting previously unselected package libjson-simple-java.\n",
            "Preparing to unpack .../11-libjson-simple-java_2.3.0-1_all.deb ...\n",
            "Unpacking libjson-simple-java (2.3.0-1) ...\n",
            "Selecting previously unselected package libngs-sdk2:amd64.\n",
            "Preparing to unpack .../12-libngs-sdk2_2.11.2-3_amd64.deb ...\n",
            "Unpacking libngs-sdk2:amd64 (2.11.2-3) ...\n",
            "Selecting previously unselected package libngs-sdk-dev:amd64.\n",
            "Preparing to unpack .../13-libngs-sdk-dev_2.11.2-3_amd64.deb ...\n",
            "Unpacking libngs-sdk-dev:amd64 (2.11.2-3) ...\n",
            "Selecting previously unselected package libngs-java:amd64.\n",
            "Preparing to unpack .../14-libngs-java_2.11.2-3_amd64.deb ...\n",
            "Unpacking libngs-java:amd64 (2.11.2-3) ...\n",
            "Selecting previously unselected package libsnappy-jni.\n",
            "Preparing to unpack .../15-libsnappy-jni_1.1.8.3-1_amd64.deb ...\n",
            "Unpacking libsnappy-jni (1.1.8.3-1) ...\n",
            "Selecting previously unselected package libsnappy-java.\n",
            "Preparing to unpack .../16-libsnappy-java_1.1.8.3-1_all.deb ...\n",
            "Unpacking libsnappy-java (1.1.8.3-1) ...\n",
            "Selecting previously unselected package libhtsjdk-java.\n",
            "Preparing to unpack .../17-libhtsjdk-java_2.24.1+dfsg-1_all.deb ...\n",
            "Unpacking libhtsjdk-java (2.24.1+dfsg-1) ...\n",
            "Selecting previously unselected package libfindbin-libs-perl.\n",
            "Preparing to unpack .../18-libfindbin-libs-perl_2.190.02-1_all.deb ...\n",
            "Unpacking libfindbin-libs-perl (2.190.02-1) ...\n",
            "Selecting previously unselected package libargs4j-java.\n",
            "Preparing to unpack .../19-libargs4j-java_2.33-1.1_all.deb ...\n",
            "Unpacking libargs4j-java (2.33-1.1) ...\n",
            "Selecting previously unselected package libcommons-io-java.\n",
            "Preparing to unpack .../20-libcommons-io-java_2.11.0-2_all.deb ...\n",
            "Unpacking libcommons-io-java (2.11.0-2) ...\n",
            "Selecting previously unselected package libcommons-lang3-java.\n",
            "Preparing to unpack .../21-libcommons-lang3-java_3.11-1_all.deb ...\n",
            "Unpacking libcommons-lang3-java (3.11-1) ...\n",
            "Selecting previously unselected package libsis-base-jni.\n",
            "Preparing to unpack .../22-libsis-base-jni_18.09~pre1+git20180928.45fbd31+dfsg-3_amd64.deb ...\n",
            "Unpacking libsis-base-jni (18.09~pre1+git20180928.45fbd31+dfsg-3) ...\n",
            "Selecting previously unselected package libsis-base-java.\n",
            "Preparing to unpack .../23-libsis-base-java_18.09~pre1+git20180928.45fbd31+dfsg-3_all.deb ...\n",
            "Unpacking libsis-base-java (18.09~pre1+git20180928.45fbd31+dfsg-3) ...\n",
            "Selecting previously unselected package libsis-jhdf5-jni.\n",
            "Preparing to unpack .../24-libsis-jhdf5-jni_19.04.0+dfsg-4_amd64.deb ...\n",
            "Unpacking libsis-jhdf5-jni (19.04.0+dfsg-4) ...\n",
            "Selecting previously unselected package libsis-jhdf5-java.\n",
            "Preparing to unpack .../25-libsis-jhdf5-java_19.04.0+dfsg-4_all.deb ...\n",
            "Unpacking libsis-jhdf5-java (19.04.0+dfsg-4) ...\n",
            "Selecting previously unselected package fastqc.\n",
            "Preparing to unpack .../26-fastqc_0.11.9+dfsg-5_all.deb ...\n",
            "Unpacking fastqc (0.11.9+dfsg-5) ...\n",
            "Selecting previously unselected package fonts-dejavu-core.\n",
            "Preparing to unpack .../27-fonts-dejavu-core_2.37-2build1_all.deb ...\n",
            "Unpacking fonts-dejavu-core (2.37-2build1) ...\n",
            "Selecting previously unselected package fonts-dejavu-extra.\n",
            "Preparing to unpack .../28-fonts-dejavu-extra_2.37-2build1_all.deb ...\n",
            "Unpacking fonts-dejavu-extra (2.37-2build1) ...\n",
            "Selecting previously unselected package libfontenc1:amd64.\n",
            "Preparing to unpack .../29-libfontenc1_1%3a1.1.4-1build3_amd64.deb ...\n",
            "Unpacking libfontenc1:amd64 (1:1.1.4-1build3) ...\n",
            "Selecting previously unselected package libxkbfile1:amd64.\n",
            "Preparing to unpack .../30-libxkbfile1_1%3a1.1.0-1build3_amd64.deb ...\n",
            "Unpacking libxkbfile1:amd64 (1:1.1.0-1build3) ...\n",
            "Selecting previously unselected package libxxf86dga1:amd64.\n",
            "Preparing to unpack .../31-libxxf86dga1_2%3a1.1.5-0ubuntu3_amd64.deb ...\n",
            "Unpacking libxxf86dga1:amd64 (2:1.1.5-0ubuntu3) ...\n",
            "Selecting previously unselected package x11-utils.\n",
            "Preparing to unpack .../32-x11-utils_7.7+5build2_amd64.deb ...\n",
            "Unpacking x11-utils (7.7+5build2) ...\n",
            "Selecting previously unselected package libatk-wrapper-java.\n",
            "Preparing to unpack .../33-libatk-wrapper-java_0.38.0-5build1_all.deb ...\n",
            "Unpacking libatk-wrapper-java (0.38.0-5build1) ...\n",
            "Selecting previously unselected package libatk-wrapper-java-jni:amd64.\n",
            "Preparing to unpack .../34-libatk-wrapper-java-jni_0.38.0-5build1_amd64.deb ...\n",
            "Unpacking libatk-wrapper-java-jni:amd64 (0.38.0-5build1) ...\n",
            "Setting up default-jre-headless (2:1.11-72build2) ...\n",
            "Setting up libxtst6:amd64 (2:1.2.3-1build4) ...\n",
            "Setting up libngs-sdk2:amd64 (2.11.2-3) ...\n",
            "Setting up libxxf86dga1:amd64 (2:1.1.5-0ubuntu3) ...\n",
            "Setting up openjdk-11-jre:amd64 (11.0.24+8-1ubuntu3~22.04) ...\n",
            "Setting up libfindbin-libs-perl (2.190.02-1) ...\n",
            "Setting up default-jre (2:1.11-72build2) ...\n",
            "Setting up libjson-simple-java (2.3.0-1) ...\n",
            "Setting up libsis-base-jni (18.09~pre1+git20180928.45fbd31+dfsg-3) ...\n",
            "Setting up libfontenc1:amd64 (1:1.1.4-1build3) ...\n",
            "Setting up libapache-pom-java (18-1) ...\n",
            "Setting up fonts-dejavu-core (2.37-2build1) ...\n",
            "Setting up libjbzip2-java (0.9.1-7) ...\n",
            "Setting up libngs-sdk-dev:amd64 (2.11.2-3) ...\n",
            "Setting up fonts-dejavu-extra (2.37-2build1) ...\n",
            "Setting up libcommons-math3-java (3.6.1-3) ...\n",
            "Setting up libsis-jhdf5-jni (19.04.0+dfsg-4) ...\n",
            "Setting up libxkbfile1:amd64 (1:1.1.0-1build3) ...\n",
            "Setting up libngs-java:amd64 (2.11.2-3) ...\n",
            "Setting up libsnappy-jni (1.1.8.3-1) ...\n",
            "Setting up libargs4j-java (2.33-1.1) ...\n",
            "Setting up libcommons-parent-java (43-1) ...\n",
            "Setting up libcommons-logging-java (1.2-2) ...\n",
            "Setting up libcommons-jexl2-java (2.1.1-5) ...\n",
            "Setting up libcommons-lang3-java (3.11-1) ...\n",
            "Setting up x11-utils (7.7+5build2) ...\n",
            "Setting up libsnappy-java (1.1.8.3-1) ...\n",
            "Setting up libatk-wrapper-java (0.38.0-5build1) ...\n",
            "Setting up libcommons-compress-java (1.21-1) ...\n",
            "Setting up libcommons-io-java (2.11.0-2) ...\n",
            "Setting up libatk-wrapper-java-jni:amd64 (0.38.0-5build1) ...\n",
            "Setting up libsis-base-java (18.09~pre1+git20180928.45fbd31+dfsg-3) ...\n",
            "Setting up libhtsjdk-java (2.24.1+dfsg-1) ...\n",
            "Setting up libsis-jhdf5-java (19.04.0+dfsg-4) ...\n",
            "Setting up fastqc (0.11.9+dfsg-5) ...\n",
            "Processing triggers for man-db (2.10.2-1) ...\n",
            "Processing triggers for fontconfig (2.13.1-4.2ubuntu5) ...\n",
            "Processing triggers for hicolor-icon-theme (0.17-2) ...\n",
            "Processing triggers for libc-bin (2.35-0ubuntu3.4) ...\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_level_zero.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_5.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_opencl.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbb.so.12 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_0.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_loader.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc_proxy.so.2 is not a symbolic link\n",
            "\n"
          ]
        }
      ],
      "source": [
        "!apt-get install fastqc"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "fF34tM8wuugl"
      },
      "outputs": [],
      "source": [
        "!fastqc /content/drive/MyDrive/SRR17159271_1.fastq -o /content/drive/MyDrive/QC_reports/\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 18,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "EiLIsBSduxwo",
        "outputId": "836a387e-785c-44e9-90ef-8914b6c3538a"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2024-10-27 10:48:24--  https://github.com/FelixKrueger/TrimGalore/archive/refs/tags/0.6.10.zip\n",
            "Resolving github.com (github.com)... 20.27.177.113\n",
            "Connecting to github.com (github.com)|20.27.177.113|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://codeload.github.com/FelixKrueger/TrimGalore/zip/refs/tags/0.6.10 [following]\n",
            "--2024-10-27 10:48:25--  https://codeload.github.com/FelixKrueger/TrimGalore/zip/refs/tags/0.6.10\n",
            "Resolving codeload.github.com (codeload.github.com)... 20.27.177.114\n",
            "Connecting to codeload.github.com (codeload.github.com)|20.27.177.114|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: unspecified [application/zip]\n",
            "Saving to: ‘0.6.10.zip’\n",
            "\n",
            "0.6.10.zip              [       <=>          ]  25.59M  17.5MB/s    in 1.5s    \n",
            "\n",
            "2024-10-27 10:48:27 (17.5 MB/s) - ‘0.6.10.zip’ saved [26835264]\n",
            "\n",
            "Archive:  0.6.10.zip\n",
            "4edff97d22f3837d42a29e4afbfaeb6e07ffb11b\n",
            "   creating: TrimGalore-0.6.10/\n",
            "  inflating: TrimGalore-0.6.10/.travis.yml  \n",
            "  inflating: TrimGalore-0.6.10/CHANGELOG.md  \n",
            "   creating: TrimGalore-0.6.10/Docs/\n",
            "   creating: TrimGalore-0.6.10/Docs/Images/\n",
            "  inflating: TrimGalore-0.6.10/Docs/Images/Thumbs.db  \n",
            "  inflating: TrimGalore-0.6.10/Docs/Images/adapters.png  \n",
            "  inflating: TrimGalore-0.6.10/Docs/Images/adapters_fixed.png  \n",
            "  inflating: TrimGalore-0.6.10/Docs/Images/fixed_quals.png  \n",
            "  inflating: TrimGalore-0.6.10/Docs/Images/fixed_quals_per_sequence.png  \n",
            "  inflating: TrimGalore-0.6.10/Docs/Images/logo.png  \n",
            "  inflating: TrimGalore-0.6.10/Docs/Images/pigz_bench.png  \n",
            "  inflating: TrimGalore-0.6.10/Docs/Images/poor_q_per_sequence.png  \n",
            "  inflating: TrimGalore-0.6.10/Docs/Images/poor_quals.png  \n",
            "  inflating: TrimGalore-0.6.10/Docs/RRBS_Guide.pdf  \n",
            "  inflating: TrimGalore-0.6.10/Docs/Trim_Galore_User_Guide.md  \n",
            "  inflating: TrimGalore-0.6.10/LICENSE  \n",
            "  inflating: TrimGalore-0.6.10/README.md  \n",
            "   creating: TrimGalore-0.6.10/test_files/\n",
            " extracting: TrimGalore-0.6.10/test_files/4_seqs_with_Ns.fastq.gz  \n",
            " extracting: TrimGalore-0.6.10/test_files/clock_10K_R1.fastq.gz  \n",
            " extracting: TrimGalore-0.6.10/test_files/clock_10K_R2.fastq.gz  \n",
            "  inflating: TrimGalore-0.6.10/test_files/colorspace_file.fastq  \n",
            " extracting: TrimGalore-0.6.10/test_files/empty_file.fastq  \n",
            " extracting: TrimGalore-0.6.10/test_files/illumina_100K.fastq.gz  \n",
            "  inflating: TrimGalore-0.6.10/test_files/illumina_100K.fastq.gz_trimming_report.txt  \n",
            " extracting: TrimGalore-0.6.10/test_files/illumina_10K.fastq.gz  \n",
            "  inflating: TrimGalore-0.6.10/test_files/illumina_10K.fastq.gz_trimming_report.txt  \n",
            " extracting: TrimGalore-0.6.10/test_files/illumina_10K_trimmed.fq.gz  \n",
            " extracting: TrimGalore-0.6.10/test_files/nextera_100K.fastq.gz  \n",
            "  inflating: TrimGalore-0.6.10/test_files/nextera_100K.fastq.gz_trimming_report.txt  \n",
            " extracting: TrimGalore-0.6.10/test_files/smallRNA_100K.fastq.gz  \n",
            "  inflating: TrimGalore-0.6.10/test_files/smallRNA_100K.fastq.gz_trimming_report.txt  \n",
            "  inflating: TrimGalore-0.6.10/test_files/smallRNA_100K_R1.fastq.gz_trimming_report.txt  \n",
            " extracting: TrimGalore-0.6.10/test_files/smallRNA_100K_R2.fastq.gz  \n",
            "  inflating: TrimGalore-0.6.10/test_files/smallRNA_100K_R2.fastq.gz_trimming_report.txt  \n",
            "  inflating: TrimGalore-0.6.10/test_files/truncated.fq.gz  \n",
            "  inflating: TrimGalore-0.6.10/trim_galore  \n"
          ]
        }
      ],
      "source": [
        "!wget https://github.com/FelixKrueger/TrimGalore/archive/refs/tags/0.6.10.zip\n",
        "!unzip 0.6.10.zip\n",
        "!chmod +x TrimGalore-0.6.10/trim_galore"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "hsiCKOng1Lgw",
        "outputId": "7bce5ff5-858a-4252-96a6-84d9fb22380b"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount(\"/content/drive\", force_remount=True).\n"
          ]
        }
      ],
      "source": [
        "from google.colab import drive\n",
        "drive.mount('/content/drive')\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 19,
      "metadata": {
        "id": "a28tBcQmu1lk"
      },
      "outputs": [],
      "source": [
        "!mv TrimGalore-0.6.10/trim_galore /usr/local/bin/"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 20,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "3RvUKtYau2Vt",
        "outputId": "13c4e8e9-913a-4f08-bea0-aeb5b7d084e8"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting cutadapt\n",
            "  Downloading cutadapt-4.9-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (3.4 kB)\n",
            "Collecting dnaio>=1.2.0 (from cutadapt)\n",
            "  Downloading dnaio-1.2.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (3.5 kB)\n",
            "Collecting xopen>=1.6.0 (from cutadapt)\n",
            "  Downloading xopen-2.0.2-py3-none-any.whl.metadata (15 kB)\n",
            "Collecting isal>=1.6.1 (from xopen>=1.6.0->cutadapt)\n",
            "  Downloading isal-1.7.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (10 kB)\n",
            "Collecting zlib-ng>=0.4.1 (from xopen>=1.6.0->cutadapt)\n",
            "  Downloading zlib_ng-0.5.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (6.8 kB)\n",
            "Downloading cutadapt-4.9-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (287 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m287.7/287.7 kB\u001b[0m \u001b[31m6.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading dnaio-1.2.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (111 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m111.7/111.7 kB\u001b[0m \u001b[31m7.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading xopen-2.0.2-py3-none-any.whl (17 kB)\n",
            "Downloading isal-1.7.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (258 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m258.9/258.9 kB\u001b[0m \u001b[31m17.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading zlib_ng-0.5.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (108 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m108.1/108.1 kB\u001b[0m \u001b[31m7.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: zlib-ng, isal, xopen, dnaio, cutadapt\n",
            "Successfully installed cutadapt-4.9 dnaio-1.2.2 isal-1.7.1 xopen-2.0.2 zlib-ng-0.5.1\n"
          ]
        }
      ],
      "source": [
        "!pip install cutadapt"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 21,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "1Qxum3YYu7bJ",
        "outputId": "63a86ba0-de1a-4f34-c07e-970187f458b9"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "The following additional packages will be installed:\n",
            "  libisal2\n",
            "The following NEW packages will be installed:\n",
            "  isal libisal2\n",
            "0 upgraded, 2 newly installed, 0 to remove and 50 not upgraded.\n",
            "Need to get 121 kB of archives.\n",
            "After this operation, 403 kB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libisal2 amd64 2.30.0-4 [110 kB]\n",
            "Get:2 http://archive.ubuntu.com/ubuntu jammy/universe amd64 isal amd64 2.30.0-4 [10.7 kB]\n",
            "Fetched 121 kB in 1s (89.5 kB/s)\n",
            "Selecting previously unselected package libisal2.\n",
            "(Reading database ... 124456 files and directories currently installed.)\n",
            "Preparing to unpack .../libisal2_2.30.0-4_amd64.deb ...\n",
            "Unpacking libisal2 (2.30.0-4) ...\n",
            "Selecting previously unselected package isal.\n",
            "Preparing to unpack .../isal_2.30.0-4_amd64.deb ...\n",
            "Unpacking isal (2.30.0-4) ...\n",
            "Setting up libisal2 (2.30.0-4) ...\n",
            "Setting up isal (2.30.0-4) ...\n",
            "Processing triggers for man-db (2.10.2-1) ...\n",
            "Processing triggers for libc-bin (2.35-0ubuntu3.4) ...\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_level_zero.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_5.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_opencl.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbb.so.12 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_0.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_loader.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc_proxy.so.2 is not a symbolic link\n",
            "\n"
          ]
        }
      ],
      "source": [
        "!apt-get install -y isal"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 22,
      "metadata": {
        "id": "stLKTNuhDeZg",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "9b323ee6-57a9-4e84-cd60-cba270707f96"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Multicore support not enabled. Proceeding with single-core trimming.\n",
            "Path to Cutadapt set as: 'cutadapt' (default)\n",
            "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
            "Cutadapt version: 4.9\n",
            "single-core operation.\n",
            "igzip command line interface 2.30.0\n",
            "igzip detected. Using igzip for decompressing\n",
            "\n",
            "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
            "\n",
            "Output will be written into the directory: /content/drive/MyDrive/trimmed_output/\n",
            "\n",
            "\n",
            "AUTO-DETECTING ADAPTER TYPE\n",
            "===========================\n",
            "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> /content/drive/MyDrive/SRR17158981_1.fastq <<)\n",
            "\n",
            "Found perfect matches for the following adapter sequences:\n",
            "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
            "Illumina\t8248\tAGATCGGAAGAGC\t1000000\t0.82\n",
            "Nextera\t1\tCTGTCTCTTATA\t1000000\t0.00\n",
            "smallRNA\t1\tTGGAATTCTCGG\t1000000\t0.00\n",
            "Using Illumina adapter for trimming (count: 8248). Second best hit was Nextera (count: 1)\n",
            "\n",
            "Writing report to '/content/drive/MyDrive/trimmed_output/SRR17158981_1.fastq_trimming_report.txt'\n",
            "\n",
            "SUMMARISING RUN PARAMETERS\n",
            "==========================\n",
            "Input filename: /content/drive/MyDrive/SRR17158981_1.fastq\n",
            "Trimming mode: paired-end\n",
            "Trim Galore version: 0.6.10\n",
            "Cutadapt version: 4.9\n",
            "Number of cores used for trimming: 1\n",
            "Quality Phred score cutoff: 20\n",
            "Quality encoding type selected: ASCII+33\n",
            "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
            "Maximum trimming error rate: 0.1 (default)\n",
            "Minimum required adapter overlap (stringency): 1 bp\n",
            "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
            "\n",
            "Cutadapt seems to be fairly up-to-date (version 4.9). Setting -j 1\n",
            "Writing final adapter and quality trimmed output to SRR17158981_1_trimmed.fq\n",
            "\n",
            "\n",
            "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /content/drive/MyDrive/SRR17158981_1.fastq <<< \n",
            "10000000 sequences processed\n",
            "This is cutadapt 4.9 with Python 3.10.12\n",
            "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /content/drive/MyDrive/SRR17158981_1.fastq\n",
            "Processing single-end reads on 1 core ...\n",
            "Finished in 193.447 s (14.031 µs/read; 4.28 M reads/minute).\n",
            "\n",
            "=== Summary ===\n",
            "\n",
            "Total reads processed:              13,787,431\n",
            "Reads with adapters:                 4,725,295 (34.3%)\n",
            "Reads written (passing filters):    13,787,431 (100.0%)\n",
            "\n",
            "Total basepairs processed: 1,392,530,531 bp\n",
            "Quality-trimmed:                 292,968 bp (0.0%)\n",
            "Total written (filtered):  1,382,384,613 bp (99.3%)\n",
            "\n",
            "=== Adapter 1 ===\n",
            "\n",
            "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 4725295 times\n",
            "\n",
            "Minimum overlap: 1\n",
            "No. of allowed errors:\n",
            "1-9 bp: 0; 10-13 bp: 1\n",
            "\n",
            "Bases preceding removed adapters:\n",
            "  A: 25.0%\n",
            "  C: 33.8%\n",
            "  G: 23.4%\n",
            "  T: 17.7%\n",
            "  none/other: 0.0%\n",
            "\n",
            "Overview of removed sequences\n",
            "length\tcount\texpect\tmax.err\terror counts\n",
            "1\t3078148\t3446857.8\t0\t3078148\n",
            "2\t1049658\t861714.4\t0\t1049658\n",
            "3\t277845\t215428.6\t0\t277845\n",
            "4\t72640\t53857.2\t0\t72640\n",
            "5\t33130\t13464.3\t0\t33130\n",
            "6\t19351\t3366.1\t0\t19351\n",
            "7\t17375\t841.5\t0\t17375\n",
            "8\t16008\t210.4\t0\t16008\n",
            "9\t14711\t52.6\t0\t14576 135\n",
            "10\t13505\t13.1\t1\t13164 341\n",
            "11\t11891\t3.3\t1\t11636 255\n",
            "12\t10948\t0.8\t1\t10805 143\n",
            "13\t10365\t0.2\t1\t10235 130\n",
            "14\t9586\t0.2\t1\t9463 123\n",
            "15\t8880\t0.2\t1\t8784 96\n",
            "16\t8103\t0.2\t1\t7998 105\n",
            "17\t7331\t0.2\t1\t7221 110\n",
            "18\t6382\t0.2\t1\t6306 76\n",
            "19\t5625\t0.2\t1\t5557 68\n",
            "20\t5162\t0.2\t1\t5108 54\n",
            "21\t4790\t0.2\t1\t4737 53\n",
            "22\t4310\t0.2\t1\t4248 62\n",
            "23\t3978\t0.2\t1\t3912 66\n",
            "24\t3615\t0.2\t1\t3556 59\n",
            "25\t3358\t0.2\t1\t3310 48\n",
            "26\t2865\t0.2\t1\t2814 51\n",
            "27\t2515\t0.2\t1\t2484 31\n",
            "28\t2369\t0.2\t1\t2340 29\n",
            "29\t2153\t0.2\t1\t2121 32\n",
            "30\t1745\t0.2\t1\t1725 20\n",
            "31\t1626\t0.2\t1\t1598 28\n",
            "32\t1457\t0.2\t1\t1423 34\n",
            "33\t1238\t0.2\t1\t1217 21\n",
            "34\t1226\t0.2\t1\t1209 17\n",
            "35\t1070\t0.2\t1\t1053 17\n",
            "36\t977\t0.2\t1\t955 22\n",
            "37\t946\t0.2\t1\t926 20\n",
            "38\t812\t0.2\t1\t801 11\n",
            "39\t765\t0.2\t1\t739 26\n",
            "40\t683\t0.2\t1\t672 11\n",
            "41\t530\t0.2\t1\t519 11\n",
            "42\t552\t0.2\t1\t533 19\n",
            "43\t501\t0.2\t1\t486 15\n",
            "44\t332\t0.2\t1\t329 3\n",
            "45\t568\t0.2\t1\t549 19\n",
            "46\t72\t0.2\t1\t61 11\n",
            "47\t253\t0.2\t1\t232 21\n",
            "48\t286\t0.2\t1\t270 16\n",
            "49\t263\t0.2\t1\t258 5\n",
            "50\t182\t0.2\t1\t178 4\n",
            "51\t210\t0.2\t1\t204 6\n",
            "52\t155\t0.2\t1\t146 9\n",
            "53\t151\t0.2\t1\t143 8\n",
            "54\t112\t0.2\t1\t102 10\n",
            "55\t118\t0.2\t1\t114 4\n",
            "56\t81\t0.2\t1\t72 9\n",
            "57\t82\t0.2\t1\t78 4\n",
            "58\t86\t0.2\t1\t79 7\n",
            "59\t52\t0.2\t1\t50 2\n",
            "60\t71\t0.2\t1\t65 6\n",
            "61\t37\t0.2\t1\t25 12\n",
            "62\t50\t0.2\t1\t37 13\n",
            "63\t50\t0.2\t1\t45 5\n",
            "64\t24\t0.2\t1\t19 5\n",
            "65\t29\t0.2\t1\t24 5\n",
            "66\t36\t0.2\t1\t28 8\n",
            "67\t19\t0.2\t1\t14 5\n",
            "68\t21\t0.2\t1\t19 2\n",
            "69\t21\t0.2\t1\t14 7\n",
            "70\t21\t0.2\t1\t11 10\n",
            "71\t2\t0.2\t1\t0 2\n",
            "72\t10\t0.2\t1\t3 7\n",
            "73\t13\t0.2\t1\t7 6\n",
            "74\t14\t0.2\t1\t5 9\n",
            "75\t12\t0.2\t1\t8 4\n",
            "76\t11\t0.2\t1\t3 8\n",
            "77\t15\t0.2\t1\t5 10\n",
            "78\t12\t0.2\t1\t7 5\n",
            "79\t14\t0.2\t1\t3 11\n",
            "80\t28\t0.2\t1\t3 25\n",
            "81\t9\t0.2\t1\t1 8\n",
            "82\t23\t0.2\t1\t7 16\n",
            "83\t20\t0.2\t1\t17 3\n",
            "84\t11\t0.2\t1\t9 2\n",
            "85\t5\t0.2\t1\t2 3\n",
            "86\t4\t0.2\t1\t2 2\n",
            "87\t23\t0.2\t1\t8 15\n",
            "88\t3\t0.2\t1\t1 2\n",
            "89\t7\t0.2\t1\t4 3\n",
            "90\t6\t0.2\t1\t4 2\n",
            "91\t28\t0.2\t1\t15 13\n",
            "92\t5\t0.2\t1\t1 4\n",
            "93\t16\t0.2\t1\t14 2\n",
            "94\t5\t0.2\t1\t3 2\n",
            "95\t16\t0.2\t1\t7 9\n",
            "96\t13\t0.2\t1\t8 5\n",
            "97\t2\t0.2\t1\t1 1\n",
            "98\t10\t0.2\t1\t3 7\n",
            "99\t2\t0.2\t1\t0 2\n",
            "100\t1\t0.2\t1\t0 1\n",
            "101\t848\t0.2\t1\t1 847\n",
            "\n",
            "RUN STATISTICS FOR INPUT FILE: /content/drive/MyDrive/SRR17158981_1.fastq\n",
            "=============================================\n",
            "13787431 sequences processed in total\n",
            "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
            "\n",
            "Writing report to '/content/drive/MyDrive/trimmed_output/SRR17158981_2.fastq_trimming_report.txt'\n",
            "\n",
            "SUMMARISING RUN PARAMETERS\n",
            "==========================\n",
            "Input filename: /content/drive/MyDrive/SRR17158981_2.fastq\n",
            "Trimming mode: paired-end\n",
            "Trim Galore version: 0.6.10\n",
            "Cutadapt version: 4.9\n",
            "Number of cores used for trimming: 1\n",
            "Quality Phred score cutoff: 20\n",
            "Quality encoding type selected: ASCII+33\n",
            "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
            "Maximum trimming error rate: 0.1 (default)\n",
            "Minimum required adapter overlap (stringency): 1 bp\n",
            "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
            "\n",
            "Cutadapt seems to be fairly up-to-date (version 4.9). Setting -j -j 1\n",
            "Writing final adapter and quality trimmed output to SRR17158981_2_trimmed.fq\n",
            "\n",
            "\n",
            "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /content/drive/MyDrive/SRR17158981_2.fastq <<< \n",
            "10000000 sequences processed\n",
            "This is cutadapt 4.9 with Python 3.10.12\n",
            "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /content/drive/MyDrive/SRR17158981_2.fastq\n",
            "Processing single-end reads on 1 core ...\n",
            "Finished in 220.216 s (15.972 µs/read; 3.76 M reads/minute).\n",
            "\n",
            "=== Summary ===\n",
            "\n",
            "Total reads processed:              13,787,431\n",
            "Reads with adapters:                 5,169,732 (37.5%)\n",
            "Reads written (passing filters):    13,787,431 (100.0%)\n",
            "\n",
            "Total basepairs processed: 1,392,530,531 bp\n",
            "Quality-trimmed:               2,045,301 bp (0.1%)\n",
            "Total written (filtered):  1,380,003,657 bp (99.1%)\n",
            "\n",
            "=== Adapter 1 ===\n",
            "\n",
            "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 5169732 times\n",
            "\n",
            "Minimum overlap: 1\n",
            "No. of allowed errors:\n",
            "1-9 bp: 0; 10-13 bp: 1\n",
            "\n",
            "Bases preceding removed adapters:\n",
            "  A: 30.4%\n",
            "  C: 30.9%\n",
            "  G: 23.3%\n",
            "  T: 15.4%\n",
            "  none/other: 0.0%\n",
            "\n",
            "Overview of removed sequences\n",
            "length\tcount\texpect\tmax.err\terror counts\n",
            "1\t3383323\t3446857.8\t0\t3383323\n",
            "2\t1053643\t861714.4\t0\t1053643\n",
            "3\t405132\t215428.6\t0\t405132\n",
            "4\t80675\t53857.2\t0\t80675\n",
            "5\t32362\t13464.3\t0\t32362\n",
            "6\t20468\t3366.1\t0\t20468\n",
            "7\t17536\t841.5\t0\t17536\n",
            "8\t16902\t210.4\t0\t16902\n",
            "9\t13859\t52.6\t0\t13515 344\n",
            "10\t14426\t13.1\t1\t13763 663\n",
            "11\t11535\t3.3\t1\t11041 494\n",
            "12\t11034\t0.8\t1\t10872 162\n",
            "13\t10392\t0.2\t1\t10258 134\n",
            "14\t10188\t0.2\t1\t10052 136\n",
            "15\t8305\t0.2\t1\t8234 71\n",
            "16\t7954\t0.2\t1\t7851 103\n",
            "17\t8464\t0.2\t1\t8377 87\n",
            "18\t5015\t0.2\t1\t4951 64\n",
            "19\t7184\t0.2\t1\t7097 87\n",
            "20\t4382\t0.2\t1\t4338 44\n",
            "21\t4094\t0.2\t1\t4042 52\n",
            "22\t4321\t0.2\t1\t4261 60\n",
            "23\t3966\t0.2\t1\t3919 47\n",
            "24\t3747\t0.2\t1\t3705 42\n",
            "25\t3400\t0.2\t1\t3364 36\n",
            "26\t2817\t0.2\t1\t2773 44\n",
            "27\t2392\t0.2\t1\t2353 39\n",
            "28\t2617\t0.2\t1\t2591 26\n",
            "29\t1841\t0.2\t1\t1824 17\n",
            "30\t1948\t0.2\t1\t1921 27\n",
            "31\t1428\t0.2\t1\t1399 29\n",
            "32\t1447\t0.2\t1\t1427 20\n",
            "33\t1266\t0.2\t1\t1231 35\n",
            "34\t1252\t0.2\t1\t1233 19\n",
            "35\t1000\t0.2\t1\t984 16\n",
            "36\t1034\t0.2\t1\t1013 21\n",
            "37\t932\t0.2\t1\t912 20\n",
            "38\t825\t0.2\t1\t811 14\n",
            "39\t789\t0.2\t1\t772 17\n",
            "40\t630\t0.2\t1\t615 15\n",
            "41\t517\t0.2\t1\t505 12\n",
            "42\t601\t0.2\t1\t592 9\n",
            "43\t434\t0.2\t1\t428 6\n",
            "44\t1082\t0.2\t1\t1058 24\n",
            "45\t373\t0.2\t1\t369 4\n",
            "46\t265\t0.2\t1\t254 11\n",
            "47\t188\t0.2\t1\t172 16\n",
            "48\t63\t0.2\t1\t58 5\n",
            "49\t256\t0.2\t1\t249 7\n",
            "50\t44\t0.2\t1\t36 8\n",
            "51\t57\t0.2\t1\t52 5\n",
            "52\t55\t0.2\t1\t51 4\n",
            "53\t134\t0.2\t1\t130 4\n",
            "54\t172\t0.2\t1\t163 9\n",
            "55\t31\t0.2\t1\t29 2\n",
            "56\t84\t0.2\t1\t77 7\n",
            "57\t178\t0.2\t1\t175 3\n",
            "58\t50\t0.2\t1\t46 4\n",
            "59\t22\t0.2\t1\t19 3\n",
            "60\t115\t0.2\t1\t105 10\n",
            "61\t21\t0.2\t1\t20 1\n",
            "62\t11\t0.2\t1\t5 6\n",
            "63\t62\t0.2\t1\t57 5\n",
            "64\t71\t0.2\t1\t63 8\n",
            "65\t10\t0.2\t1\t4 6\n",
            "66\t5\t0.2\t1\t4 1\n",
            "67\t5\t0.2\t1\t1 4\n",
            "68\t13\t0.2\t1\t2 11\n",
            "69\t8\t0.2\t1\t7 1\n",
            "70\t10\t0.2\t1\t8 2\n",
            "71\t4\t0.2\t1\t3 1\n",
            "72\t7\t0.2\t1\t2 5\n",
            "73\t5\t0.2\t1\t3 2\n",
            "74\t9\t0.2\t1\t4 5\n",
            "75\t11\t0.2\t1\t5 6\n",
            "76\t7\t0.2\t1\t2 5\n",
            "77\t3\t0.2\t1\t2 1\n",
            "78\t9\t0.2\t1\t5 4\n",
            "79\t11\t0.2\t1\t2 9\n",
            "80\t14\t0.2\t1\t2 12\n",
            "81\t11\t0.2\t1\t3 8\n",
            "82\t9\t0.2\t1\t7 2\n",
            "83\t18\t0.2\t1\t14 4\n",
            "84\t17\t0.2\t1\t9 8\n",
            "85\t8\t0.2\t1\t2 6\n",
            "86\t5\t0.2\t1\t2 3\n",
            "87\t10\t0.2\t1\t9 1\n",
            "88\t6\t0.2\t1\t1 5\n",
            "89\t7\t0.2\t1\t4 3\n",
            "90\t6\t0.2\t1\t4 2\n",
            "91\t17\t0.2\t1\t15 2\n",
            "92\t9\t0.2\t1\t1 8\n",
            "93\t20\t0.2\t1\t14 6\n",
            "94\t6\t0.2\t1\t2 4\n",
            "95\t7\t0.2\t1\t6 1\n",
            "96\t9\t0.2\t1\t8 1\n",
            "97\t7\t0.2\t1\t1 6\n",
            "98\t6\t0.2\t1\t3 3\n",
            "99\t4\t0.2\t1\t0 4\n",
            "100\t4\t0.2\t1\t0 4\n",
            "101\t4\t0.2\t1\t1 3\n",
            "\n",
            "RUN STATISTICS FOR INPUT FILE: /content/drive/MyDrive/SRR17158981_2.fastq\n",
            "=============================================\n",
            "13787431 sequences processed in total\n",
            "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
            "\n",
            "Validate paired-end files SRR17158981_1_trimmed.fq and SRR17158981_2_trimmed.fq\n",
            "file_1: SRR17158981_1_trimmed.fq, file_2: SRR17158981_2_trimmed.fq\n",
            "\n",
            "\n",
            ">>>>> Now validing the length of the 2 paired-end infiles: SRR17158981_1_trimmed.fq and SRR17158981_2_trimmed.fq <<<<<\n",
            "Writing validated paired-end Read 1 reads to SRR17158981_1_val_1.fq\n",
            "Writing validated paired-end Read 2 reads to SRR17158981_2_val_2.fq\n",
            "\n",
            "Total number of sequences analysed: 13787431\n",
            "\n",
            "Number of sequence pairs removed because at least one read was shorter than the length cutoff (20 bp): 13776 (0.10%)\n",
            "\n",
            "Deleting both intermediate output files SRR17158981_1_trimmed.fq and SRR17158981_2_trimmed.fq\n",
            "\n",
            "====================================================================================================\n",
            "\n"
          ]
        }
      ],
      "source": [
        "!trim_galore --paired --quality 20 --length 20 /content/drive/MyDrive/SRR17158981_1.fastq /content/drive/MyDrive/SRR17158981_2.fastq -o /content/drive/MyDrive/trimmed_output/\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!hisat2 -x /content/drive/MyDrive/reference_index/GRCm39_index \\\n",
        "       -1 /content/drive/MyDrive/trimmed_output/SRR17158981_1_val_1.fq\\\n",
        "       -2 /content/drive/MyDrive/trimmed_output/SRR17158981_2_val_2.fq \\\n",
        "       -S /content/drive/MyDrive/SRR17158981_aligned.sam"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "3p4tkD_6owo5",
        "outputId": "6981f1b2-92d8-4115-c04f-0d254d1d8786"
      },
      "execution_count": 24,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "13773655 reads; of these:\n",
            "  13773655 (100.00%) were paired; of these:\n",
            "    320822 (2.33%) aligned concordantly 0 times\n",
            "    10935239 (79.39%) aligned concordantly exactly 1 time\n",
            "    2517594 (18.28%) aligned concordantly >1 times\n",
            "    ----\n",
            "    320822 pairs aligned concordantly 0 times; of these:\n",
            "      37426 (11.67%) aligned discordantly 1 time\n",
            "    ----\n",
            "    283396 pairs aligned 0 times concordantly or discordantly; of these:\n",
            "      566792 mates make up the pairs; of these:\n",
            "        304699 (53.76%) aligned 0 times\n",
            "        196159 (34.61%) aligned exactly 1 time\n",
            "        65934 (11.63%) aligned >1 times\n",
            "98.89% overall alignment rate\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!samtools view -bS /content/drive/MyDrive/SRR17158981_aligned.sam | samtools sort -o /content/drive/MyDrive/SRR17158981_sorted.bam\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "LjP8or6YoxGJ",
        "outputId": "279e40ea-915e-499c-e8f6-55850adf8539"
      },
      "execution_count": 25,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[bam_sort_core] merging from 15 files and 1 in-memory blocks...\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!samtools index /content/drive/MyDrive/SRR17158981_sorted.bam"
      ],
      "metadata": {
        "id": "KH6a-5IpoxIC"
      },
      "execution_count": 26,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!samtools index /content/drive/MyDrive/SRR17159271_sorted.bam"
      ],
      "metadata": {
        "id": "ljCAyuKnoxLv"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!samtools flagstat /content/drive/MyDrive/SRR17158981_sorted.bam"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "IVugcmcb5b0U",
        "outputId": "0b5efa46-f58a-4838-b15e-aa9b3c7a3d6d"
      },
      "execution_count": 27,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "40732515 + 0 in total (QC-passed reads + QC-failed reads)\n",
            "27547310 + 0 primary\n",
            "13185205 + 0 secondary\n",
            "0 + 0 supplementary\n",
            "0 + 0 duplicates\n",
            "0 + 0 primary duplicates\n",
            "40427816 + 0 mapped (99.25% : N/A)\n",
            "27242611 + 0 primary mapped (98.89% : N/A)\n",
            "27547310 + 0 paired in sequencing\n",
            "13773655 + 0 read1\n",
            "13773655 + 0 read2\n",
            "26905666 + 0 properly paired (97.67% : N/A)\n",
            "27109944 + 0 with itself and mate mapped\n",
            "132667 + 0 singletons (0.48% : N/A)\n",
            "192472 + 0 with mate mapped to a different chr\n",
            "148778 + 0 with mate mapped to a different chr (mapQ>=5)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!featureCounts -a /content/drive/MyDrive/gencode.vM35.basic.annotation.gtf -o /content/drive/MyDrive/SRR17158981_counts.txt -T 4 -p /content/drive/MyDrive/SRR17158981_sorted.bam\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "0ermZRSQ5cW9",
        "outputId": "ddfd256c-8c6a-4705-97fb-63926a33cc07"
      },
      "execution_count": 28,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
            "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
            "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
            "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
            "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
            "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
            "\t  v2.0.3\n",
            "\n",
            "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
            "||  \u001b[0m                                                                          ||\n",
            "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "||                           \u001b[36mSRR17158981_sorted.bam\u001b[0m \u001b[0m                          ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "||             Output file : \u001b[36mSRR17158981_counts.txt\u001b[0m \u001b[0m                          ||\n",
            "||                 Summary : \u001b[36mSRR17158981_counts.txt.summary\u001b[0m \u001b[0m                  ||\n",
            "||              Paired-end : \u001b[36myes\u001b[0m \u001b[0m                                             ||\n",
            "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
            "||              Annotation : \u001b[36mgencode.vM35.basic.annotation.gtf (GTF)\u001b[0m \u001b[0m         ||\n",
            "||      Dir for temp files : \u001b[36m/content/drive/MyDrive\u001b[0m \u001b[0m                          ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "||                 Threads : \u001b[36m4\u001b[0m \u001b[0m                                               ||\n",
            "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
            "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
            "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
            "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "\\\\============================================================================//\n",
            "\n",
            "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
            "||  \u001b[0m                                                                          ||\n",
            "|| Load annotation file gencode.vM35.basic.annotation.gtf \u001b[0m... \u001b[0m                ||\n",
            "||    Features : \u001b[36m532131\u001b[0m \u001b[0m                                                      ||\n",
            "||    Meta-features : \u001b[36m57132\u001b[0m \u001b[0m                                                  ||\n",
            "||    Chromosomes/contigs : \u001b[36m22\u001b[0m \u001b[0m                                               ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "|| Process BAM file SRR17158981_sorted.bam... \u001b[0m                                ||\n",
            "||    Paired-end reads are included. \u001b[0m                                         ||\n",
            "||    The reads are assigned on the single-end mode. \u001b[0m                         ||\n",
            "||    Total alignments : \u001b[36m40732515\u001b[0m \u001b[0m                                            ||\n",
            "||    Successfully assigned alignments : \u001b[36m20351434 (50.0%)\u001b[0m \u001b[0m                    ||\n",
            "||    Running time : \u001b[36m0.76 minutes\u001b[0m \u001b[0m                                            ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "|| Write the final count table. \u001b[0m                                              ||\n",
            "|| Write the read assignment summary. \u001b[0m                                        ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "|| Summary of counting results can be found in file \"/content/drive/MyDrive/ \u001b[0m ||\n",
            "|| SRR17158981_counts.txt.summary\" \u001b[0m                                           ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "\\\\============================================================================//\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "KAj_ZDbm1b_6",
        "outputId": "ba1b0431-d169-4540-8042-966128f43383"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Multicore support not enabled. Proceeding with single-core trimming.\n",
            "Path to Cutadapt set as: 'cutadapt' (default)\n",
            "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
            "Cutadapt version: 4.9\n",
            "single-core operation.\n",
            "igzip command line interface 2.30.0\n",
            "igzip detected. Using igzip for decompressing\n",
            "\n",
            "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
            "\n",
            "Output will be written into the directory: /content/drive/MyDrive/trimmed_output/\n",
            "\n",
            "\n",
            "AUTO-DETECTING ADAPTER TYPE\n",
            "===========================\n",
            "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> /content/drive/MyDrive/SRR17159236_1.fastq <<)\n",
            "\n",
            "Found perfect matches for the following adapter sequences:\n",
            "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
            "Illumina\t28983\tAGATCGGAAGAGC\t1000000\t2.90\n",
            "smallRNA\t1\tTGGAATTCTCGG\t1000000\t0.00\n",
            "Nextera\t0\tCTGTCTCTTATA\t1000000\t0.00\n",
            "Using Illumina adapter for trimming (count: 28983). Second best hit was smallRNA (count: 1)\n",
            "\n",
            "Writing report to '/content/drive/MyDrive/trimmed_output/SRR17159236_1.fastq_trimming_report.txt'\n",
            "\n",
            "SUMMARISING RUN PARAMETERS\n",
            "==========================\n",
            "Input filename: /content/drive/MyDrive/SRR17159236_1.fastq\n",
            "Trimming mode: paired-end\n",
            "Trim Galore version: 0.6.10\n",
            "Cutadapt version: 4.9\n",
            "Number of cores used for trimming: 1\n",
            "Quality Phred score cutoff: 20\n",
            "Quality encoding type selected: ASCII+33\n",
            "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
            "Maximum trimming error rate: 0.1 (default)\n",
            "Minimum required adapter overlap (stringency): 1 bp\n",
            "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
            "\n",
            "Cutadapt seems to be fairly up-to-date (version 4.9). Setting -j 1\n",
            "Writing final adapter and quality trimmed output to SRR17159236_1_trimmed.fq\n",
            "\n",
            "\n",
            "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /content/drive/MyDrive/SRR17159236_1.fastq <<< \n",
            "10000000 sequences processed\n",
            "20000000 sequences processed\n",
            "30000000 sequences processed\n",
            "This is cutadapt 4.9 with Python 3.10.12\n",
            "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /content/drive/MyDrive/SRR17159236_1.fastq\n",
            "Processing single-end reads on 1 core ...\n",
            "Finished in 440.346 s (13.644 µs/read; 4.40 M reads/minute).\n",
            "\n",
            "=== Summary ===\n",
            "\n",
            "Total reads processed:              32,275,135\n",
            "Reads with adapters:                12,176,195 (37.7%)\n",
            "Reads written (passing filters):    32,275,135 (100.0%)\n",
            "\n",
            "Total basepairs processed: 3,259,788,635 bp\n",
            "Quality-trimmed:                 777,230 bp (0.0%)\n",
            "Total written (filtered):  3,214,046,540 bp (98.6%)\n",
            "\n",
            "=== Adapter 1 ===\n",
            "\n",
            "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 12176195 times\n",
            "\n",
            "Minimum overlap: 1\n",
            "No. of allowed errors:\n",
            "1-9 bp: 0; 10-13 bp: 1\n",
            "\n",
            "Bases preceding removed adapters:\n",
            "  A: 25.9%\n",
            "  C: 29.5%\n",
            "  G: 24.0%\n",
            "  T: 20.6%\n",
            "  none/other: 0.1%\n",
            "\n",
            "Overview of removed sequences\n",
            "length\tcount\texpect\tmax.err\terror counts\n",
            "1\t7154153\t8068783.8\t0\t7154153\n",
            "2\t2413174\t2017195.9\t0\t2413174\n",
            "3\t685882\t504299.0\t0\t685882\n",
            "4\t232118\t126074.7\t0\t232118\n",
            "5\t127930\t31518.7\t0\t127930\n",
            "6\t102100\t7879.7\t0\t102100\n",
            "7\t95407\t1969.9\t0\t95407\n",
            "8\t89600\t492.5\t0\t89600\n",
            "9\t81833\t123.1\t0\t81372 461\n",
            "10\t76079\t30.8\t1\t74674 1405\n",
            "11\t72195\t7.7\t1\t70864 1331\n",
            "12\t67594\t1.9\t1\t66506 1088\n",
            "13\t67202\t0.5\t1\t66186 1016\n",
            "14\t64013\t0.5\t1\t62888 1125\n",
            "15\t62216\t0.5\t1\t61242 974\n",
            "16\t59448\t0.5\t1\t58393 1055\n",
            "17\t55646\t0.5\t1\t54662 984\n",
            "18\t51163\t0.5\t1\t50266 897\n",
            "19\t47317\t0.5\t1\t46423 894\n",
            "20\t73976\t0.5\t1\t73005 971\n",
            "21\t6647\t0.5\t1\t6358 289\n",
            "22\t35901\t0.5\t1\t35318 583\n",
            "23\t34430\t0.5\t1\t33815 615\n",
            "24\t33926\t0.5\t1\t33294 632\n",
            "25\t32368\t0.5\t1\t31833 535\n",
            "26\t30159\t0.5\t1\t29678 481\n",
            "27\t28819\t0.5\t1\t28322 497\n",
            "28\t26656\t0.5\t1\t26221 435\n",
            "29\t23630\t0.5\t1\t23149 481\n",
            "30\t21516\t0.5\t1\t21153 363\n",
            "31\t18851\t0.5\t1\t18516 335\n",
            "32\t17349\t0.5\t1\t17040 309\n",
            "33\t16278\t0.5\t1\t15981 297\n",
            "34\t14913\t0.5\t1\t14636 277\n",
            "35\t13823\t0.5\t1\t13531 292\n",
            "36\t13263\t0.5\t1\t13045 218\n",
            "37\t12476\t0.5\t1\t12228 248\n",
            "38\t12024\t0.5\t1\t11781 243\n",
            "39\t10360\t0.5\t1\t10160 200\n",
            "40\t9579\t0.5\t1\t9382 197\n",
            "41\t8656\t0.5\t1\t8467 189\n",
            "42\t7566\t0.5\t1\t7392 174\n",
            "43\t6990\t0.5\t1\t6861 129\n",
            "44\t5403\t0.5\t1\t5296 107\n",
            "45\t9457\t0.5\t1\t9297 160\n",
            "46\t1062\t0.5\t1\t1011 51\n",
            "47\t3101\t0.5\t1\t3004 97\n",
            "48\t5551\t0.5\t1\t5449 102\n",
            "49\t3415\t0.5\t1\t3335 80\n",
            "50\t2434\t0.5\t1\t2365 69\n",
            "51\t3425\t0.5\t1\t3356 69\n",
            "52\t2449\t0.5\t1\t2398 51\n",
            "53\t2010\t0.5\t1\t1971 39\n",
            "54\t1924\t0.5\t1\t1869 55\n",
            "55\t2169\t0.5\t1\t2108 61\n",
            "56\t996\t0.5\t1\t953 43\n",
            "57\t1321\t0.5\t1\t1261 60\n",
            "58\t850\t0.5\t1\t815 35\n",
            "59\t1420\t0.5\t1\t1378 42\n",
            "60\t587\t0.5\t1\t549 38\n",
            "61\t269\t0.5\t1\t245 24\n",
            "62\t781\t0.5\t1\t751 30\n",
            "63\t520\t0.5\t1\t487 33\n",
            "64\t230\t0.5\t1\t201 29\n",
            "65\t256\t0.5\t1\t218 38\n",
            "66\t433\t0.5\t1\t395 38\n",
            "67\t137\t0.5\t1\t124 13\n",
            "68\t253\t0.5\t1\t230 23\n",
            "69\t185\t0.5\t1\t167 18\n",
            "70\t162\t0.5\t1\t126 36\n",
            "71\t17\t0.5\t1\t6 11\n",
            "72\t28\t0.5\t1\t12 16\n",
            "73\t79\t0.5\t1\t36 43\n",
            "74\t59\t0.5\t1\t40 19\n",
            "75\t70\t0.5\t1\t37 33\n",
            "76\t61\t0.5\t1\t40 21\n",
            "77\t51\t0.5\t1\t25 26\n",
            "78\t45\t0.5\t1\t25 20\n",
            "79\t72\t0.5\t1\t33 39\n",
            "80\t84\t0.5\t1\t36 48\n",
            "81\t57\t0.5\t1\t33 24\n",
            "82\t71\t0.5\t1\t42 29\n",
            "83\t67\t0.5\t1\t28 39\n",
            "84\t61\t0.5\t1\t37 24\n",
            "85\t121\t0.5\t1\t93 28\n",
            "86\t59\t0.5\t1\t38 21\n",
            "87\t68\t0.5\t1\t35 33\n",
            "88\t53\t0.5\t1\t35 18\n",
            "89\t74\t0.5\t1\t61 13\n",
            "90\t40\t0.5\t1\t31 9\n",
            "91\t44\t0.5\t1\t27 17\n",
            "92\t94\t0.5\t1\t61 33\n",
            "93\t80\t0.5\t1\t66 14\n",
            "94\t81\t0.5\t1\t67 14\n",
            "95\t64\t0.5\t1\t29 35\n",
            "96\t81\t0.5\t1\t41 40\n",
            "97\t32\t0.5\t1\t18 14\n",
            "98\t149\t0.5\t1\t104 45\n",
            "99\t25\t0.5\t1\t11 14\n",
            "100\t50\t0.5\t1\t12 38\n",
            "101\t8232\t0.5\t1\t12 8220\n",
            "\n",
            "RUN STATISTICS FOR INPUT FILE: /content/drive/MyDrive/SRR17159236_1.fastq\n",
            "=============================================\n",
            "32275135 sequences processed in total\n",
            "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
            "\n",
            "Writing report to '/content/drive/MyDrive/trimmed_output/SRR17159236_2.fastq_trimming_report.txt'\n",
            "\n",
            "SUMMARISING RUN PARAMETERS\n",
            "==========================\n",
            "Input filename: /content/drive/MyDrive/SRR17159236_2.fastq\n",
            "Trimming mode: paired-end\n",
            "Trim Galore version: 0.6.10\n",
            "Cutadapt version: 4.9\n",
            "Number of cores used for trimming: 1\n",
            "Quality Phred score cutoff: 20\n",
            "Quality encoding type selected: ASCII+33\n",
            "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
            "Maximum trimming error rate: 0.1 (default)\n",
            "Minimum required adapter overlap (stringency): 1 bp\n",
            "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
            "\n",
            "Cutadapt seems to be fairly up-to-date (version 4.9). Setting -j -j 1\n",
            "Writing final adapter and quality trimmed output to SRR17159236_2_trimmed.fq\n",
            "\n",
            "\n",
            "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /content/drive/MyDrive/SRR17159236_2.fastq <<< \n",
            "10000000 sequences processed\n",
            "20000000 sequences processed\n",
            "30000000 sequences processed\n",
            "This is cutadapt 4.9 with Python 3.10.12\n",
            "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /content/drive/MyDrive/SRR17159236_2.fastq\n",
            "Processing single-end reads on 1 core ...\n",
            "Finished in 495.893 s (15.365 µs/read; 3.91 M reads/minute).\n",
            "\n",
            "=== Summary ===\n",
            "\n",
            "Total reads processed:              32,275,135\n",
            "Reads with adapters:                12,872,948 (39.9%)\n",
            "Reads written (passing filters):    32,275,135 (100.0%)\n",
            "\n",
            "Total basepairs processed: 3,259,788,635 bp\n",
            "Quality-trimmed:               6,120,015 bp (0.2%)\n",
            "Total written (filtered):  3,208,996,223 bp (98.4%)\n",
            "\n",
            "=== Adapter 1 ===\n",
            "\n",
            "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 12872948 times\n",
            "\n",
            "Minimum overlap: 1\n",
            "No. of allowed errors:\n",
            "1-9 bp: 0; 10-13 bp: 1\n",
            "\n",
            "Bases preceding removed adapters:\n",
            "  A: 27.0%\n",
            "  C: 32.2%\n",
            "  G: 23.1%\n",
            "  T: 17.6%\n",
            "  none/other: 0.0%\n",
            "\n",
            "Overview of removed sequences\n",
            "length\tcount\texpect\tmax.err\terror counts\n",
            "1\t7838587\t8068783.8\t0\t7838587\n",
            "2\t2331921\t2017195.9\t0\t2331921\n",
            "3\t789666\t504299.0\t0\t789666\n",
            "4\t227533\t126074.7\t0\t227533\n",
            "5\t131403\t31518.7\t0\t131403\n",
            "6\t105648\t7879.7\t0\t105648\n",
            "7\t96455\t1969.9\t0\t96455\n",
            "8\t91854\t492.5\t0\t91854\n",
            "9\t79395\t123.1\t0\t78693 702\n",
            "10\t77662\t30.8\t1\t76405 1257\n",
            "11\t71370\t7.7\t1\t69784 1586\n",
            "12\t67877\t1.9\t1\t67010 867\n",
            "13\t67112\t0.5\t1\t66509 603\n",
            "14\t66877\t0.5\t1\t66257 620\n",
            "15\t59204\t0.5\t1\t58729 475\n",
            "16\t58267\t0.5\t1\t57762 505\n",
            "17\t64679\t0.5\t1\t64152 527\n",
            "18\t40340\t0.5\t1\t39962 378\n",
            "19\t51823\t0.5\t1\t51367 456\n",
            "20\t39505\t0.5\t1\t39177 328\n",
            "21\t36577\t0.5\t1\t36237 340\n",
            "22\t35880\t0.5\t1\t35574 306\n",
            "23\t34347\t0.5\t1\t34045 302\n",
            "24\t33968\t0.5\t1\t33680 288\n",
            "25\t33374\t0.5\t1\t33111 263\n",
            "26\t29349\t0.5\t1\t29116 233\n",
            "27\t28054\t0.5\t1\t27756 298\n",
            "28\t27688\t0.5\t1\t27473 215\n",
            "29\t21943\t0.5\t1\t21760 183\n",
            "30\t22176\t0.5\t1\t22002 174\n",
            "31\t18140\t0.5\t1\t17987 153\n",
            "32\t17291\t0.5\t1\t17121 170\n",
            "33\t16187\t0.5\t1\t16005 182\n",
            "34\t14906\t0.5\t1\t14772 134\n",
            "35\t15287\t0.5\t1\t15115 172\n",
            "36\t34881\t0.5\t1\t34640 241\n",
            "37\t5782\t0.5\t1\t5700 82\n",
            "38\t2927\t0.5\t1\t2866 61\n",
            "39\t13040\t0.5\t1\t12943 97\n",
            "40\t3425\t0.5\t1\t3368 57\n",
            "41\t7577\t0.5\t1\t7512 65\n",
            "42\t13318\t0.5\t1\t13233 85\n",
            "43\t3011\t0.5\t1\t2953 58\n",
            "44\t13687\t0.5\t1\t13614 73\n",
            "45\t6510\t0.5\t1\t6456 54\n",
            "46\t3224\t0.5\t1\t3193 31\n",
            "47\t1229\t0.5\t1\t1187 42\n",
            "48\t1023\t0.5\t1\t982 41\n",
            "49\t3811\t0.5\t1\t3772 39\n",
            "50\t387\t0.5\t1\t359 28\n",
            "51\t846\t0.5\t1\t817 29\n",
            "52\t602\t0.5\t1\t576 26\n",
            "53\t1789\t0.5\t1\t1759 30\n",
            "54\t2489\t0.5\t1\t2454 35\n",
            "55\t690\t0.5\t1\t667 23\n",
            "56\t1154\t0.5\t1\t1131 23\n",
            "57\t2194\t0.5\t1\t2170 24\n",
            "58\t718\t0.5\t1\t691 27\n",
            "59\t555\t0.5\t1\t537 18\n",
            "60\t1596\t0.5\t1\t1570 26\n",
            "61\t529\t0.5\t1\t500 29\n",
            "62\t197\t0.5\t1\t188 9\n",
            "63\t648\t0.5\t1\t631 17\n",
            "64\t690\t0.5\t1\t663 27\n",
            "65\t117\t0.5\t1\t100 17\n",
            "66\t118\t0.5\t1\t97 21\n",
            "67\t51\t0.5\t1\t43 8\n",
            "68\t91\t0.5\t1\t66 25\n",
            "69\t56\t0.5\t1\t39 17\n",
            "70\t86\t0.5\t1\t78 8\n",
            "71\t32\t0.5\t1\t16 16\n",
            "72\t36\t0.5\t1\t21 15\n",
            "73\t42\t0.5\t1\t27 15\n",
            "74\t38\t0.5\t1\t24 14\n",
            "75\t45\t0.5\t1\t21 24\n",
            "76\t41\t0.5\t1\t25 16\n",
            "77\t7\t0.5\t1\t3 4\n",
            "78\t34\t0.5\t1\t16 18\n",
            "79\t39\t0.5\t1\t22 17\n",
            "80\t48\t0.5\t1\t28 20\n",
            "81\t40\t0.5\t1\t27 13\n",
            "82\t58\t0.5\t1\t38 20\n",
            "83\t42\t0.5\t1\t28 14\n",
            "84\t53\t0.5\t1\t38 15\n",
            "85\t119\t0.5\t1\t89 30\n",
            "86\t53\t0.5\t1\t36 17\n",
            "87\t58\t0.5\t1\t33 25\n",
            "88\t56\t0.5\t1\t30 26\n",
            "89\t76\t0.5\t1\t55 21\n",
            "90\t41\t0.5\t1\t32 9\n",
            "91\t35\t0.5\t1\t28 7\n",
            "92\t76\t0.5\t1\t55 21\n",
            "93\t77\t0.5\t1\t39 38\n",
            "94\t73\t0.5\t1\t36 37\n",
            "95\t41\t0.5\t1\t22 19\n",
            "96\t51\t0.5\t1\t34 17\n",
            "97\t36\t0.5\t1\t15 21\n",
            "98\t137\t0.5\t1\t80 57\n",
            "99\t23\t0.5\t1\t9 14\n",
            "100\t29\t0.5\t1\t6 23\n",
            "101\t19\t0.5\t1\t7 12\n",
            "\n",
            "RUN STATISTICS FOR INPUT FILE: /content/drive/MyDrive/SRR17159236_2.fastq\n",
            "=============================================\n",
            "32275135 sequences processed in total\n",
            "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
            "\n",
            "Validate paired-end files SRR17159236_1_trimmed.fq and SRR17159236_2_trimmed.fq\n",
            "file_1: SRR17159236_1_trimmed.fq, file_2: SRR17159236_2_trimmed.fq\n",
            "\n",
            "\n",
            ">>>>> Now validing the length of the 2 paired-end infiles: SRR17159236_1_trimmed.fq and SRR17159236_2_trimmed.fq <<<<<\n",
            "Writing validated paired-end Read 1 reads to SRR17159236_1_val_1.fq\n",
            "Writing validated paired-end Read 2 reads to SRR17159236_2_val_2.fq\n",
            "\n",
            "Total number of sequences analysed: 32275135\n",
            "\n",
            "Number of sequence pairs removed because at least one read was shorter than the length cutoff (20 bp): 48352 (0.15%)\n",
            "\n",
            "Deleting both intermediate output files SRR17159236_1_trimmed.fq and SRR17159236_2_trimmed.fq\n",
            "\n",
            "====================================================================================================\n",
            "\n"
          ]
        }
      ],
      "source": [
        "!trim_galore --paired --quality 20 --length 20 /content/drive/MyDrive/SRR17159236_1.fastq /content/drive/MyDrive/SRR17159236_2.fastq -o /content/drive/MyDrive/trimmed_output/\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "-5WLFqAqvGwz"
      },
      "outputs": [],
      "source": [
        "!trim_galore --paired --quality 20 --length 20 /content/drive/MyDrive/SRR17159271_1.fastq /content/drive/MyDrive/SRR17159271_2.fastq -o /content/drive/MyDrive/trimmed_output/\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "k2iGd8b_vGzW",
        "outputId": "1b9094ec-191b-477a-bb18-7252bf3ea1fd"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\r0% [Working]\r            \rGet:1 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease [3,626 B]\n",
            "Get:2 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  InRelease [1,581 B]\n",
            "Get:3 http://security.ubuntu.com/ubuntu jammy-security InRelease [129 kB]\n",
            "Get:4 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  Packages [1,071 kB]\n",
            "Hit:5 http://archive.ubuntu.com/ubuntu jammy InRelease\n",
            "Get:6 http://archive.ubuntu.com/ubuntu jammy-updates InRelease [128 kB]\n",
            "Ign:7 https://r2u.stat.illinois.edu/ubuntu jammy InRelease\n",
            "Hit:8 https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu jammy InRelease\n",
            "Get:9 https://r2u.stat.illinois.edu/ubuntu jammy Release [5,713 B]\n",
            "Hit:10 https://ppa.launchpadcontent.net/graphics-drivers/ppa/ubuntu jammy InRelease\n",
            "Get:11 https://r2u.stat.illinois.edu/ubuntu jammy Release.gpg [793 B]\n",
            "Hit:12 https://ppa.launchpadcontent.net/ubuntugis/ppa/ubuntu jammy InRelease\n",
            "Get:13 http://archive.ubuntu.com/ubuntu jammy-backports InRelease [127 kB]\n",
            "Get:14 https://r2u.stat.illinois.edu/ubuntu jammy/main amd64 Packages [2,602 kB]\n",
            "Get:15 http://archive.ubuntu.com/ubuntu jammy-updates/universe amd64 Packages [1,450 kB]\n",
            "Get:16 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 Packages [2,654 kB]\n",
            "Get:17 https://r2u.stat.illinois.edu/ubuntu jammy/main all Packages [8,428 kB]\n",
            "Fetched 16.6 MB in 5s (3,661 kB/s)\n",
            "Reading package lists... Done\n",
            "W: Skipping acquire of configured file 'main/source/Sources' as repository 'https://r2u.stat.illinois.edu/ubuntu jammy InRelease' does not seem to provide it (sources.list entry misspelt?)\n",
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "The following additional packages will be installed:\n",
            "  bcftools libhts3 libhtscodecs2 python3-hisat2 samtools\n",
            "Suggested packages:\n",
            "  python3-numpy python3-matplotlib texlive-latex-recommended cwltool\n",
            "The following NEW packages will be installed:\n",
            "  bcftools hisat2 libhts3 libhtscodecs2 python3-hisat2 samtools\n",
            "0 upgraded, 6 newly installed, 0 to remove and 50 not upgraded.\n",
            "Need to get 5,505 kB of archives.\n",
            "After this operation, 17.1 MB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libhtscodecs2 amd64 1.1.1-3 [53.2 kB]\n",
            "Get:2 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libhts3 amd64 1.13+ds-2build1 [390 kB]\n",
            "Get:3 http://archive.ubuntu.com/ubuntu jammy/universe amd64 bcftools amd64 1.13-1 [697 kB]\n",
            "Get:4 http://archive.ubuntu.com/ubuntu jammy/universe amd64 hisat2 amd64 2.2.1-3 [3,832 kB]\n",
            "Get:5 http://archive.ubuntu.com/ubuntu jammy/universe amd64 python3-hisat2 all 2.2.1-3 [12.7 kB]\n",
            "Get:6 http://archive.ubuntu.com/ubuntu jammy/universe amd64 samtools amd64 1.13-4 [520 kB]\n",
            "Fetched 5,505 kB in 3s (2,095 kB/s)\n",
            "Selecting previously unselected package libhtscodecs2:amd64.\n",
            "(Reading database ... 124221 files and directories currently installed.)\n",
            "Preparing to unpack .../0-libhtscodecs2_1.1.1-3_amd64.deb ...\n",
            "Unpacking libhtscodecs2:amd64 (1.1.1-3) ...\n",
            "Selecting previously unselected package libhts3:amd64.\n",
            "Preparing to unpack .../1-libhts3_1.13+ds-2build1_amd64.deb ...\n",
            "Unpacking libhts3:amd64 (1.13+ds-2build1) ...\n",
            "Selecting previously unselected package bcftools.\n",
            "Preparing to unpack .../2-bcftools_1.13-1_amd64.deb ...\n",
            "Unpacking bcftools (1.13-1) ...\n",
            "Selecting previously unselected package hisat2.\n",
            "Preparing to unpack .../3-hisat2_2.2.1-3_amd64.deb ...\n",
            "Unpacking hisat2 (2.2.1-3) ...\n",
            "Selecting previously unselected package python3-hisat2.\n",
            "Preparing to unpack .../4-python3-hisat2_2.2.1-3_all.deb ...\n",
            "Unpacking python3-hisat2 (2.2.1-3) ...\n",
            "Selecting previously unselected package samtools.\n",
            "Preparing to unpack .../5-samtools_1.13-4_amd64.deb ...\n",
            "Unpacking samtools (1.13-4) ...\n",
            "Setting up libhtscodecs2:amd64 (1.1.1-3) ...\n",
            "Setting up libhts3:amd64 (1.13+ds-2build1) ...\n",
            "Setting up bcftools (1.13-1) ...\n",
            "Setting up samtools (1.13-4) ...\n",
            "Setting up hisat2 (2.2.1-3) ...\n",
            "Setting up python3-hisat2 (2.2.1-3) ...\n",
            "Processing triggers for man-db (2.10.2-1) ...\n",
            "Processing triggers for libc-bin (2.35-0ubuntu3.4) ...\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_level_zero.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_5.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_opencl.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbb.so.12 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_0.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_loader.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc_proxy.so.2 is not a symbolic link\n",
            "\n"
          ]
        }
      ],
      "source": [
        "!apt-get update\n",
        "!apt-get install -y hisat2"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "66zSDyFyvG3C",
        "outputId": "c340fe73-4f02-44ed-a54c-b86c673d7e9f"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "/bin/bash: line 1: hisat2: command not found\n"
          ]
        }
      ],
      "source": [
        "!hisat2 -x /content/drive/MyDrive/reference_index/GRCm39_index \\\n",
        "       -1 /content/drive/MyDrive/trimmed_output/SRR17159236_1_val_1.fq\\\n",
        "       -2 /content/drive/MyDrive/trimmed_output/SRR17159236_2_val_2.fq \\\n",
        "       -S /content/drive/MyDrive/SRR17159236_aligned.sam\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!hisat2 -x /content/drive/MyDrive/reference_index/GRCm39_index \\\n",
        "       -1 /content/drive/MyDrive/trimmed_output/SRR17158946_1_val_1.fq\\\n",
        "       -2 /content/drive/MyDrive/trimmed_output/SRR17158946_2_val_2.fq \\\n",
        "       -S /content/drive/MyDrive/SRR17158946_aligned.sam"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "k55d37_xpHh-",
        "outputId": "96584f92-ff65-4029-97bb-4e00319bcc0f"
      },
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "26287704 reads; of these:\n",
            "  26287704 (100.00%) were paired; of these:\n",
            "    839435 (3.19%) aligned concordantly 0 times\n",
            "    20748935 (78.93%) aligned concordantly exactly 1 time\n",
            "    4699334 (17.88%) aligned concordantly >1 times\n",
            "    ----\n",
            "    839435 pairs aligned concordantly 0 times; of these:\n",
            "      61759 (7.36%) aligned discordantly 1 time\n",
            "    ----\n",
            "    777676 pairs aligned 0 times concordantly or discordantly; of these:\n",
            "      1555352 mates make up the pairs; of these:\n",
            "        905206 (58.20%) aligned 0 times\n",
            "        492915 (31.69%) aligned exactly 1 time\n",
            "        157231 (10.11%) aligned >1 times\n",
            "98.28% overall alignment rate\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!samtools view -bS /content/drive/MyDrive/SRR17158946_aligned.sam | samtools sort -o /content/drive/MyDrive/SRR17158946_sorted.bam\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "9lT5wVFDXsKD",
        "outputId": "7ac10152-7b7f-47ee-e3a3-61d7642805f4"
      },
      "execution_count": 13,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[bam_sort_core] merging from 30 files and 1 in-memory blocks...\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!samtools index /content/drive/MyDrive/SRR17158946_sorted.bam"
      ],
      "metadata": {
        "id": "3cPV5OHDiPRo"
      },
      "execution_count": 14,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!samtools flagstat /content/drive/MyDrive/SRR17158946_sorted.bam"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "12nWrYi4kjNl",
        "outputId": "636beee4-a58e-47e1-ae91-bb5d437bd5ce"
      },
      "execution_count": 15,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "79217508 + 0 in total (QC-passed reads + QC-failed reads)\n",
            "52575408 + 0 primary\n",
            "26642100 + 0 secondary\n",
            "0 + 0 supplementary\n",
            "0 + 0 duplicates\n",
            "0 + 0 primary duplicates\n",
            "78312302 + 0 mapped (98.86% : N/A)\n",
            "51670202 + 0 primary mapped (98.28% : N/A)\n",
            "52575408 + 0 paired in sequencing\n",
            "26287704 + 0 read1\n",
            "26287704 + 0 read2\n",
            "50896538 + 0 properly paired (96.81% : N/A)\n",
            "51315670 + 0 with itself and mate mapped\n",
            "354532 + 0 singletons (0.67% : N/A)\n",
            "391752 + 0 with mate mapped to a different chr\n",
            "293962 + 0 with mate mapped to a different chr (mapQ>=5)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!featureCounts -a /content/drive/MyDrive/gencode.vM35.basic.annotation.gtf -o /content/drive/MyDrive/SRR17158946_counts.txt -T 4 -p /content/drive/MyDrive/SRR17158946_sorted.bam\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "NRQLpecJkyzl",
        "outputId": "8c709a4a-4e34-4cd1-d85e-1287132ea2af"
      },
      "execution_count": 16,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
            "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
            "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
            "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
            "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
            "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
            "\t  v2.0.3\n",
            "\n",
            "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
            "||  \u001b[0m                                                                          ||\n",
            "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "||                           \u001b[36mSRR17158946_sorted.bam\u001b[0m \u001b[0m                          ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "||             Output file : \u001b[36mSRR17158946_counts.txt\u001b[0m \u001b[0m                          ||\n",
            "||                 Summary : \u001b[36mSRR17158946_counts.txt.summary\u001b[0m \u001b[0m                  ||\n",
            "||              Paired-end : \u001b[36myes\u001b[0m \u001b[0m                                             ||\n",
            "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
            "||              Annotation : \u001b[36mgencode.vM35.basic.annotation.gtf (GTF)\u001b[0m \u001b[0m         ||\n",
            "||      Dir for temp files : \u001b[36m/content/drive/MyDrive\u001b[0m \u001b[0m                          ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "||                 Threads : \u001b[36m4\u001b[0m \u001b[0m                                               ||\n",
            "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
            "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
            "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
            "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "\\\\============================================================================//\n",
            "\n",
            "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
            "||  \u001b[0m                                                                          ||\n",
            "|| Load annotation file gencode.vM35.basic.annotation.gtf \u001b[0m... \u001b[0m                ||\n",
            "||    Features : \u001b[36m532131\u001b[0m \u001b[0m                                                      ||\n",
            "||    Meta-features : \u001b[36m57132\u001b[0m \u001b[0m                                                  ||\n",
            "||    Chromosomes/contigs : \u001b[36m22\u001b[0m \u001b[0m                                               ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "|| Process BAM file SRR17158946_sorted.bam... \u001b[0m                                ||\n",
            "||    Paired-end reads are included. \u001b[0m                                         ||\n",
            "||    The reads are assigned on the single-end mode. \u001b[0m                         ||\n",
            "||    Total alignments : \u001b[36m79217508\u001b[0m \u001b[0m                                            ||\n",
            "||    Successfully assigned alignments : \u001b[36m38129903 (48.1%)\u001b[0m \u001b[0m                    ||\n",
            "||    Running time : \u001b[36m1.49 minutes\u001b[0m \u001b[0m                                            ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "|| Write the final count table. \u001b[0m                                              ||\n",
            "|| Write the read assignment summary. \u001b[0m                                        ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "|| Summary of counting results can be found in file \"/content/drive/MyDrive/ \u001b[0m ||\n",
            "|| SRR17158946_counts.txt.summary\" \u001b[0m                                           ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "\\\\============================================================================//\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!hisat2 -x /content/drive/MyDrive/reference_index/GRCm39_index \\\n",
        "       -1 /content/drive/MyDrive/trimmed_output/SRR17158978_1_val_1.fq\\\n",
        "       -2 /content/drive/MyDrive/trimmed_output/SRR17158978_2_val_2.fq \\\n",
        "       -S /content/drive/MyDrive/SRR17158978_aligned.sam"
      ],
      "metadata": {
        "id": "p_SQJHyqugvs"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "iQwQEoSBvXRT",
        "outputId": "2cf2112f-a89f-4885-89c2-9d2f59b3e7e4"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "samtools is already the newest version (1.13-4).\n",
            "0 upgraded, 0 newly installed, 0 to remove and 49 not upgraded.\n"
          ]
        }
      ],
      "source": [
        "!apt-get install -y samtools"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "3DwSpz6EvXTj",
        "outputId": "319b1022-89ae-4b49-cb9c-d3928e619169"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "[bam_sort_core] merging from 29 files and 1 in-memory blocks...\n"
          ]
        }
      ],
      "source": [
        "!samtools view -bS /content/drive/MyDrive/SRR17159236_aligned.sam | samtools sort -o /content/drive/MyDrive/SRR17159236_sorted.bam\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "j5_zuSMMvXWZ"
      },
      "outputs": [],
      "source": [
        "!samtools index /content/drive/MyDrive/SRR17159236_sorted.bam"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!samtools index /content/drive/MyDrive/SRR17159271_sorted.bam"
      ],
      "metadata": {
        "id": "bBhukuf_iVUs"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!samtools flagstat /content/drive/MyDrive/SRR17159271_sorted.bam"
      ],
      "metadata": {
        "id": "-AEnAbYbiVYc"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "pUslN-YLvXa3",
        "outputId": "e656b476-5490-4678-dd7b-39114ebbb0dd"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "77413698 + 0 in total (QC-passed reads + QC-failed reads)\n",
            "64453566 + 0 primary\n",
            "12960132 + 0 secondary\n",
            "0 + 0 supplementary\n",
            "0 + 0 duplicates\n",
            "0 + 0 primary duplicates\n",
            "76837208 + 0 mapped (99.26% : N/A)\n",
            "63877076 + 0 primary mapped (99.11% : N/A)\n",
            "64453566 + 0 paired in sequencing\n",
            "32226783 + 0 read1\n",
            "32226783 + 0 read2\n",
            "63229614 + 0 properly paired (98.10% : N/A)\n",
            "63506552 + 0 with itself and mate mapped\n",
            "370524 + 0 singletons (0.57% : N/A)\n",
            "246752 + 0 with mate mapped to a different chr\n",
            "191859 + 0 with mate mapped to a different chr (mapQ>=5)\n"
          ]
        }
      ],
      "source": [
        "!samtools flagstat /content/drive/MyDrive/SRR17159236_sorted.bam"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "v3g3VZPOvsHz"
      },
      "outputs": [],
      "source": [
        "!samtools view -bS /content/drive/MyDrive/SRR17159271_aligned.sam | samtools sort -o /content/drive/MyDrive/SRR17159271_sorted.bam"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "b0ipmWgKvsNW"
      },
      "outputs": [],
      "source": [
        "!samtools index /content/drive/MyDrive/SRR17159271_sorted.bam"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "QQhxA8nPvsP_"
      },
      "outputs": [],
      "source": [
        "!samtools flagstat /content/drive/MyDrive/SRR17159271_sorted.bam"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "3W82DqKov1UW",
        "outputId": "1eebfa87-39d0-4e8b-9969-8b9f58acc93a"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "The following NEW packages will be installed:\n",
            "  subread\n",
            "0 upgraded, 1 newly installed, 0 to remove and 50 not upgraded.\n",
            "Need to get 565 kB of archives.\n",
            "After this operation, 1,928 kB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 subread amd64 2.0.3+dfsg-1 [565 kB]\n",
            "Fetched 565 kB in 2s (353 kB/s)\n",
            "Selecting previously unselected package subread.\n",
            "(Reading database ... 124421 files and directories currently installed.)\n",
            "Preparing to unpack .../subread_2.0.3+dfsg-1_amd64.deb ...\n",
            "Unpacking subread (2.0.3+dfsg-1) ...\n",
            "Setting up subread (2.0.3+dfsg-1) ...\n",
            "Processing triggers for man-db (2.10.2-1) ...\n"
          ]
        }
      ],
      "source": [
        "!apt-get install -y subread"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "cIRXFZMHv1WZ",
        "outputId": "b5914799-f3cb-499b-9ed5-a42323fa86bf"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "\n",
            "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
            "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
            "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
            "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
            "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
            "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
            "\t  v2.0.3\n",
            "\n",
            "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
            "||  \u001b[0m                                                                          ||\n",
            "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "||                           \u001b[36mSRR17159236_sorted.bam\u001b[0m \u001b[0m                          ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "||             Output file : \u001b[36mSRR17159236_counts.txt\u001b[0m \u001b[0m                          ||\n",
            "||                 Summary : \u001b[36mSRR17159236_counts.txt.summary\u001b[0m \u001b[0m                  ||\n",
            "||              Paired-end : \u001b[36myes\u001b[0m \u001b[0m                                             ||\n",
            "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
            "||              Annotation : \u001b[36mgencode.vM35.basic.annotation.gtf (GTF)\u001b[0m \u001b[0m         ||\n",
            "||      Dir for temp files : \u001b[36m/content/drive/MyDrive\u001b[0m \u001b[0m                          ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "||                 Threads : \u001b[36m4\u001b[0m \u001b[0m                                               ||\n",
            "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
            "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
            "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
            "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "\\\\============================================================================//\n",
            "\n",
            "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
            "||  \u001b[0m                                                                          ||\n",
            "|| Load annotation file gencode.vM35.basic.annotation.gtf \u001b[0m... \u001b[0m                ||\n",
            "||    Features : \u001b[36m532131\u001b[0m \u001b[0m                                                      ||\n",
            "||    Meta-features : \u001b[36m57132\u001b[0m \u001b[0m                                                  ||\n",
            "||    Chromosomes/contigs : \u001b[36m22\u001b[0m \u001b[0m                                               ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "|| Process BAM file SRR17159236_sorted.bam... \u001b[0m                                ||\n",
            "||    Paired-end reads are included. \u001b[0m                                         ||\n",
            "||    The reads are assigned on the single-end mode. \u001b[0m                         ||\n",
            "||    Total alignments : \u001b[36m77413698\u001b[0m \u001b[0m                                            ||\n",
            "||    Successfully assigned alignments : \u001b[36m48828682 (63.1%)\u001b[0m \u001b[0m                    ||\n",
            "||    Running time : \u001b[36m1.65 minutes\u001b[0m \u001b[0m                                            ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "|| Write the final count table. \u001b[0m                                              ||\n",
            "|| Write the read assignment summary. \u001b[0m                                        ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "|| Summary of counting results can be found in file \"/content/drive/MyDrive/ \u001b[0m ||\n",
            "|| SRR17159236_counts.txt.summary\" \u001b[0m                                           ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "\\\\============================================================================//\n",
            "\n"
          ]
        }
      ],
      "source": [
        "!featureCounts -a /content/drive/MyDrive/gencode.vM35.basic.annotation.gtf -o /content/drive/MyDrive/SRR17159236_counts.txt -T 4 -p /content/drive/MyDrive/SRR17159236_sorted.bam\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "WD-EMXV8v1aE",
        "outputId": "3ce6ba6d-79a0-47c8-bb85-9122fad4af78"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Started analysis of SRR17159276_2.fastq\n",
            "Approx 5% complete for SRR17159276_2.fastq\n",
            "Approx 10% complete for SRR17159276_2.fastq\n",
            "Approx 15% complete for SRR17159276_2.fastq\n",
            "Approx 20% complete for SRR17159276_2.fastq\n",
            "Approx 25% complete for SRR17159276_2.fastq\n",
            "Approx 30% complete for SRR17159276_2.fastq\n",
            "Approx 35% complete for SRR17159276_2.fastq\n",
            "Approx 40% complete for SRR17159276_2.fastq\n",
            "Approx 45% complete for SRR17159276_2.fastq\n",
            "Approx 50% complete for SRR17159276_2.fastq\n",
            "Approx 55% complete for SRR17159276_2.fastq\n",
            "Approx 60% complete for SRR17159276_2.fastq\n",
            "Approx 65% complete for SRR17159276_2.fastq\n",
            "Approx 70% complete for SRR17159276_2.fastq\n",
            "Approx 75% complete for SRR17159276_2.fastq\n",
            "Approx 80% complete for SRR17159276_2.fastq\n",
            "Approx 85% complete for SRR17159276_2.fastq\n",
            "Approx 90% complete for SRR17159276_2.fastq\n",
            "Approx 95% complete for SRR17159276_2.fastq\n",
            "Analysis complete for SRR17159276_2.fastq\n"
          ]
        }
      ],
      "source": [
        "!fastqc /content/drive/MyDrive/SRR17159276_2.fastq -o /content/drive/MyDrive/QC_reports/"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "MkJfYmwvl4dS",
        "outputId": "63b0195d-f27e-42a9-b519-d06520ebb181"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Started analysis of SRR17158978_2.fastq\n",
            "Approx 5% complete for SRR17158978_2.fastq\n",
            "Approx 10% complete for SRR17158978_2.fastq\n",
            "Approx 15% complete for SRR17158978_2.fastq\n",
            "Approx 20% complete for SRR17158978_2.fastq\n",
            "Approx 25% complete for SRR17158978_2.fastq\n",
            "Approx 30% complete for SRR17158978_2.fastq\n",
            "Approx 35% complete for SRR17158978_2.fastq\n",
            "Approx 40% complete for SRR17158978_2.fastq\n",
            "Approx 45% complete for SRR17158978_2.fastq\n",
            "Approx 50% complete for SRR17158978_2.fastq\n",
            "Approx 55% complete for SRR17158978_2.fastq\n",
            "Approx 60% complete for SRR17158978_2.fastq\n",
            "Approx 65% complete for SRR17158978_2.fastq\n",
            "Approx 70% complete for SRR17158978_2.fastq\n",
            "Approx 75% complete for SRR17158978_2.fastq\n",
            "Approx 80% complete for SRR17158978_2.fastq\n",
            "Approx 85% complete for SRR17158978_2.fastq\n",
            "Approx 90% complete for SRR17158978_2.fastq\n",
            "Approx 95% complete for SRR17158978_2.fastq\n",
            "Analysis complete for SRR17158978_2.fastq\n"
          ]
        }
      ],
      "source": [
        "!fastqc /content/drive/MyDrive/SRR17158978_2.fastq -o /content/drive/MyDrive/QC_reports/"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "jNzmLtuUl4aY",
        "outputId": "e196a1a4-bb54-445c-a9fc-123f1fb06cee"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Multicore support not enabled. Proceeding with single-core trimming.\n",
            "Path to Cutadapt set as: 'cutadapt' (default)\n",
            "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
            "Cutadapt version: 4.9\n",
            "single-core operation.\n",
            "igzip command line interface 2.30.0\n",
            "igzip detected. Using igzip for decompressing\n",
            "\n",
            "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
            "\n",
            "Output will be written into the directory: /content/drive/MyDrive/trimmed_output/\n",
            "\n",
            "\n",
            "AUTO-DETECTING ADAPTER TYPE\n",
            "===========================\n",
            "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> /content/drive/MyDrive/SRR17159238_1.fastq <<)\n",
            "\n",
            "Found perfect matches for the following adapter sequences:\n",
            "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
            "Illumina\t32723\tAGATCGGAAGAGC\t1000000\t3.27\n",
            "smallRNA\t0\tTGGAATTCTCGG\t1000000\t0.00\n",
            "Nextera\t0\tCTGTCTCTTATA\t1000000\t0.00\n",
            "Using Illumina adapter for trimming (count: 32723). Second best hit was smallRNA (count: 0)\n",
            "\n",
            "Writing report to '/content/drive/MyDrive/trimmed_output/SRR17159238_1.fastq_trimming_report.txt'\n",
            "\n",
            "SUMMARISING RUN PARAMETERS\n",
            "==========================\n",
            "Input filename: /content/drive/MyDrive/SRR17159238_1.fastq\n",
            "Trimming mode: paired-end\n",
            "Trim Galore version: 0.6.10\n",
            "Cutadapt version: 4.9\n",
            "Number of cores used for trimming: 1\n",
            "Quality Phred score cutoff: 20\n",
            "Quality encoding type selected: ASCII+33\n",
            "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
            "Maximum trimming error rate: 0.1 (default)\n",
            "Minimum required adapter overlap (stringency): 1 bp\n",
            "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
            "\n",
            "Cutadapt seems to be fairly up-to-date (version 4.9). Setting -j 1\n",
            "Writing final adapter and quality trimmed output to SRR17159238_1_trimmed.fq\n",
            "\n",
            "\n",
            "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /content/drive/MyDrive/SRR17159238_1.fastq <<< \n",
            "10000000 sequences processed\n",
            "20000000 sequences processed\n",
            "30000000 sequences processed\n",
            "This is cutadapt 4.9 with Python 3.10.12\n",
            "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /content/drive/MyDrive/SRR17159238_1.fastq\n",
            "Processing single-end reads on 1 core ...\n",
            "Finished in 492.590 s (12.422 µs/read; 4.83 M reads/minute).\n",
            "\n",
            "=== Summary ===\n",
            "\n",
            "Total reads processed:              39,655,191\n",
            "Reads with adapters:                15,157,996 (38.2%)\n",
            "Reads written (passing filters):    39,655,191 (100.0%)\n",
            "\n",
            "Total basepairs processed: 4,005,174,291 bp\n",
            "Quality-trimmed:                 955,811 bp (0.0%)\n",
            "Total written (filtered):  3,940,604,305 bp (98.4%)\n",
            "\n",
            "=== Adapter 1 ===\n",
            "\n",
            "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 15157996 times\n",
            "\n",
            "Minimum overlap: 1\n",
            "No. of allowed errors:\n",
            "1-9 bp: 0; 10-13 bp: 1\n",
            "\n",
            "Bases preceding removed adapters:\n",
            "  A: 25.7%\n",
            "  C: 29.6%\n",
            "  G: 23.9%\n",
            "  T: 20.5%\n",
            "  none/other: 0.3%\n",
            "\n",
            "Overview of removed sequences\n",
            "length\tcount\texpect\tmax.err\terror counts\n",
            "1\t8762043\t9913797.8\t0\t8762043\n",
            "2\t2938210\t2478449.4\t0\t2938210\n",
            "3\t851809\t619612.4\t0\t851809\n",
            "4\t286443\t154903.1\t0\t286443\n",
            "5\t162501\t38725.8\t0\t162501\n",
            "6\t130657\t9681.4\t0\t130657\n",
            "7\t120850\t2420.4\t0\t120850\n",
            "8\t114300\t605.1\t0\t114300\n",
            "9\t104657\t151.3\t0\t104158 499\n",
            "10\t99270\t37.8\t1\t97037 2233\n",
            "11\t94266\t9.5\t1\t92071 2195\n",
            "12\t89582\t2.4\t1\t87840 1742\n",
            "13\t88685\t0.6\t1\t86985 1700\n",
            "14\t85780\t0.6\t1\t83960 1820\n",
            "15\t82826\t0.6\t1\t81169 1657\n",
            "16\t79147\t0.6\t1\t77479 1668\n",
            "17\t75319\t0.6\t1\t73652 1667\n",
            "18\t68347\t0.6\t1\t66895 1452\n",
            "19\t61929\t0.6\t1\t60664 1265\n",
            "20\t57575\t0.6\t1\t56354 1221\n",
            "21\t54256\t0.6\t1\t53089 1167\n",
            "22\t50707\t0.6\t1\t49772 935\n",
            "23\t48554\t0.6\t1\t47572 982\n",
            "24\t47488\t0.6\t1\t46459 1029\n",
            "25\t45763\t0.6\t1\t44869 894\n",
            "26\t43412\t0.6\t1\t42542 870\n",
            "27\t41110\t0.6\t1\t40198 912\n",
            "28\t37937\t0.6\t1\t37183 754\n",
            "29\t34331\t0.6\t1\t33558 773\n",
            "30\t30957\t0.6\t1\t30255 702\n",
            "31\t27375\t0.6\t1\t26805 570\n",
            "32\t25831\t0.6\t1\t25286 545\n",
            "33\t23689\t0.6\t1\t23180 509\n",
            "34\t21727\t0.6\t1\t21267 460\n",
            "35\t20858\t0.6\t1\t20379 479\n",
            "36\t20002\t0.6\t1\t19519 483\n",
            "37\t18498\t0.6\t1\t18102 396\n",
            "38\t17418\t0.6\t1\t17040 378\n",
            "39\t15901\t0.6\t1\t15537 364\n",
            "40\t14709\t0.6\t1\t14416 293\n",
            "41\t13523\t0.6\t1\t13193 330\n",
            "42\t12038\t0.6\t1\t11747 291\n",
            "43\t10386\t0.6\t1\t10078 308\n",
            "44\t8642\t0.6\t1\t8433 209\n",
            "45\t14534\t0.6\t1\t14219 315\n",
            "46\t1621\t0.6\t1\t1534 87\n",
            "47\t5922\t0.6\t1\t5802 120\n",
            "48\t12969\t0.6\t1\t12700 269\n",
            "49\t5666\t0.6\t1\t5502 164\n",
            "50\t2374\t0.6\t1\t2267 107\n",
            "51\t7127\t0.6\t1\t6995 132\n",
            "52\t2811\t0.6\t1\t2716 95\n",
            "53\t2039\t0.6\t1\t1975 64\n",
            "54\t4301\t0.6\t1\t4202 99\n",
            "55\t4841\t0.6\t1\t4730 111\n",
            "56\t1404\t0.6\t1\t1352 52\n",
            "57\t1338\t0.6\t1\t1264 74\n",
            "58\t514\t0.6\t1\t464 50\n",
            "59\t2112\t0.6\t1\t2045 67\n",
            "60\t279\t0.6\t1\t223 56\n",
            "61\t191\t0.6\t1\t147 44\n",
            "62\t1503\t0.6\t1\t1442 61\n",
            "63\t480\t0.6\t1\t447 33\n",
            "64\t107\t0.6\t1\t79 28\n",
            "65\t473\t0.6\t1\t420 53\n",
            "66\t657\t0.6\t1\t609 48\n",
            "67\t73\t0.6\t1\t62 11\n",
            "68\t324\t0.6\t1\t284 40\n",
            "69\t304\t0.6\t1\t281 23\n",
            "70\t295\t0.6\t1\t255 40\n",
            "71\t33\t0.6\t1\t11 22\n",
            "72\t23\t0.6\t1\t8 15\n",
            "73\t99\t0.6\t1\t51 48\n",
            "74\t105\t0.6\t1\t83 22\n",
            "75\t78\t0.6\t1\t49 29\n",
            "76\t70\t0.6\t1\t50 20\n",
            "77\t68\t0.6\t1\t43 25\n",
            "78\t89\t0.6\t1\t68 21\n",
            "79\t78\t0.6\t1\t47 31\n",
            "80\t98\t0.6\t1\t41 57\n",
            "81\t71\t0.6\t1\t47 24\n",
            "82\t81\t0.6\t1\t50 31\n",
            "83\t87\t0.6\t1\t46 41\n",
            "84\t78\t0.6\t1\t56 22\n",
            "85\t161\t0.6\t1\t132 29\n",
            "86\t85\t0.6\t1\t60 25\n",
            "87\t103\t0.6\t1\t68 35\n",
            "88\t106\t0.6\t1\t64 42\n",
            "89\t95\t0.6\t1\t70 25\n",
            "90\t73\t0.6\t1\t54 19\n",
            "91\t55\t0.6\t1\t38 17\n",
            "92\t104\t0.6\t1\t77 27\n",
            "93\t125\t0.6\t1\t107 18\n",
            "94\t115\t0.6\t1\t99 16\n",
            "95\t97\t0.6\t1\t52 45\n",
            "96\t95\t0.6\t1\t54 41\n",
            "97\t53\t0.6\t1\t37 16\n",
            "98\t183\t0.6\t1\t128 55\n",
            "99\t65\t0.6\t1\t24 41\n",
            "100\t110\t0.6\t1\t26 84\n",
            "101\t41746\t0.6\t1\t14 41732\n",
            "\n",
            "RUN STATISTICS FOR INPUT FILE: /content/drive/MyDrive/SRR17159238_1.fastq\n",
            "=============================================\n",
            "39655191 sequences processed in total\n",
            "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
            "\n",
            "Writing report to '/content/drive/MyDrive/trimmed_output/SRR17159238_2.fastq_trimming_report.txt'\n",
            "\n",
            "SUMMARISING RUN PARAMETERS\n",
            "==========================\n",
            "Input filename: /content/drive/MyDrive/SRR17159238_2.fastq\n",
            "Trimming mode: paired-end\n",
            "Trim Galore version: 0.6.10\n",
            "Cutadapt version: 4.9\n",
            "Number of cores used for trimming: 1\n",
            "Quality Phred score cutoff: 20\n",
            "Quality encoding type selected: ASCII+33\n",
            "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
            "Maximum trimming error rate: 0.1 (default)\n",
            "Minimum required adapter overlap (stringency): 1 bp\n",
            "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
            "\n",
            "Cutadapt seems to be fairly up-to-date (version 4.9). Setting -j -j 1\n",
            "Writing final adapter and quality trimmed output to SRR17159238_2_trimmed.fq\n",
            "\n",
            "\n",
            "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /content/drive/MyDrive/SRR17159238_2.fastq <<< \n",
            "10000000 sequences processed\n",
            "20000000 sequences processed\n",
            "30000000 sequences processed\n",
            "This is cutadapt 4.9 with Python 3.10.12\n",
            "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /content/drive/MyDrive/SRR17159238_2.fastq\n",
            "Processing single-end reads on 1 core ...\n",
            "Finished in 639.567 s (16.128 µs/read; 3.72 M reads/minute).\n",
            "\n",
            "=== Summary ===\n",
            "\n",
            "Total reads processed:              39,655,191\n",
            "Reads with adapters:                15,977,862 (40.3%)\n",
            "Reads written (passing filters):    39,655,191 (100.0%)\n",
            "\n",
            "Total basepairs processed: 4,005,174,291 bp\n",
            "Quality-trimmed:              12,479,090 bp (0.3%)\n",
            "Total written (filtered):  3,932,727,290 bp (98.2%)\n",
            "\n",
            "=== Adapter 1 ===\n",
            "\n",
            "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 15977862 times\n",
            "\n",
            "Minimum overlap: 1\n",
            "No. of allowed errors:\n",
            "1-9 bp: 0; 10-13 bp: 1\n",
            "\n",
            "Bases preceding removed adapters:\n",
            "  A: 27.1%\n",
            "  C: 32.5%\n",
            "  G: 22.9%\n",
            "  T: 17.5%\n",
            "  none/other: 0.0%\n",
            "\n",
            "Overview of removed sequences\n",
            "length\tcount\texpect\tmax.err\terror counts\n",
            "1\t9647883\t9913797.8\t0\t9647883\n",
            "2\t2818966\t2478449.4\t0\t2818966\n",
            "3\t960317\t619612.4\t0\t960317\n",
            "4\t280494\t154903.1\t0\t280494\n",
            "5\t166289\t38725.8\t0\t166289\n",
            "6\t134868\t9681.4\t0\t134868\n",
            "7\t122087\t2420.4\t0\t122087\n",
            "8\t117858\t605.1\t0\t117858\n",
            "9\t100655\t151.3\t0\t99724 931\n",
            "10\t102223\t37.8\t1\t100530 1693\n",
            "11\t90897\t9.5\t1\t88838 2059\n",
            "12\t89817\t2.4\t1\t88573 1244\n",
            "13\t87634\t0.6\t1\t86772 862\n",
            "14\t90993\t0.6\t1\t89966 1027\n",
            "15\t77041\t0.6\t1\t76259 782\n",
            "16\t76764\t0.6\t1\t76065 699\n",
            "17\t88211\t0.6\t1\t87286 925\n",
            "18\t52241\t0.6\t1\t51716 525\n",
            "19\t71313\t0.6\t1\t70505 808\n",
            "20\t52134\t0.6\t1\t51589 545\n",
            "21\t49330\t0.6\t1\t48835 495\n",
            "22\t50382\t0.6\t1\t49897 485\n",
            "23\t47884\t0.6\t1\t47400 484\n",
            "24\t46915\t0.6\t1\t46436 479\n",
            "25\t46169\t0.6\t1\t45723 446\n",
            "26\t42694\t0.6\t1\t42237 457\n",
            "27\t41057\t0.6\t1\t40545 512\n",
            "28\t38536\t0.6\t1\t38212 324\n",
            "29\t32388\t0.6\t1\t32049 339\n",
            "30\t31124\t0.6\t1\t30844 280\n",
            "31\t26612\t0.6\t1\t26360 252\n",
            "32\t24990\t0.6\t1\t24773 217\n",
            "33\t23714\t0.6\t1\t23467 247\n",
            "34\t22455\t0.6\t1\t22182 273\n",
            "35\t21004\t0.6\t1\t20751 253\n",
            "36\t19007\t0.6\t1\t18828 179\n",
            "37\t18001\t0.6\t1\t17844 157\n",
            "38\t18121\t0.6\t1\t17955 166\n",
            "39\t15808\t0.6\t1\t15642 166\n",
            "40\t15690\t0.6\t1\t15557 133\n",
            "41\t11485\t0.6\t1\t11347 138\n",
            "42\t12263\t0.6\t1\t12143 120\n",
            "43\t9011\t0.6\t1\t8937 74\n",
            "44\t8779\t0.6\t1\t8679 100\n",
            "45\t8778\t0.6\t1\t8688 90\n",
            "46\t7181\t0.6\t1\t7098 83\n",
            "47\t11754\t0.6\t1\t11655 99\n",
            "48\t2862\t0.6\t1\t2803 59\n",
            "49\t7404\t0.6\t1\t7308 96\n",
            "50\t3679\t0.6\t1\t3623 56\n",
            "51\t3734\t0.6\t1\t3678 56\n",
            "52\t2829\t0.6\t1\t2782 47\n",
            "53\t3516\t0.6\t1\t3468 48\n",
            "54\t4250\t0.6\t1\t4212 38\n",
            "55\t1120\t0.6\t1\t1086 34\n",
            "56\t2126\t0.6\t1\t2082 44\n",
            "57\t4645\t0.6\t1\t4591 54\n",
            "58\t988\t0.6\t1\t958 30\n",
            "59\t696\t0.6\t1\t671 25\n",
            "60\t2638\t0.6\t1\t2591 47\n",
            "61\t589\t0.6\t1\t550 39\n",
            "62\t535\t0.6\t1\t511 24\n",
            "63\t1874\t0.6\t1\t1836 38\n",
            "64\t535\t0.6\t1\t506 29\n",
            "65\t236\t0.6\t1\t208 28\n",
            "66\t275\t0.6\t1\t258 17\n",
            "67\t275\t0.6\t1\t257 18\n",
            "68\t227\t0.6\t1\t194 33\n",
            "69\t230\t0.6\t1\t214 16\n",
            "70\t265\t0.6\t1\t238 27\n",
            "71\t158\t0.6\t1\t136 22\n",
            "72\t116\t0.6\t1\t95 21\n",
            "73\t104\t0.6\t1\t90 14\n",
            "74\t66\t0.6\t1\t46 20\n",
            "75\t72\t0.6\t1\t36 36\n",
            "76\t39\t0.6\t1\t19 20\n",
            "77\t52\t0.6\t1\t33 19\n",
            "78\t69\t0.6\t1\t52 17\n",
            "79\t57\t0.6\t1\t37 20\n",
            "80\t65\t0.6\t1\t38 27\n",
            "81\t66\t0.6\t1\t44 22\n",
            "82\t79\t0.6\t1\t46 33\n",
            "83\t64\t0.6\t1\t43 21\n",
            "84\t79\t0.6\t1\t56 23\n",
            "85\t148\t0.6\t1\t127 21\n",
            "86\t87\t0.6\t1\t59 28\n",
            "87\t87\t0.6\t1\t63 24\n",
            "88\t85\t0.6\t1\t55 30\n",
            "89\t84\t0.6\t1\t63 21\n",
            "90\t68\t0.6\t1\t53 15\n",
            "91\t46\t0.6\t1\t35 11\n",
            "92\t96\t0.6\t1\t75 21\n",
            "93\t133\t0.6\t1\t74 59\n",
            "94\t114\t0.6\t1\t48 66\n",
            "95\t56\t0.6\t1\t36 20\n",
            "96\t75\t0.6\t1\t46 29\n",
            "97\t47\t0.6\t1\t28 19\n",
            "98\t188\t0.6\t1\t95 93\n",
            "99\t42\t0.6\t1\t20 22\n",
            "100\t49\t0.6\t1\t5 44\n",
            "101\t26\t0.6\t1\t9 17\n",
            "\n",
            "RUN STATISTICS FOR INPUT FILE: /content/drive/MyDrive/SRR17159238_2.fastq\n",
            "=============================================\n",
            "39655191 sequences processed in total\n",
            "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
            "\n",
            "Validate paired-end files SRR17159238_1_trimmed.fq and SRR17159238_2_trimmed.fq\n",
            "file_1: SRR17159238_1_trimmed.fq, file_2: SRR17159238_2_trimmed.fq\n",
            "\n",
            "\n",
            ">>>>> Now validing the length of the 2 paired-end infiles: SRR17159238_1_trimmed.fq and SRR17159238_2_trimmed.fq <<<<<\n",
            "Writing validated paired-end Read 1 reads to SRR17159238_1_val_1.fq\n",
            "Writing validated paired-end Read 2 reads to SRR17159238_2_val_2.fq\n",
            "\n",
            "Total number of sequences analysed: 39655191\n",
            "\n",
            "Number of sequence pairs removed because at least one read was shorter than the length cutoff (20 bp): 132871 (0.34%)\n",
            "\n",
            "Deleting both intermediate output files SRR17159238_1_trimmed.fq and SRR17159238_2_trimmed.fq\n",
            "\n",
            "====================================================================================================\n",
            "\n"
          ]
        }
      ],
      "source": [
        "!trim_galore --paired --quality 20 --length 20 /content/drive/MyDrive/SRR17159238_1.fastq /content/drive/MyDrive/SRR17159238_2.fastq -o /content/drive/MyDrive/trimmed_output/\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "7HTOzq_-l4Rg",
        "outputId": "aa09cdc5-bb0c-41bd-ed8c-a9fec02fd4e9"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "39522320 reads; of these:\n",
            "  39522320 (100.00%) were paired; of these:\n",
            "    852828 (2.16%) aligned concordantly 0 times\n",
            "    31906198 (80.73%) aligned concordantly exactly 1 time\n",
            "    6763294 (17.11%) aligned concordantly >1 times\n",
            "    ----\n",
            "    852828 pairs aligned concordantly 0 times; of these:\n",
            "      65714 (7.71%) aligned discordantly 1 time\n",
            "    ----\n",
            "    787114 pairs aligned 0 times concordantly or discordantly; of these:\n",
            "      1574228 mates make up the pairs; of these:\n",
            "        781927 (49.67%) aligned 0 times\n",
            "        614838 (39.06%) aligned exactly 1 time\n",
            "        177463 (11.27%) aligned >1 times\n",
            "99.01% overall alignment rate\n"
          ]
        }
      ],
      "source": [
        "!hisat2 -x /content/drive/MyDrive/reference_index/GRCm39_index \\\n",
        "       -1 /content/drive/MyDrive/trimmed_output/SRR17159238_1_val_1.fq \\\n",
        "       -2 /content/drive/MyDrive/trimmed_output/SRR17159238_2_val_2.fq \\\n",
        "       -S /content/drive/MyDrive/SRR17159238_aligned.sam\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "QMECKOIW_HbF"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "39522320 reads; of these:\n",
        "  39522320 (100.00%) were paired; of these:\n",
        "    852828 (2.16%) aligned concordantly 0 times\n",
        "    31906198 (80.73%) aligned concordantly exactly 1 time\n",
        "    6763294 (17.11%) aligned concordantly >1 times\n",
        "    ----\n",
        "    852828 pairs aligned concordantly 0 times; of these:\n",
        "      65714 (7.71%) aligned discordantly 1 time\n",
        "    ----\n",
        "    787114 pairs aligned 0 times concordantly or discordantly; of these:\n",
        "      1574228 mates make up the pairs; of these:\n",
        "        781927 (49.67%) aligned 0 times\n",
        "        614838 (39.06%) aligned exactly 1 time\n",
        "        177463 (11.27%) aligned >1 times\n",
        "99.01% overall alignment rate"
      ],
      "metadata": {
        "id": "nEdU9TV--OaD"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!hisat2 -x /content/drive/MyDrive/reference_index/GRCm39_index \\\n",
        "       -1 /content/drive/MyDrive/trimmed_output/SRR17158978_1_val_1.fq \\\n",
        "       -2 /content/drive/MyDrive/trimmed_output/SRR17158978_2_val_2.fq \\\n",
        "       -S /content/drive/MyDrive/SRR17158978_aligned.sam"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "FmOD7NWO_L9x",
        "outputId": "ce4aeb56-e20a-4288-d250-224777035368"
      },
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "18789132 reads; of these:\n",
            "  18789132 (100.00%) were paired; of these:\n",
            "    452122 (2.41%) aligned concordantly 0 times\n",
            "    14660851 (78.03%) aligned concordantly exactly 1 time\n",
            "    3676159 (19.57%) aligned concordantly >1 times\n",
            "    ----\n",
            "    452122 pairs aligned concordantly 0 times; of these:\n",
            "      35346 (7.82%) aligned discordantly 1 time\n",
            "    ----\n",
            "    416776 pairs aligned 0 times concordantly or discordantly; of these:\n",
            "      833552 mates make up the pairs; of these:\n",
            "        501903 (60.21%) aligned 0 times\n",
            "        251503 (30.17%) aligned exactly 1 time\n",
            "        80146 (9.61%) aligned >1 times\n",
            "98.66% overall alignment rate\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Wz4ZCA3xDgdQ",
        "outputId": "baf8dd69-1cc4-4fad-8464-2f46112ceb50"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[bam_sort_core] merging from 36 files and 1 in-memory blocks...\n"
          ]
        }
      ],
      "source": [
        "!samtools view -bS /content/drive/MyDrive/SRR17159238_aligned.sam | samtools sort -o /content/drive/MyDrive/SRR17159238_sorted.bam\n"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "ip9S6peeXUIa"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!samtools view -bS /content/drive/MyDrive/SRR17158978_aligned.sam | samtools sort -o /content/drive/MyDrive/SRR17158978_sorted.bam\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "2ctAFUqcXS86",
        "outputId": "a44ade3c-c8b4-4aa4-e26f-d25c535fca67"
      },
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[bam_sort_core] merging from 22 files and 1 in-memory blocks...\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!samtools index /content/drive/MyDrive/SRR17158978_sorted.bam"
      ],
      "metadata": {
        "id": "gEUDhbgzd95v"
      },
      "execution_count": 9,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!samtools flagstat /content/drive/MyDrive/SRR17158978_sorted.bam"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "FoziVReoeGmd",
        "outputId": "569ab0b5-4051-45cf-b169-02ffdf8562ba"
      },
      "execution_count": 10,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "58309766 + 0 in total (QC-passed reads + QC-failed reads)\n",
            "37578264 + 0 primary\n",
            "20731502 + 0 secondary\n",
            "0 + 0 supplementary\n",
            "0 + 0 duplicates\n",
            "0 + 0 primary duplicates\n",
            "57807863 + 0 mapped (99.14% : N/A)\n",
            "37076361 + 0 primary mapped (98.66% : N/A)\n",
            "37578264 + 0 paired in sequencing\n",
            "18789132 + 0 read1\n",
            "18789132 + 0 read2\n",
            "36674020 + 0 properly paired (97.59% : N/A)\n",
            "36893632 + 0 with itself and mate mapped\n",
            "182729 + 0 singletons (0.49% : N/A)\n",
            "201706 + 0 with mate mapped to a different chr\n",
            "153554 + 0 with mate mapped to a different chr (mapQ>=5)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "vkhOKBnjDuXo"
      },
      "outputs": [],
      "source": [
        "!samtools index /content/drive/MyDrive/SRR17159238_sorted.bam"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "eOzn-6O6Dzr8",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "82606b44-0411-4d28-be14-51172823af5f"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "95387991 + 0 in total (QC-passed reads + QC-failed reads)\n",
            "79044640 + 0 primary\n",
            "16343351 + 0 secondary\n",
            "0 + 0 supplementary\n",
            "0 + 0 duplicates\n",
            "0 + 0 primary duplicates\n",
            "94606064 + 0 mapped (99.18% : N/A)\n",
            "78262713 + 0 primary mapped (99.01% : N/A)\n",
            "79044640 + 0 paired in sequencing\n",
            "39522320 + 0 read1\n",
            "39522320 + 0 read2\n",
            "77338984 + 0 properly paired (97.84% : N/A)\n",
            "77687442 + 0 with itself and mate mapped\n",
            "575271 + 0 singletons (0.73% : N/A)\n",
            "304454 + 0 with mate mapped to a different chr\n",
            "222808 + 0 with mate mapped to a different chr (mapQ>=5)\n"
          ]
        }
      ],
      "source": [
        "!samtools flagstat /content/drive/MyDrive/SRR17159238_sorted.bam"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!featureCounts -a /content/drive/MyDrive/gencode.vM35.basic.annotation.gtf -o /content/drive/MyDrive/SRR17158978_counts.txt -T 4 -p /content/drive/MyDrive/SRR17158978_sorted.bam\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "8FlApjBmeFQo",
        "outputId": "1766a0ed-c729-414c-8eb1-9d9e26347de5"
      },
      "execution_count": 12,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
            "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
            "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
            "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
            "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
            "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
            "\t  v2.0.3\n",
            "\n",
            "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
            "||  \u001b[0m                                                                          ||\n",
            "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "||                           \u001b[36mSRR17158978_sorted.bam\u001b[0m \u001b[0m                          ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "||             Output file : \u001b[36mSRR17158978_counts.txt\u001b[0m \u001b[0m                          ||\n",
            "||                 Summary : \u001b[36mSRR17158978_counts.txt.summary\u001b[0m \u001b[0m                  ||\n",
            "||              Paired-end : \u001b[36myes\u001b[0m \u001b[0m                                             ||\n",
            "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
            "||              Annotation : \u001b[36mgencode.vM35.basic.annotation.gtf (GTF)\u001b[0m \u001b[0m         ||\n",
            "||      Dir for temp files : \u001b[36m/content/drive/MyDrive\u001b[0m \u001b[0m                          ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "||                 Threads : \u001b[36m4\u001b[0m \u001b[0m                                               ||\n",
            "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
            "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
            "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
            "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "\\\\============================================================================//\n",
            "\n",
            "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
            "||  \u001b[0m                                                                          ||\n",
            "|| Load annotation file gencode.vM35.basic.annotation.gtf \u001b[0m... \u001b[0m                ||\n",
            "||    Features : \u001b[36m532131\u001b[0m \u001b[0m                                                      ||\n",
            "||    Meta-features : \u001b[36m57132\u001b[0m \u001b[0m                                                  ||\n",
            "||    Chromosomes/contigs : \u001b[36m22\u001b[0m \u001b[0m                                               ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "|| Process BAM file SRR17158978_sorted.bam... \u001b[0m                                ||\n",
            "||    Paired-end reads are included. \u001b[0m                                         ||\n",
            "||    The reads are assigned on the single-end mode. \u001b[0m                         ||\n",
            "||    Total alignments : \u001b[36m58309766\u001b[0m \u001b[0m                                            ||\n",
            "||    Successfully assigned alignments : \u001b[36m27254467 (46.7%)\u001b[0m \u001b[0m                    ||\n",
            "||    Running time : \u001b[36m1.08 minutes\u001b[0m \u001b[0m                                            ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "|| Write the final count table. \u001b[0m                                              ||\n",
            "|| Write the read assignment summary. \u001b[0m                                        ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "|| Summary of counting results can be found in file \"/content/drive/MyDrive/ \u001b[0m ||\n",
            "|| SRR17158978_counts.txt.summary\" \u001b[0m                                           ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "\\\\============================================================================//\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "W4Tnx4tKDzve",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "d4b17ccf-d2de-4d79-8763-bbc493949b8a"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
            "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
            "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
            "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
            "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
            "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
            "\t  v2.0.3\n",
            "\n",
            "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
            "||  \u001b[0m                                                                          ||\n",
            "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "||                           \u001b[36mSRR17159238_sorted.bam\u001b[0m \u001b[0m                          ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "||             Output file : \u001b[36mSRR17159238_counts.txt\u001b[0m \u001b[0m                          ||\n",
            "||                 Summary : \u001b[36mSRR17159238_counts.txt.summary\u001b[0m \u001b[0m                  ||\n",
            "||              Paired-end : \u001b[36myes\u001b[0m \u001b[0m                                             ||\n",
            "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
            "||              Annotation : \u001b[36mgencode.vM35.basic.annotation.gtf (GTF)\u001b[0m \u001b[0m         ||\n",
            "||      Dir for temp files : \u001b[36m/content/drive/MyDrive\u001b[0m \u001b[0m                          ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "||                 Threads : \u001b[36m4\u001b[0m \u001b[0m                                               ||\n",
            "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
            "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
            "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
            "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "\\\\============================================================================//\n",
            "\n",
            "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
            "||  \u001b[0m                                                                          ||\n",
            "|| Load annotation file gencode.vM35.basic.annotation.gtf \u001b[0m... \u001b[0m                ||\n",
            "||    Features : \u001b[36m532131\u001b[0m \u001b[0m                                                      ||\n",
            "||    Meta-features : \u001b[36m57132\u001b[0m \u001b[0m                                                  ||\n",
            "||    Chromosomes/contigs : \u001b[36m22\u001b[0m \u001b[0m                                               ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "|| Process BAM file SRR17159238_sorted.bam... \u001b[0m                                ||\n",
            "||    Paired-end reads are included. \u001b[0m                                         ||\n",
            "||    The reads are assigned on the single-end mode. \u001b[0m                         ||\n",
            "||    Total alignments : \u001b[36m95387991\u001b[0m \u001b[0m                                            ||\n",
            "||    Successfully assigned alignments : \u001b[36m59398891 (62.3%)\u001b[0m \u001b[0m                    ||\n",
            "||    Running time : \u001b[36m1.80 minutes\u001b[0m \u001b[0m                                            ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "|| Write the final count table. \u001b[0m                                              ||\n",
            "|| Write the read assignment summary. \u001b[0m                                        ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "|| Summary of counting results can be found in file \"/content/drive/MyDrive/ \u001b[0m ||\n",
            "|| SRR17159238_counts.txt.summary\" \u001b[0m                                           ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "\\\\============================================================================//\n",
            "\n"
          ]
        }
      ],
      "source": [
        "!featureCounts -a /content/drive/MyDrive/gencode.vM35.basic.annotation.gtf -o /content/drive/MyDrive/SRR17159238_counts.txt -T 4 -p /content/drive/MyDrive/SRR17159238_sorted.bam\n"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "mount_file_id": "1p3ZqinHYgHEJr1ttOoyiCPpMqAlsMufA",
      "authorship_tag": "ABX9TyNqPFn8Rl3UV37tfd8OhRcr",
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}