{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "toc_visible": true,
      "mount_file_id": "1bn29FdsEBO5hGQB3vs18xI_xYR9G8w9S",
      "authorship_tag": "ABX9TyPb8B5kts4HZuV0BLKsCdEd",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/MoumitaRoy13/RNASequence_Analysis_Project/blob/main/Untitled3.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "T5nEf8F3_jNg",
        "outputId": "1bbe6b85-a102-4271-ef51-be9c6c24e012"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2024-10-27 05:57:28--  https://github.com/FelixKrueger/TrimGalore/archive/refs/tags/0.6.10.zip\n",
            "Resolving github.com (github.com)... 20.27.177.113\n",
            "Connecting to github.com (github.com)|20.27.177.113|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://codeload.github.com/FelixKrueger/TrimGalore/zip/refs/tags/0.6.10 [following]\n",
            "--2024-10-27 05:57:29--  https://codeload.github.com/FelixKrueger/TrimGalore/zip/refs/tags/0.6.10\n",
            "Resolving codeload.github.com (codeload.github.com)... 20.27.177.114\n",
            "Connecting to codeload.github.com (codeload.github.com)|20.27.177.114|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: unspecified [application/zip]\n",
            "Saving to: ‘0.6.10.zip’\n",
            "\n",
            "0.6.10.zip              [      <=>           ]  25.59M  16.9MB/s    in 1.5s    \n",
            "\n",
            "2024-10-27 05:57:31 (16.9 MB/s) - ‘0.6.10.zip’ saved [26835264]\n",
            "\n",
            "Archive:  0.6.10.zip\n",
            "4edff97d22f3837d42a29e4afbfaeb6e07ffb11b\n",
            "   creating: TrimGalore-0.6.10/\n",
            "  inflating: TrimGalore-0.6.10/.travis.yml  \n",
            "  inflating: TrimGalore-0.6.10/CHANGELOG.md  \n",
            "   creating: TrimGalore-0.6.10/Docs/\n",
            "   creating: TrimGalore-0.6.10/Docs/Images/\n",
            "  inflating: TrimGalore-0.6.10/Docs/Images/Thumbs.db  \n",
            "  inflating: TrimGalore-0.6.10/Docs/Images/adapters.png  \n",
            "  inflating: TrimGalore-0.6.10/Docs/Images/adapters_fixed.png  \n",
            "  inflating: TrimGalore-0.6.10/Docs/Images/fixed_quals.png  \n",
            "  inflating: TrimGalore-0.6.10/Docs/Images/fixed_quals_per_sequence.png  \n",
            "  inflating: TrimGalore-0.6.10/Docs/Images/logo.png  \n",
            "  inflating: TrimGalore-0.6.10/Docs/Images/pigz_bench.png  \n",
            "  inflating: TrimGalore-0.6.10/Docs/Images/poor_q_per_sequence.png  \n",
            "  inflating: TrimGalore-0.6.10/Docs/Images/poor_quals.png  \n",
            "  inflating: TrimGalore-0.6.10/Docs/RRBS_Guide.pdf  \n",
            "  inflating: TrimGalore-0.6.10/Docs/Trim_Galore_User_Guide.md  \n",
            "  inflating: TrimGalore-0.6.10/LICENSE  \n",
            "  inflating: TrimGalore-0.6.10/README.md  \n",
            "   creating: TrimGalore-0.6.10/test_files/\n",
            " extracting: TrimGalore-0.6.10/test_files/4_seqs_with_Ns.fastq.gz  \n",
            " extracting: TrimGalore-0.6.10/test_files/clock_10K_R1.fastq.gz  \n",
            " extracting: TrimGalore-0.6.10/test_files/clock_10K_R2.fastq.gz  \n",
            "  inflating: TrimGalore-0.6.10/test_files/colorspace_file.fastq  \n",
            " extracting: TrimGalore-0.6.10/test_files/empty_file.fastq  \n",
            " extracting: TrimGalore-0.6.10/test_files/illumina_100K.fastq.gz  \n",
            "  inflating: TrimGalore-0.6.10/test_files/illumina_100K.fastq.gz_trimming_report.txt  \n",
            " extracting: TrimGalore-0.6.10/test_files/illumina_10K.fastq.gz  \n",
            "  inflating: TrimGalore-0.6.10/test_files/illumina_10K.fastq.gz_trimming_report.txt  \n",
            " extracting: TrimGalore-0.6.10/test_files/illumina_10K_trimmed.fq.gz  \n",
            " extracting: TrimGalore-0.6.10/test_files/nextera_100K.fastq.gz  \n",
            "  inflating: TrimGalore-0.6.10/test_files/nextera_100K.fastq.gz_trimming_report.txt  \n",
            " extracting: TrimGalore-0.6.10/test_files/smallRNA_100K.fastq.gz  \n",
            "  inflating: TrimGalore-0.6.10/test_files/smallRNA_100K.fastq.gz_trimming_report.txt  \n",
            "  inflating: TrimGalore-0.6.10/test_files/smallRNA_100K_R1.fastq.gz_trimming_report.txt  \n",
            " extracting: TrimGalore-0.6.10/test_files/smallRNA_100K_R2.fastq.gz  \n",
            "  inflating: TrimGalore-0.6.10/test_files/smallRNA_100K_R2.fastq.gz_trimming_report.txt  \n",
            "  inflating: TrimGalore-0.6.10/test_files/truncated.fq.gz  \n",
            "  inflating: TrimGalore-0.6.10/trim_galore  \n"
          ]
        }
      ],
      "source": [
        "!wget https://github.com/FelixKrueger/TrimGalore/archive/refs/tags/0.6.10.zip\n",
        "!unzip 0.6.10.zip\n",
        "!chmod +x TrimGalore-0.6.10/trim_galore"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from google.colab import drive\n",
        "drive.mount('/content/drive')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "gvk6ztk9_uT0",
        "outputId": "c948fc19-0138-4395-bfd2-378b754288a6"
      },
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Mounted at /content/drive\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!trim_galore --paired --quality 20 --length 20 /content/drive/MyDrive/SRR17159276_1.fastq /content/drive/MyDrive/SRR17159276_2.fastq -o /content/drive/MyDrive/trimmed_output/\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Quo3s8KlAQpl",
        "outputId": "e230dba5-beaa-4d01-a43a-74346cdd351a"
      },
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Multicore support not enabled. Proceeding with single-core trimming.\n",
            "Path to Cutadapt set as: 'cutadapt' (default)\n",
            "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
            "Cutadapt version: 4.9\n",
            "single-core operation.\n",
            "igzip command line interface 2.30.0\n",
            "igzip detected. Using igzip for decompressing\n",
            "\n",
            "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
            "\n",
            "Output will be written into the directory: /content/drive/MyDrive/trimmed_output/\n",
            "\n",
            "\n",
            "AUTO-DETECTING ADAPTER TYPE\n",
            "===========================\n",
            "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> /content/drive/MyDrive/SRR17159276_1.fastq <<)\n",
            "\n",
            "Found perfect matches for the following adapter sequences:\n",
            "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
            "Illumina\t291075\tAGATCGGAAGAGC\t1000000\t29.11\n",
            "smallRNA\t8\tTGGAATTCTCGG\t1000000\t0.00\n",
            "Nextera\t0\tCTGTCTCTTATA\t1000000\t0.00\n",
            "Using Illumina adapter for trimming (count: 291075). Second best hit was smallRNA (count: 8)\n",
            "\n",
            "Writing report to '/content/drive/MyDrive/trimmed_output/SRR17159276_1.fastq_trimming_report.txt'\n",
            "\n",
            "SUMMARISING RUN PARAMETERS\n",
            "==========================\n",
            "Input filename: /content/drive/MyDrive/SRR17159276_1.fastq\n",
            "Trimming mode: paired-end\n",
            "Trim Galore version: 0.6.10\n",
            "Cutadapt version: 4.9\n",
            "Number of cores used for trimming: 1\n",
            "Quality Phred score cutoff: 20\n",
            "Quality encoding type selected: ASCII+33\n",
            "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
            "Maximum trimming error rate: 0.1 (default)\n",
            "Minimum required adapter overlap (stringency): 1 bp\n",
            "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
            "\n",
            "Cutadapt seems to be fairly up-to-date (version 4.9). Setting -j 1\n",
            "Writing final adapter and quality trimmed output to SRR17159276_1_trimmed.fq\n",
            "\n",
            "\n",
            "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /content/drive/MyDrive/SRR17159276_1.fastq <<< \n",
            "10000000 sequences processed\n",
            "This is cutadapt 4.9 with Python 3.10.12\n",
            "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /content/drive/MyDrive/SRR17159276_1.fastq\n",
            "Processing single-end reads on 1 core ...\n",
            "Finished in 293.156 s (18.402 µs/read; 3.26 M reads/minute).\n",
            "\n",
            "=== Summary ===\n",
            "\n",
            "Total reads processed:              15,930,786\n",
            "Reads with adapters:                 9,525,086 (59.8%)\n",
            "Reads written (passing filters):    15,930,786 (100.0%)\n",
            "\n",
            "Total basepairs processed: 2,389,617,900 bp\n",
            "Quality-trimmed:               1,199,472 bp (0.1%)\n",
            "Total written (filtered):  2,202,262,675 bp (92.2%)\n",
            "\n",
            "=== Adapter 1 ===\n",
            "\n",
            "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 9525086 times\n",
            "\n",
            "Minimum overlap: 1\n",
            "No. of allowed errors:\n",
            "1-9 bp: 0; 10-13 bp: 1\n",
            "\n",
            "Bases preceding removed adapters:\n",
            "  A: 23.0%\n",
            "  C: 26.2%\n",
            "  G: 31.7%\n",
            "  T: 18.8%\n",
            "  none/other: 0.2%\n",
            "\n",
            "Overview of removed sequences\n",
            "length\tcount\texpect\tmax.err\terror counts\n",
            "1\t2377912\t3982696.5\t0\t2377912\n",
            "2\t824951\t995674.1\t0\t824951\n",
            "3\t313277\t248918.5\t0\t313277\n",
            "4\t167883\t62229.6\t0\t167883\n",
            "5\t133966\t15557.4\t0\t133966\n",
            "6\t121053\t3889.4\t0\t121053\n",
            "7\t122892\t972.3\t0\t122892\n",
            "8\t126770\t243.1\t0\t126770\n",
            "9\t125642\t60.8\t0\t125467 175\n",
            "10\t130549\t15.2\t1\t128029 2520\n",
            "11\t130449\t3.8\t1\t127605 2844\n",
            "12\t130206\t0.9\t1\t127377 2829\n",
            "13\t128592\t0.2\t1\t125809 2783\n",
            "14\t129851\t0.2\t1\t126925 2926\n",
            "15\t126450\t0.2\t1\t123739 2711\n",
            "16\t129258\t0.2\t1\t126282 2976\n",
            "17\t129426\t0.2\t1\t126449 2977\n",
            "18\t128738\t0.2\t1\t126113 2625\n",
            "19\t129139\t0.2\t1\t126513 2626\n",
            "20\t129225\t0.2\t1\t126473 2752\n",
            "21\t126526\t0.2\t1\t123891 2635\n",
            "22\t125932\t0.2\t1\t123197 2735\n",
            "23\t126547\t0.2\t1\t123847 2700\n",
            "24\t123753\t0.2\t1\t120777 2976\n",
            "25\t118978\t0.2\t1\t116506 2472\n",
            "26\t118664\t0.2\t1\t116344 2320\n",
            "27\t118002\t0.2\t1\t115634 2368\n",
            "28\t115841\t0.2\t1\t113556 2285\n",
            "29\t116274\t0.2\t1\t113689 2585\n",
            "30\t112975\t0.2\t1\t110832 2143\n",
            "31\t112123\t0.2\t1\t109976 2147\n",
            "32\t111958\t0.2\t1\t109830 2128\n",
            "33\t109723\t0.2\t1\t107408 2315\n",
            "34\t107417\t0.2\t1\t105049 2368\n",
            "35\t99928\t0.2\t1\t98067 1861\n",
            "36\t94422\t0.2\t1\t92608 1814\n",
            "37\t94756\t0.2\t1\t92923 1833\n",
            "38\t92208\t0.2\t1\t89923 2285\n",
            "39\t89934\t0.2\t1\t87771 2163\n",
            "40\t83344\t0.2\t1\t81259 2085\n",
            "41\t83444\t0.2\t1\t81582 1862\n",
            "42\t84161\t0.2\t1\t82066 2095\n",
            "43\t83170\t0.2\t1\t80967 2203\n",
            "44\t77069\t0.2\t1\t75203 1866\n",
            "45\t117107\t0.2\t1\t114417 2690\n",
            "46\t26958\t0.2\t1\t26185 773\n",
            "47\t61068\t0.2\t1\t59678 1390\n",
            "48\t67210\t0.2\t1\t65528 1682\n",
            "49\t61447\t0.2\t1\t60023 1424\n",
            "50\t50020\t0.2\t1\t48861 1159\n",
            "51\t64750\t0.2\t1\t63516 1234\n",
            "52\t48949\t0.2\t1\t48024 925\n",
            "53\t45228\t0.2\t1\t44429 799\n",
            "54\t48437\t0.2\t1\t47567 870\n",
            "55\t50629\t0.2\t1\t49618 1011\n",
            "56\t41960\t0.2\t1\t41157 803\n",
            "57\t45033\t0.2\t1\t44238 795\n",
            "58\t34214\t0.2\t1\t33569 645\n",
            "59\t39725\t0.2\t1\t38932 793\n",
            "60\t27748\t0.2\t1\t27279 469\n",
            "61\t23226\t0.2\t1\t22809 417\n",
            "62\t34026\t0.2\t1\t33447 579\n",
            "63\t32577\t0.2\t1\t31999 578\n",
            "64\t20813\t0.2\t1\t20412 401\n",
            "65\t22908\t0.2\t1\t22474 434\n",
            "66\t39929\t0.2\t1\t39265 664\n",
            "67\t17497\t0.2\t1\t17143 354\n",
            "68\t27349\t0.2\t1\t26900 449\n",
            "69\t19893\t0.2\t1\t19509 384\n",
            "70\t21879\t0.2\t1\t21486 393\n",
            "71\t548\t0.2\t1\t491 57\n",
            "72\t871\t0.2\t1\t828 43\n",
            "73\t5828\t0.2\t1\t5715 113\n",
            "74\t9847\t0.2\t1\t9654 193\n",
            "75\t11090\t0.2\t1\t10899 191\n",
            "76\t10924\t0.2\t1\t10699 225\n",
            "77\t10059\t0.2\t1\t9871 188\n",
            "78\t9087\t0.2\t1\t8901 186\n",
            "79\t8138\t0.2\t1\t7983 155\n",
            "80\t7159\t0.2\t1\t7005 154\n",
            "81\t6703\t0.2\t1\t6584 119\n",
            "82\t6259\t0.2\t1\t6137 122\n",
            "83\t5706\t0.2\t1\t5583 123\n",
            "84\t5176\t0.2\t1\t5093 83\n",
            "85\t4917\t0.2\t1\t4820 97\n",
            "86\t4818\t0.2\t1\t4739 79\n",
            "87\t4610\t0.2\t1\t4519 91\n",
            "88\t4131\t0.2\t1\t4051 80\n",
            "89\t3856\t0.2\t1\t3775 81\n",
            "90\t3502\t0.2\t1\t3429 73\n",
            "91\t3032\t0.2\t1\t2950 82\n",
            "92\t2514\t0.2\t1\t2450 64\n",
            "93\t2267\t0.2\t1\t2221 46\n",
            "94\t2032\t0.2\t1\t1984 48\n",
            "95\t1668\t0.2\t1\t1632 36\n",
            "96\t1715\t0.2\t1\t1683 32\n",
            "97\t1655\t0.2\t1\t1628 27\n",
            "98\t1624\t0.2\t1\t1599 25\n",
            "99\t1453\t0.2\t1\t1422 31\n",
            "100\t1339\t0.2\t1\t1306 33\n",
            "101\t1116\t0.2\t1\t1084 32\n",
            "102\t966\t0.2\t1\t938 28\n",
            "103\t829\t0.2\t1\t819 10\n",
            "104\t695\t0.2\t1\t676 19\n",
            "105\t519\t0.2\t1\t498 21\n",
            "106\t522\t0.2\t1\t498 24\n",
            "107\t451\t0.2\t1\t442 9\n",
            "108\t452\t0.2\t1\t431 21\n",
            "109\t448\t0.2\t1\t425 23\n",
            "110\t370\t0.2\t1\t352 18\n",
            "111\t323\t0.2\t1\t304 19\n",
            "112\t234\t0.2\t1\t221 13\n",
            "113\t166\t0.2\t1\t158 8\n",
            "114\t148\t0.2\t1\t135 13\n",
            "115\t155\t0.2\t1\t147 8\n",
            "116\t138\t0.2\t1\t131 7\n",
            "117\t102\t0.2\t1\t83 19\n",
            "118\t105\t0.2\t1\t94 11\n",
            "119\t93\t0.2\t1\t71 22\n",
            "120\t68\t0.2\t1\t60 8\n",
            "121\t48\t0.2\t1\t41 7\n",
            "122\t47\t0.2\t1\t30 17\n",
            "123\t32\t0.2\t1\t23 9\n",
            "124\t29\t0.2\t1\t19 10\n",
            "125\t23\t0.2\t1\t13 10\n",
            "126\t32\t0.2\t1\t18 14\n",
            "127\t29\t0.2\t1\t19 10\n",
            "128\t24\t0.2\t1\t10 14\n",
            "129\t44\t0.2\t1\t17 27\n",
            "130\t24\t0.2\t1\t13 11\n",
            "131\t23\t0.2\t1\t9 14\n",
            "132\t32\t0.2\t1\t19 13\n",
            "133\t25\t0.2\t1\t18 7\n",
            "134\t62\t0.2\t1\t49 13\n",
            "135\t29\t0.2\t1\t25 4\n",
            "136\t34\t0.2\t1\t19 15\n",
            "137\t39\t0.2\t1\t28 11\n",
            "138\t42\t0.2\t1\t28 14\n",
            "139\t29\t0.2\t1\t24 5\n",
            "140\t16\t0.2\t1\t8 8\n",
            "141\t33\t0.2\t1\t25 8\n",
            "142\t34\t0.2\t1\t26 8\n",
            "143\t30\t0.2\t1\t27 3\n",
            "144\t21\t0.2\t1\t13 8\n",
            "145\t33\t0.2\t1\t21 12\n",
            "146\t15\t0.2\t1\t10 5\n",
            "147\t63\t0.2\t1\t48 15\n",
            "148\t22\t0.2\t1\t14 8\n",
            "149\t33\t0.2\t1\t8 25\n",
            "150\t19753\t0.2\t1\t10 19743\n",
            "\n",
            "RUN STATISTICS FOR INPUT FILE: /content/drive/MyDrive/SRR17159276_1.fastq\n",
            "=============================================\n",
            "15930786 sequences processed in total\n",
            "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
            "\n",
            "Writing report to '/content/drive/MyDrive/trimmed_output/SRR17159276_2.fastq_trimming_report.txt'\n",
            "\n",
            "SUMMARISING RUN PARAMETERS\n",
            "==========================\n",
            "Input filename: /content/drive/MyDrive/SRR17159276_2.fastq\n",
            "Trimming mode: paired-end\n",
            "Trim Galore version: 0.6.10\n",
            "Cutadapt version: 4.9\n",
            "Number of cores used for trimming: 1\n",
            "Quality Phred score cutoff: 20\n",
            "Quality encoding type selected: ASCII+33\n",
            "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
            "Maximum trimming error rate: 0.1 (default)\n",
            "Minimum required adapter overlap (stringency): 1 bp\n",
            "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
            "\n",
            "Cutadapt seems to be fairly up-to-date (version 4.9). Setting -j -j 1\n",
            "Writing final adapter and quality trimmed output to SRR17159276_2_trimmed.fq\n",
            "\n",
            "\n",
            "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /content/drive/MyDrive/SRR17159276_2.fastq <<< \n",
            "10000000 sequences processed\n",
            "This is cutadapt 4.9 with Python 3.10.12\n",
            "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /content/drive/MyDrive/SRR17159276_2.fastq\n",
            "Processing single-end reads on 1 core ...\n",
            "Finished in 332.170 s (20.851 µs/read; 2.88 M reads/minute).\n",
            "\n",
            "=== Summary ===\n",
            "\n",
            "Total reads processed:              15,930,786\n",
            "Reads with adapters:                 9,733,132 (61.1%)\n",
            "Reads written (passing filters):    15,930,786 (100.0%)\n",
            "\n",
            "Total basepairs processed: 2,389,617,900 bp\n",
            "Quality-trimmed:               8,542,132 bp (0.4%)\n",
            "Total written (filtered):  2,200,403,963 bp (92.1%)\n",
            "\n",
            "=== Adapter 1 ===\n",
            "\n",
            "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 9733132 times\n",
            "\n",
            "Minimum overlap: 1\n",
            "No. of allowed errors:\n",
            "1-9 bp: 0; 10-13 bp: 1\n",
            "\n",
            "Bases preceding removed adapters:\n",
            "  A: 21.5%\n",
            "  C: 34.8%\n",
            "  G: 31.4%\n",
            "  T: 12.4%\n",
            "  none/other: 0.0%\n",
            "\n",
            "Overview of removed sequences\n",
            "length\tcount\texpect\tmax.err\terror counts\n",
            "1\t2686088\t3982696.5\t0\t2686088\n",
            "2\t760405\t995674.1\t0\t760405\n",
            "3\t342074\t248918.5\t0\t342074\n",
            "4\t172344\t62229.6\t0\t172344\n",
            "5\t134702\t15557.4\t0\t134702\n",
            "6\t120870\t3889.4\t0\t120870\n",
            "7\t123589\t972.3\t0\t123589\n",
            "8\t130188\t243.1\t0\t130188\n",
            "9\t120636\t60.8\t0\t120315 321\n",
            "10\t130194\t15.2\t1\t128058 2136\n",
            "11\t128986\t3.8\t1\t126359 2627\n",
            "12\t130079\t0.9\t1\t127916 2163\n",
            "13\t126732\t0.2\t1\t124569 2163\n",
            "14\t129008\t0.2\t1\t126804 2204\n",
            "15\t123956\t0.2\t1\t122109 1847\n",
            "16\t128249\t0.2\t1\t126067 2182\n",
            "17\t126444\t0.2\t1\t124370 2074\n",
            "18\t126554\t0.2\t1\t124691 1863\n",
            "19\t134722\t0.2\t1\t132439 2283\n",
            "20\t128147\t0.2\t1\t126304 1843\n",
            "21\t123210\t0.2\t1\t121288 1922\n",
            "22\t122687\t0.2\t1\t120971 1716\n",
            "23\t124215\t0.2\t1\t122389 1826\n",
            "24\t129195\t0.2\t1\t126752 2443\n",
            "25\t129605\t0.2\t1\t128008 1597\n",
            "26\t103745\t0.2\t1\t102355 1390\n",
            "27\t113778\t0.2\t1\t112095 1683\n",
            "28\t115579\t0.2\t1\t113950 1629\n",
            "29\t113243\t0.2\t1\t111505 1738\n",
            "30\t112104\t0.2\t1\t110446 1658\n",
            "31\t111360\t0.2\t1\t109912 1448\n",
            "32\t111332\t0.2\t1\t109764 1568\n",
            "33\t106947\t0.2\t1\t105542 1405\n",
            "34\t108422\t0.2\t1\t106715 1707\n",
            "35\t99015\t0.2\t1\t97787 1228\n",
            "36\t101761\t0.2\t1\t100223 1538\n",
            "37\t103725\t0.2\t1\t102477 1248\n",
            "38\t77028\t0.2\t1\t75871 1157\n",
            "39\t83944\t0.2\t1\t82942 1002\n",
            "40\t83250\t0.2\t1\t82324 926\n",
            "41\t82518\t0.2\t1\t81505 1013\n",
            "42\t96439\t0.2\t1\t95168 1271\n",
            "43\t67481\t0.2\t1\t66776 705\n",
            "44\t94227\t0.2\t1\t92936 1291\n",
            "45\t77392\t0.2\t1\t76553 839\n",
            "46\t71479\t0.2\t1\t70666 813\n",
            "47\t105105\t0.2\t1\t104034 1071\n",
            "48\t27786\t0.2\t1\t27378 408\n",
            "49\t70398\t0.2\t1\t69751 647\n",
            "50\t35590\t0.2\t1\t35177 413\n",
            "51\t46954\t0.2\t1\t46495 459\n",
            "52\t35824\t0.2\t1\t35449 375\n",
            "53\t49445\t0.2\t1\t48913 532\n",
            "54\t70968\t0.2\t1\t70249 719\n",
            "55\t22629\t0.2\t1\t22334 295\n",
            "56\t43221\t0.2\t1\t42691 530\n",
            "57\t122291\t0.2\t1\t121296 995\n",
            "58\t21705\t0.2\t1\t21386 319\n",
            "59\t10981\t0.2\t1\t10776 205\n",
            "60\t68373\t0.2\t1\t67823 550\n",
            "61\t11209\t0.2\t1\t11028 181\n",
            "62\t13504\t0.2\t1\t13309 195\n",
            "63\t70695\t0.2\t1\t70153 542\n",
            "64\t31170\t0.2\t1\t30771 399\n",
            "65\t5827\t0.2\t1\t5723 104\n",
            "66\t14792\t0.2\t1\t14619 173\n",
            "67\t12092\t0.2\t1\t11979 113\n",
            "68\t9114\t0.2\t1\t9020 94\n",
            "69\t9942\t0.2\t1\t9835 107\n",
            "70\t17027\t0.2\t1\t16884 143\n",
            "71\t6984\t0.2\t1\t6912 72\n",
            "72\t6633\t0.2\t1\t6575 58\n",
            "73\t8726\t0.2\t1\t8637 89\n",
            "74\t6521\t0.2\t1\t6449 72\n",
            "75\t6284\t0.2\t1\t6214 70\n",
            "76\t4231\t0.2\t1\t4165 66\n",
            "77\t5997\t0.2\t1\t5935 62\n",
            "78\t7305\t0.2\t1\t7236 69\n",
            "79\t6966\t0.2\t1\t6900 66\n",
            "80\t6224\t0.2\t1\t6155 69\n",
            "81\t5998\t0.2\t1\t5937 61\n",
            "82\t5686\t0.2\t1\t5627 59\n",
            "83\t5315\t0.2\t1\t5244 71\n",
            "84\t4878\t0.2\t1\t4820 58\n",
            "85\t4746\t0.2\t1\t4674 72\n",
            "86\t4414\t0.2\t1\t4373 41\n",
            "87\t4242\t0.2\t1\t4198 44\n",
            "88\t3758\t0.2\t1\t3717 41\n",
            "89\t3598\t0.2\t1\t3571 27\n",
            "90\t3406\t0.2\t1\t3351 55\n",
            "91\t2833\t0.2\t1\t2802 31\n",
            "92\t2380\t0.2\t1\t2351 29\n",
            "93\t2204\t0.2\t1\t2177 27\n",
            "94\t1881\t0.2\t1\t1854 27\n",
            "95\t1599\t0.2\t1\t1573 26\n",
            "96\t1629\t0.2\t1\t1606 23\n",
            "97\t1594\t0.2\t1\t1573 21\n",
            "98\t1602\t0.2\t1\t1567 35\n",
            "99\t1446\t0.2\t1\t1421 25\n",
            "100\t1358\t0.2\t1\t1335 23\n",
            "101\t1065\t0.2\t1\t1056 9\n",
            "102\t918\t0.2\t1\t908 10\n",
            "103\t798\t0.2\t1\t788 10\n",
            "104\t649\t0.2\t1\t641 8\n",
            "105\t508\t0.2\t1\t493 15\n",
            "106\t488\t0.2\t1\t476 12\n",
            "107\t431\t0.2\t1\t422 9\n",
            "108\t438\t0.2\t1\t431 7\n",
            "109\t421\t0.2\t1\t405 16\n",
            "110\t364\t0.2\t1\t356 8\n",
            "111\t313\t0.2\t1\t303 10\n",
            "112\t237\t0.2\t1\t227 10\n",
            "113\t166\t0.2\t1\t161 5\n",
            "114\t148\t0.2\t1\t136 12\n",
            "115\t143\t0.2\t1\t134 9\n",
            "116\t128\t0.2\t1\t125 3\n",
            "117\t83\t0.2\t1\t81 2\n",
            "118\t99\t0.2\t1\t96 3\n",
            "119\t79\t0.2\t1\t69 10\n",
            "120\t76\t0.2\t1\t62 14\n",
            "121\t52\t0.2\t1\t44 8\n",
            "122\t36\t0.2\t1\t30 6\n",
            "123\t30\t0.2\t1\t23 7\n",
            "124\t35\t0.2\t1\t20 15\n",
            "125\t24\t0.2\t1\t14 10\n",
            "126\t21\t0.2\t1\t17 4\n",
            "127\t26\t0.2\t1\t17 9\n",
            "128\t17\t0.2\t1\t13 4\n",
            "129\t31\t0.2\t1\t17 14\n",
            "130\t20\t0.2\t1\t12 8\n",
            "131\t18\t0.2\t1\t10 8\n",
            "132\t22\t0.2\t1\t18 4\n",
            "133\t20\t0.2\t1\t18 2\n",
            "134\t52\t0.2\t1\t48 4\n",
            "135\t31\t0.2\t1\t26 5\n",
            "136\t28\t0.2\t1\t18 10\n",
            "137\t40\t0.2\t1\t28 12\n",
            "138\t30\t0.2\t1\t25 5\n",
            "139\t24\t0.2\t1\t22 2\n",
            "140\t9\t0.2\t1\t8 1\n",
            "141\t35\t0.2\t1\t21 14\n",
            "142\t33\t0.2\t1\t14 19\n",
            "143\t35\t0.2\t1\t17 18\n",
            "144\t16\t0.2\t1\t7 9\n",
            "145\t22\t0.2\t1\t16 6\n",
            "146\t15\t0.2\t1\t4 11\n",
            "147\t71\t0.2\t1\t36 35\n",
            "148\t16\t0.2\t1\t10 6\n",
            "149\t21\t0.2\t1\t3 18\n",
            "150\t22\t0.2\t1\t6 16\n",
            "\n",
            "RUN STATISTICS FOR INPUT FILE: /content/drive/MyDrive/SRR17159276_2.fastq\n",
            "=============================================\n",
            "15930786 sequences processed in total\n",
            "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
            "\n",
            "Validate paired-end files SRR17159276_1_trimmed.fq and SRR17159276_2_trimmed.fq\n",
            "file_1: SRR17159276_1_trimmed.fq, file_2: SRR17159276_2_trimmed.fq\n",
            "\n",
            "\n",
            ">>>>> Now validing the length of the 2 paired-end infiles: SRR17159276_1_trimmed.fq and SRR17159276_2_trimmed.fq <<<<<\n",
            "Writing validated paired-end Read 1 reads to SRR17159276_1_val_1.fq\n",
            "Writing validated paired-end Read 2 reads to SRR17159276_2_val_2.fq\n",
            "\n",
            "Total number of sequences analysed: 15930786\n",
            "\n",
            "Number of sequence pairs removed because at least one read was shorter than the length cutoff (20 bp): 47025 (0.30%)\n",
            "\n",
            "Deleting both intermediate output files SRR17159276_1_trimmed.fq and SRR17159276_2_trimmed.fq\n",
            "\n",
            "====================================================================================================\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "mNj5p61eEIP-"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install cutadapt"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "iF0d8DF3Anmm",
        "outputId": "f6cc2059-342d-4954-d79e-1e68bec1f314"
      },
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting cutadapt\n",
            "  Downloading cutadapt-4.9-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (3.4 kB)\n",
            "Collecting dnaio>=1.2.0 (from cutadapt)\n",
            "  Downloading dnaio-1.2.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (3.5 kB)\n",
            "Collecting xopen>=1.6.0 (from cutadapt)\n",
            "  Downloading xopen-2.0.2-py3-none-any.whl.metadata (15 kB)\n",
            "Collecting isal>=1.6.1 (from xopen>=1.6.0->cutadapt)\n",
            "  Downloading isal-1.7.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (10 kB)\n",
            "Collecting zlib-ng>=0.4.1 (from xopen>=1.6.0->cutadapt)\n",
            "  Downloading zlib_ng-0.5.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (6.8 kB)\n",
            "Downloading cutadapt-4.9-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (287 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m287.7/287.7 kB\u001b[0m \u001b[31m6.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading dnaio-1.2.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (111 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m111.7/111.7 kB\u001b[0m \u001b[31m9.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading xopen-2.0.2-py3-none-any.whl (17 kB)\n",
            "Downloading isal-1.7.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (258 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m258.9/258.9 kB\u001b[0m \u001b[31m14.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading zlib_ng-0.5.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (108 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m108.1/108.1 kB\u001b[0m \u001b[31m6.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: zlib-ng, isal, xopen, dnaio, cutadapt\n",
            "Successfully installed cutadapt-4.9 dnaio-1.2.2 isal-1.7.1 xopen-2.0.2 zlib-ng-0.5.1\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!apt-get install -y isal"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "nDuOskI6ArGN",
        "outputId": "f708af00-2995-4a8c-c005-7532b85af525"
      },
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "The following additional packages will be installed:\n",
            "  libisal2\n",
            "The following NEW packages will be installed:\n",
            "  isal libisal2\n",
            "0 upgraded, 2 newly installed, 0 to remove and 49 not upgraded.\n",
            "Need to get 121 kB of archives.\n",
            "After this operation, 403 kB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libisal2 amd64 2.30.0-4 [110 kB]\n",
            "Get:2 http://archive.ubuntu.com/ubuntu jammy/universe amd64 isal amd64 2.30.0-4 [10.7 kB]\n",
            "Fetched 121 kB in 1s (135 kB/s)\n",
            "Selecting previously unselected package libisal2.\n",
            "(Reading database ... 123622 files and directories currently installed.)\n",
            "Preparing to unpack .../libisal2_2.30.0-4_amd64.deb ...\n",
            "Unpacking libisal2 (2.30.0-4) ...\n",
            "Selecting previously unselected package isal.\n",
            "Preparing to unpack .../isal_2.30.0-4_amd64.deb ...\n",
            "Unpacking isal (2.30.0-4) ...\n",
            "Setting up libisal2 (2.30.0-4) ...\n",
            "Setting up isal (2.30.0-4) ...\n",
            "Processing triggers for man-db (2.10.2-1) ...\n",
            "Processing triggers for libc-bin (2.35-0ubuntu3.4) ...\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_level_zero.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_5.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_opencl.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbb.so.12 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_0.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_loader.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc_proxy.so.2 is not a symbolic link\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!mv TrimGalore-0.6.10/trim_galore /usr/local/bin/"
      ],
      "metadata": {
        "id": "anhhUxVxAhjP"
      },
      "execution_count": 3,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!hisat2 -x /content/drive/MyDrive/reference_index/GRCm39_index \\\n",
        "       -1 /content/drive/MyDrive/trimmed_output/SRR17159276_1_val_1.fq \\\n",
        "       -2 /content/drive/MyDrive/trimmed_output/SRR17159276_2_val_2.fq \\\n",
        "       -S /content/drive/MyDrive/SRR17159276_aligned.sam\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "bARFIwPiEKGi",
        "outputId": "99d424bd-b978-4440-f1f8-8c13579e806a"
      },
      "execution_count": 11,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "15883761 reads; of these:\n",
            "  15883761 (100.00%) were paired; of these:\n",
            "    384587 (2.42%) aligned concordantly 0 times\n",
            "    12371394 (77.89%) aligned concordantly exactly 1 time\n",
            "    3127780 (19.69%) aligned concordantly >1 times\n",
            "    ----\n",
            "    384587 pairs aligned concordantly 0 times; of these:\n",
            "      17653 (4.59%) aligned discordantly 1 time\n",
            "    ----\n",
            "    366934 pairs aligned 0 times concordantly or discordantly; of these:\n",
            "      733868 mates make up the pairs; of these:\n",
            "        416386 (56.74%) aligned 0 times\n",
            "        247970 (33.79%) aligned exactly 1 time\n",
            "        69512 (9.47%) aligned >1 times\n",
            "98.69% overall alignment rate\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!samtools view -bS /content/drive/MyDrive/SRR17159276_aligned.sam | samtools sort -o /content/drive/MyDrive/SRR17159276_sorted.bam\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "9ShxXLYglgos",
        "outputId": "77b44277-d663-4deb-dedf-807e35cfa1b7"
      },
      "execution_count": 12,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[bam_sort_core] merging from 17 files and 1 in-memory blocks...\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!samtools index /content/drive/MyDrive/SRR17159276_sorted.bam"
      ],
      "metadata": {
        "id": "GMDSI78ApG55"
      },
      "execution_count": 13,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!samtools flagstat /content/drive/MyDrive/SRR17159276_sorted.bam\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Q5xd1nGOpt0A",
        "outputId": "1da6cdd8-b9be-4ad1-ddc0-96d7349d0c82"
      },
      "execution_count": 14,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "39023740 + 0 in total (QC-passed reads + QC-failed reads)\n",
            "31767522 + 0 primary\n",
            "7256218 + 0 secondary\n",
            "0 + 0 supplementary\n",
            "0 + 0 duplicates\n",
            "0 + 0 primary duplicates\n",
            "38607354 + 0 mapped (98.93% : N/A)\n",
            "31351136 + 0 primary mapped (98.69% : N/A)\n",
            "31767522 + 0 paired in sequencing\n",
            "15883761 + 0 read1\n",
            "15883761 + 0 read2\n",
            "30998348 + 0 properly paired (97.58% : N/A)\n",
            "31098996 + 0 with itself and mate mapped\n",
            "252140 + 0 singletons (0.79% : N/A)\n",
            "71492 + 0 with mate mapped to a different chr\n",
            "46259 + 0 with mate mapped to a different chr (mapQ>=5)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!apt-get install -y hisat2"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "RXEfTgFgEKJ2",
        "outputId": "61188f79-35fa-4484-b7e9-23c126acae9b"
      },
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "The following additional packages will be installed:\n",
            "  bcftools libhts3 libhtscodecs2 python3-hisat2 samtools\n",
            "Suggested packages:\n",
            "  python3-numpy python3-matplotlib texlive-latex-recommended cwltool\n",
            "The following NEW packages will be installed:\n",
            "  bcftools hisat2 libhts3 libhtscodecs2 python3-hisat2 samtools\n",
            "0 upgraded, 6 newly installed, 0 to remove and 49 not upgraded.\n",
            "Need to get 5,505 kB of archives.\n",
            "After this operation, 17.1 MB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libhtscodecs2 amd64 1.1.1-3 [53.2 kB]\n",
            "Get:2 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libhts3 amd64 1.13+ds-2build1 [390 kB]\n",
            "Get:3 http://archive.ubuntu.com/ubuntu jammy/universe amd64 bcftools amd64 1.13-1 [697 kB]\n",
            "Get:4 http://archive.ubuntu.com/ubuntu jammy/universe amd64 hisat2 amd64 2.2.1-3 [3,832 kB]\n",
            "Get:5 http://archive.ubuntu.com/ubuntu jammy/universe amd64 python3-hisat2 all 2.2.1-3 [12.7 kB]\n",
            "Get:6 http://archive.ubuntu.com/ubuntu jammy/universe amd64 samtools amd64 1.13-4 [520 kB]\n",
            "Fetched 5,505 kB in 2s (2,311 kB/s)\n",
            "Selecting previously unselected package libhtscodecs2:amd64.\n",
            "(Reading database ... 123632 files and directories currently installed.)\n",
            "Preparing to unpack .../0-libhtscodecs2_1.1.1-3_amd64.deb ...\n",
            "Unpacking libhtscodecs2:amd64 (1.1.1-3) ...\n",
            "Selecting previously unselected package libhts3:amd64.\n",
            "Preparing to unpack .../1-libhts3_1.13+ds-2build1_amd64.deb ...\n",
            "Unpacking libhts3:amd64 (1.13+ds-2build1) ...\n",
            "Selecting previously unselected package bcftools.\n",
            "Preparing to unpack .../2-bcftools_1.13-1_amd64.deb ...\n",
            "Unpacking bcftools (1.13-1) ...\n",
            "Selecting previously unselected package hisat2.\n",
            "Preparing to unpack .../3-hisat2_2.2.1-3_amd64.deb ...\n",
            "Unpacking hisat2 (2.2.1-3) ...\n",
            "Selecting previously unselected package python3-hisat2.\n",
            "Preparing to unpack .../4-python3-hisat2_2.2.1-3_all.deb ...\n",
            "Unpacking python3-hisat2 (2.2.1-3) ...\n",
            "Selecting previously unselected package samtools.\n",
            "Preparing to unpack .../5-samtools_1.13-4_amd64.deb ...\n",
            "Unpacking samtools (1.13-4) ...\n",
            "Setting up libhtscodecs2:amd64 (1.1.1-3) ...\n",
            "Setting up libhts3:amd64 (1.13+ds-2build1) ...\n",
            "Setting up bcftools (1.13-1) ...\n",
            "Setting up samtools (1.13-4) ...\n",
            "Setting up hisat2 (2.2.1-3) ...\n",
            "Setting up python3-hisat2 (2.2.1-3) ...\n",
            "Processing triggers for man-db (2.10.2-1) ...\n",
            "Processing triggers for libc-bin (2.35-0ubuntu3.4) ...\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_level_zero.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_5.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_opencl.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbb.so.12 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_0.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_loader.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc_proxy.so.2 is not a symbolic link\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!featureCounts -a /content/drive/MyDrive/gencode.vM35.basic.annotation.gtf -o /content/drive/MyDrive/SRR17159276_counts.txt -T 4 /content/drive/MyDrive/SRR17159276_sorted.bam\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "4EBpTSidrIcj",
        "outputId": "2aef5f33-c36c-49e5-e5c7-eaf9ff03242a"
      },
      "execution_count": 15,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "/bin/bash: line 1: featureCounts: command not found\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!featureCounts -a /content/drive/MyDrive/gencode.vM35.basic.annotation.gtf -o /content/drive/MyDrive/SRR17159276_counts.txt -T 4 -p /content/drive/MyDrive/SRR17159276_sorted.bam\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "pTEXyqcIrhCj",
        "outputId": "14988410-18d6-40c2-b01a-63eb8b206257"
      },
      "execution_count": 17,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
            "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
            "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
            "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
            "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
            "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
            "\t  v2.0.3\n",
            "\n",
            "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
            "||  \u001b[0m                                                                          ||\n",
            "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "||                           \u001b[36mSRR17159276_sorted.bam\u001b[0m \u001b[0m                          ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "||             Output file : \u001b[36mSRR17159276_counts.txt\u001b[0m \u001b[0m                          ||\n",
            "||                 Summary : \u001b[36mSRR17159276_counts.txt.summary\u001b[0m \u001b[0m                  ||\n",
            "||              Paired-end : \u001b[36myes\u001b[0m \u001b[0m                                             ||\n",
            "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
            "||              Annotation : \u001b[36mgencode.vM35.basic.annotation.gtf (GTF)\u001b[0m \u001b[0m         ||\n",
            "||      Dir for temp files : \u001b[36m/content/drive/MyDrive\u001b[0m \u001b[0m                          ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "||                 Threads : \u001b[36m4\u001b[0m \u001b[0m                                               ||\n",
            "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
            "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
            "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
            "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "\\\\============================================================================//\n",
            "\n",
            "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
            "||  \u001b[0m                                                                          ||\n",
            "|| Load annotation file gencode.vM35.basic.annotation.gtf \u001b[0m... \u001b[0m                ||\n",
            "||    Features : \u001b[36m532131\u001b[0m \u001b[0m                                                      ||\n",
            "||    Meta-features : \u001b[36m57132\u001b[0m \u001b[0m                                                  ||\n",
            "||    Chromosomes/contigs : \u001b[36m22\u001b[0m \u001b[0m                                               ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "|| Process BAM file SRR17159276_sorted.bam... \u001b[0m                                ||\n",
            "||    Paired-end reads are included. \u001b[0m                                         ||\n",
            "||    The reads are assigned on the single-end mode. \u001b[0m                         ||\n",
            "||    Total alignments : \u001b[36m39023740\u001b[0m \u001b[0m                                            ||\n",
            "||    Successfully assigned alignments : \u001b[36m22959129 (58.8%)\u001b[0m \u001b[0m                    ||\n",
            "||    Running time : \u001b[36m0.87 minutes\u001b[0m \u001b[0m                                            ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "|| Write the final count table. \u001b[0m                                              ||\n",
            "|| Write the read assignment summary. \u001b[0m                                        ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "|| Summary of counting results can be found in file \"/content/drive/MyDrive/ \u001b[0m ||\n",
            "|| SRR17159276_counts.txt.summary\" \u001b[0m                                           ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "\\\\============================================================================//\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!apt-get install -y subread"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "OY4KRjCJr0VF",
        "outputId": "e5168f28-fe06-4a4b-ecf9-54b9ad5b4fdc"
      },
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "The following NEW packages will be installed:\n",
            "  subread\n",
            "0 upgraded, 1 newly installed, 0 to remove and 49 not upgraded.\n",
            "Need to get 565 kB of archives.\n",
            "After this operation, 1,928 kB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 subread amd64 2.0.3+dfsg-1 [565 kB]\n",
            "Fetched 565 kB in 2s (297 kB/s)\n",
            "Selecting previously unselected package subread.\n",
            "(Reading database ... 123622 files and directories currently installed.)\n",
            "Preparing to unpack .../subread_2.0.3+dfsg-1_amd64.deb ...\n",
            "Unpacking subread (2.0.3+dfsg-1) ...\n",
            "Setting up subread (2.0.3+dfsg-1) ...\n",
            "Processing triggers for man-db (2.10.2-1) ...\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!trim_galore --paired --quality 20 --length 20 /content/drive/MyDrive/SRR17158946_1.fastq /content/drive/MyDrive/SRR17158946_2.fastq -o /content/drive/MyDrive/trimmed_output/\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "38PSdBGMjkUb",
        "outputId": "1c408066-8e93-4e47-d86e-ed17302ea70b"
      },
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Multicore support not enabled. Proceeding with single-core trimming.\n",
            "Path to Cutadapt set as: 'cutadapt' (default)\n",
            "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
            "Cutadapt version: 4.9\n",
            "single-core operation.\n",
            "Proceeding with 'gzip' for decompression\n",
            "To decrease CPU usage of decompression, please install 'igzip' and run again\n",
            "\n",
            "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
            "\n",
            "Output will be written into the directory: /content/drive/MyDrive/trimmed_output/\n",
            "\n",
            "\n",
            "AUTO-DETECTING ADAPTER TYPE\n",
            "===========================\n",
            "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> /content/drive/MyDrive/SRR17158946_1.fastq <<)\n",
            "\n",
            "Found perfect matches for the following adapter sequences:\n",
            "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
            "Illumina\t13583\tAGATCGGAAGAGC\t1000000\t1.36\n",
            "Nextera\t1\tCTGTCTCTTATA\t1000000\t0.00\n",
            "smallRNA\t0\tTGGAATTCTCGG\t1000000\t0.00\n",
            "Using Illumina adapter for trimming (count: 13583). Second best hit was Nextera (count: 1)\n",
            "\n",
            "Writing report to '/content/drive/MyDrive/trimmed_output/SRR17158946_1.fastq_trimming_report.txt'\n",
            "\n",
            "SUMMARISING RUN PARAMETERS\n",
            "==========================\n",
            "Input filename: /content/drive/MyDrive/SRR17158946_1.fastq\n",
            "Trimming mode: paired-end\n",
            "Trim Galore version: 0.6.10\n",
            "Cutadapt version: 4.9\n",
            "Number of cores used for trimming: 1\n",
            "Quality Phred score cutoff: 20\n",
            "Quality encoding type selected: ASCII+33\n",
            "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
            "Maximum trimming error rate: 0.1 (default)\n",
            "Minimum required adapter overlap (stringency): 1 bp\n",
            "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
            "\n",
            "Cutadapt seems to be fairly up-to-date (version 4.9). Setting -j 1\n",
            "Writing final adapter and quality trimmed output to SRR17158946_1_trimmed.fq\n",
            "\n",
            "\n",
            "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /content/drive/MyDrive/SRR17158946_1.fastq <<< \n",
            "10000000 sequences processed\n",
            "20000000 sequences processed\n",
            "This is cutadapt 4.9 with Python 3.10.12\n",
            "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /content/drive/MyDrive/SRR17158946_1.fastq\n",
            "Processing single-end reads on 1 core ...\n",
            "Finished in 355.295 s (13.491 µs/read; 4.45 M reads/minute).\n",
            "\n",
            "=== Summary ===\n",
            "\n",
            "Total reads processed:              26,335,934\n",
            "Reads with adapters:                 9,103,994 (34.6%)\n",
            "Reads written (passing filters):    26,335,934 (100.0%)\n",
            "\n",
            "Total basepairs processed: 2,659,929,334 bp\n",
            "Quality-trimmed:                 593,410 bp (0.0%)\n",
            "Total written (filtered):  2,635,682,121 bp (99.1%)\n",
            "\n",
            "=== Adapter 1 ===\n",
            "\n",
            "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 9103994 times\n",
            "\n",
            "Minimum overlap: 1\n",
            "No. of allowed errors:\n",
            "1-9 bp: 0; 10-13 bp: 1\n",
            "\n",
            "Bases preceding removed adapters:\n",
            "  A: 23.6%\n",
            "  C: 35.1%\n",
            "  G: 24.6%\n",
            "  T: 16.7%\n",
            "  none/other: 0.0%\n",
            "\n",
            "Overview of removed sequences\n",
            "length\tcount\texpect\tmax.err\terror counts\n",
            "1\t5650534\t6583983.5\t0\t5650534\n",
            "2\t2029504\t1645995.9\t0\t2029504\n",
            "3\t540360\t411499.0\t0\t540360\n",
            "4\t159806\t102874.7\t0\t159806\n",
            "5\t77060\t25718.7\t0\t77060\n",
            "6\t50450\t6429.7\t0\t50450\n",
            "7\t45896\t1607.4\t0\t45896\n",
            "8\t42439\t401.9\t0\t42439\n",
            "9\t40002\t100.5\t0\t39736 266\n",
            "10\t37564\t25.1\t1\t36785 779\n",
            "11\t34519\t6.3\t1\t33923 596\n",
            "12\t31042\t1.6\t1\t30595 447\n",
            "13\t29953\t0.4\t1\t29577 376\n",
            "14\t27455\t0.4\t1\t27091 364\n",
            "15\t25887\t0.4\t1\t25550 337\n",
            "16\t24059\t0.4\t1\t23732 327\n",
            "17\t21619\t0.4\t1\t21340 279\n",
            "18\t19602\t0.4\t1\t19321 281\n",
            "19\t17707\t0.4\t1\t17465 242\n",
            "20\t16719\t0.4\t1\t16513 206\n",
            "21\t15189\t0.4\t1\t14973 216\n",
            "22\t13867\t0.4\t1\t13684 183\n",
            "23\t12817\t0.4\t1\t12619 198\n",
            "24\t11890\t0.4\t1\t11723 167\n",
            "25\t11298\t0.4\t1\t11153 145\n",
            "26\t9720\t0.4\t1\t9583 137\n",
            "27\t9117\t0.4\t1\t8981 136\n",
            "28\t8294\t0.4\t1\t8179 115\n",
            "29\t7147\t0.4\t1\t7027 120\n",
            "30\t6688\t0.4\t1\t6576 112\n",
            "31\t5852\t0.4\t1\t5762 90\n",
            "32\t5579\t0.4\t1\t5485 94\n",
            "33\t5344\t0.4\t1\t5244 100\n",
            "34\t5039\t0.4\t1\t4934 105\n",
            "35\t4380\t0.4\t1\t4285 95\n",
            "36\t4116\t0.4\t1\t4050 66\n",
            "37\t3842\t0.4\t1\t3785 57\n",
            "38\t3346\t0.4\t1\t3283 63\n",
            "39\t3166\t0.4\t1\t3106 60\n",
            "40\t2645\t0.4\t1\t2591 54\n",
            "41\t2604\t0.4\t1\t2557 47\n",
            "42\t2404\t0.4\t1\t2324 80\n",
            "43\t2267\t0.4\t1\t2211 56\n",
            "44\t1842\t0.4\t1\t1813 29\n",
            "45\t3305\t0.4\t1\t3239 66\n",
            "46\t334\t0.4\t1\t303 31\n",
            "47\t1390\t0.4\t1\t1338 52\n",
            "48\t2003\t0.4\t1\t1967 36\n",
            "49\t1344\t0.4\t1\t1326 18\n",
            "50\t798\t0.4\t1\t772 26\n",
            "51\t1436\t0.4\t1\t1407 29\n",
            "52\t872\t0.4\t1\t859 13\n",
            "53\t838\t0.4\t1\t824 14\n",
            "54\t931\t0.4\t1\t914 17\n",
            "55\t1240\t0.4\t1\t1204 36\n",
            "56\t644\t0.4\t1\t607 37\n",
            "57\t695\t0.4\t1\t670 25\n",
            "58\t352\t0.4\t1\t343 9\n",
            "59\t780\t0.4\t1\t753 27\n",
            "60\t228\t0.4\t1\t200 28\n",
            "61\t162\t0.4\t1\t119 43\n",
            "62\t588\t0.4\t1\t562 26\n",
            "63\t313\t0.4\t1\t303 10\n",
            "64\t109\t0.4\t1\t98 11\n",
            "65\t203\t0.4\t1\t183 20\n",
            "66\t325\t0.4\t1\t315 10\n",
            "67\t94\t0.4\t1\t74 20\n",
            "68\t178\t0.4\t1\t165 13\n",
            "69\t161\t0.4\t1\t137 24\n",
            "70\t125\t0.4\t1\t109 16\n",
            "71\t12\t0.4\t1\t2 10\n",
            "72\t35\t0.4\t1\t14 21\n",
            "73\t60\t0.4\t1\t43 17\n",
            "74\t65\t0.4\t1\t41 24\n",
            "75\t69\t0.4\t1\t37 32\n",
            "76\t43\t0.4\t1\t29 14\n",
            "77\t49\t0.4\t1\t28 21\n",
            "78\t61\t0.4\t1\t50 11\n",
            "79\t33\t0.4\t1\t20 13\n",
            "80\t58\t0.4\t1\t32 26\n",
            "81\t44\t0.4\t1\t32 12\n",
            "82\t40\t0.4\t1\t15 25\n",
            "83\t87\t0.4\t1\t77 10\n",
            "84\t40\t0.4\t1\t32 8\n",
            "85\t24\t0.4\t1\t17 7\n",
            "86\t48\t0.4\t1\t40 8\n",
            "87\t34\t0.4\t1\t20 14\n",
            "88\t29\t0.4\t1\t22 7\n",
            "89\t39\t0.4\t1\t34 5\n",
            "90\t15\t0.4\t1\t12 3\n",
            "91\t66\t0.4\t1\t45 21\n",
            "92\t25\t0.4\t1\t20 5\n",
            "93\t43\t0.4\t1\t32 11\n",
            "94\t21\t0.4\t1\t12 9\n",
            "95\t20\t0.4\t1\t10 10\n",
            "96\t36\t0.4\t1\t13 23\n",
            "97\t12\t0.4\t1\t8 4\n",
            "98\t28\t0.4\t1\t15 13\n",
            "99\t5\t0.4\t1\t2 3\n",
            "100\t8\t0.4\t1\t2 6\n",
            "101\t2767\t0.4\t1\t1 2766\n",
            "\n",
            "RUN STATISTICS FOR INPUT FILE: /content/drive/MyDrive/SRR17158946_1.fastq\n",
            "=============================================\n",
            "26335934 sequences processed in total\n",
            "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
            "\n",
            "Writing report to '/content/drive/MyDrive/trimmed_output/SRR17158946_2.fastq_trimming_report.txt'\n",
            "\n",
            "SUMMARISING RUN PARAMETERS\n",
            "==========================\n",
            "Input filename: /content/drive/MyDrive/SRR17158946_2.fastq\n",
            "Trimming mode: paired-end\n",
            "Trim Galore version: 0.6.10\n",
            "Cutadapt version: 4.9\n",
            "Number of cores used for trimming: 1\n",
            "Quality Phred score cutoff: 20\n",
            "Quality encoding type selected: ASCII+33\n",
            "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
            "Maximum trimming error rate: 0.1 (default)\n",
            "Minimum required adapter overlap (stringency): 1 bp\n",
            "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
            "\n",
            "Cutadapt seems to be fairly up-to-date (version 4.9). Setting -j -j 1\n",
            "Writing final adapter and quality trimmed output to SRR17158946_2_trimmed.fq\n",
            "\n",
            "\n",
            "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /content/drive/MyDrive/SRR17158946_2.fastq <<< \n",
            "10000000 sequences processed\n",
            "20000000 sequences processed\n",
            "This is cutadapt 4.9 with Python 3.10.12\n",
            "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /content/drive/MyDrive/SRR17158946_2.fastq\n",
            "Processing single-end reads on 1 core ...\n",
            "Finished in 421.247 s (15.995 µs/read; 3.75 M reads/minute).\n",
            "\n",
            "=== Summary ===\n",
            "\n",
            "Total reads processed:              26,335,934\n",
            "Reads with adapters:                 9,908,693 (37.6%)\n",
            "Reads written (passing filters):    26,335,934 (100.0%)\n",
            "\n",
            "Total basepairs processed: 2,659,929,334 bp\n",
            "Quality-trimmed:               6,440,692 bp (0.2%)\n",
            "Total written (filtered):  2,628,903,516 bp (98.8%)\n",
            "\n",
            "=== Adapter 1 ===\n",
            "\n",
            "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 9908693 times\n",
            "\n",
            "Minimum overlap: 1\n",
            "No. of allowed errors:\n",
            "1-9 bp: 0; 10-13 bp: 1\n",
            "\n",
            "Bases preceding removed adapters:\n",
            "  A: 29.5%\n",
            "  C: 31.5%\n",
            "  G: 24.8%\n",
            "  T: 14.2%\n",
            "  none/other: 0.0%\n",
            "\n",
            "Overview of removed sequences\n",
            "length\tcount\texpect\tmax.err\terror counts\n",
            "1\t6212447\t6583983.5\t0\t6212447\n",
            "2\t2031347\t1645995.9\t0\t2031347\n",
            "3\t777525\t411499.0\t0\t777525\n",
            "4\t168647\t102874.7\t0\t168647\n",
            "5\t74615\t25718.7\t0\t74615\n",
            "6\t52009\t6429.7\t0\t52009\n",
            "7\t46215\t1607.4\t0\t46215\n",
            "8\t44612\t401.9\t0\t44612\n",
            "9\t37758\t100.5\t0\t37149 609\n",
            "10\t39278\t25.1\t1\t37969 1309\n",
            "11\t33681\t6.3\t1\t32537 1144\n",
            "12\t31251\t1.6\t1\t30786 465\n",
            "13\t29753\t0.4\t1\t29474 279\n",
            "14\t28552\t0.4\t1\t28248 304\n",
            "15\t24695\t0.4\t1\t24473 222\n",
            "16\t23542\t0.4\t1\t23312 230\n",
            "17\t23760\t0.4\t1\t23487 273\n",
            "18\t16859\t0.4\t1\t16677 182\n",
            "19\t21138\t0.4\t1\t20890 248\n",
            "20\t14929\t0.4\t1\t14793 136\n",
            "21\t13417\t0.4\t1\t13295 122\n",
            "22\t13819\t0.4\t1\t13686 133\n",
            "23\t12649\t0.4\t1\t12512 137\n",
            "24\t12209\t0.4\t1\t12041 168\n",
            "25\t11605\t0.4\t1\t11498 107\n",
            "26\t9503\t0.4\t1\t9413 90\n",
            "27\t8706\t0.4\t1\t8561 145\n",
            "28\t8896\t0.4\t1\t8811 85\n",
            "29\t6342\t0.4\t1\t6271 71\n",
            "30\t7207\t0.4\t1\t7139 68\n",
            "31\t5275\t0.4\t1\t5223 52\n",
            "32\t5524\t0.4\t1\t5461 63\n",
            "33\t5340\t0.4\t1\t5275 65\n",
            "34\t5094\t0.4\t1\t5038 56\n",
            "35\t4104\t0.4\t1\t4066 38\n",
            "36\t4363\t0.4\t1\t4300 63\n",
            "37\t3520\t0.4\t1\t3456 64\n",
            "38\t3526\t0.4\t1\t3484 42\n",
            "39\t3330\t0.4\t1\t3279 51\n",
            "40\t2443\t0.4\t1\t2399 44\n",
            "41\t2492\t0.4\t1\t2457 35\n",
            "42\t2685\t0.4\t1\t2653 32\n",
            "43\t1887\t0.4\t1\t1851 36\n",
            "44\t5945\t0.4\t1\t5885 60\n",
            "45\t2166\t0.4\t1\t2134 32\n",
            "46\t1056\t0.4\t1\t1035 21\n",
            "47\t1234\t0.4\t1\t1215 19\n",
            "48\t360\t0.4\t1\t348 12\n",
            "49\t1494\t0.4\t1\t1466 28\n",
            "50\t222\t0.4\t1\t210 12\n",
            "51\t382\t0.4\t1\t369 13\n",
            "52\t284\t0.4\t1\t278 6\n",
            "53\t947\t0.4\t1\t924 23\n",
            "54\t1281\t0.4\t1\t1258 23\n",
            "55\t253\t0.4\t1\t241 12\n",
            "56\t623\t0.4\t1\t612 11\n",
            "57\t1637\t0.4\t1\t1616 21\n",
            "58\t304\t0.4\t1\t290 14\n",
            "59\t230\t0.4\t1\t210 20\n",
            "60\t950\t0.4\t1\t930 20\n",
            "61\t193\t0.4\t1\t173 20\n",
            "62\t110\t0.4\t1\t99 11\n",
            "63\t594\t0.4\t1\t582 12\n",
            "64\t515\t0.4\t1\t503 12\n",
            "65\t53\t0.4\t1\t46 7\n",
            "66\t91\t0.4\t1\t82 9\n",
            "67\t51\t0.4\t1\t37 14\n",
            "68\t53\t0.4\t1\t41 12\n",
            "69\t43\t0.4\t1\t34 9\n",
            "70\t79\t0.4\t1\t64 15\n",
            "71\t33\t0.4\t1\t26 7\n",
            "72\t46\t0.4\t1\t36 10\n",
            "73\t45\t0.4\t1\t38 7\n",
            "74\t34\t0.4\t1\t28 6\n",
            "75\t33\t0.4\t1\t18 15\n",
            "76\t13\t0.4\t1\t11 2\n",
            "77\t28\t0.4\t1\t15 13\n",
            "78\t53\t0.4\t1\t42 11\n",
            "79\t35\t0.4\t1\t26 9\n",
            "80\t46\t0.4\t1\t27 19\n",
            "81\t46\t0.4\t1\t31 15\n",
            "82\t24\t0.4\t1\t19 5\n",
            "83\t87\t0.4\t1\t74 13\n",
            "84\t36\t0.4\t1\t29 7\n",
            "85\t20\t0.4\t1\t14 6\n",
            "86\t46\t0.4\t1\t39 7\n",
            "87\t27\t0.4\t1\t21 6\n",
            "88\t37\t0.4\t1\t22 15\n",
            "89\t38\t0.4\t1\t31 7\n",
            "90\t20\t0.4\t1\t12 8\n",
            "91\t48\t0.4\t1\t45 3\n",
            "92\t37\t0.4\t1\t17 20\n",
            "93\t34\t0.4\t1\t31 3\n",
            "94\t17\t0.4\t1\t12 5\n",
            "95\t13\t0.4\t1\t9 4\n",
            "96\t27\t0.4\t1\t12 15\n",
            "97\t10\t0.4\t1\t7 3\n",
            "98\t39\t0.4\t1\t14 25\n",
            "99\t4\t0.4\t1\t1 3\n",
            "100\t12\t0.4\t1\t2 10\n",
            "101\t6\t0.4\t1\t1 5\n",
            "\n",
            "RUN STATISTICS FOR INPUT FILE: /content/drive/MyDrive/SRR17158946_2.fastq\n",
            "=============================================\n",
            "26335934 sequences processed in total\n",
            "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
            "\n",
            "Validate paired-end files SRR17158946_1_trimmed.fq and SRR17158946_2_trimmed.fq\n",
            "file_1: SRR17158946_1_trimmed.fq, file_2: SRR17158946_2_trimmed.fq\n",
            "\n",
            "\n",
            ">>>>> Now validing the length of the 2 paired-end infiles: SRR17158946_1_trimmed.fq and SRR17158946_2_trimmed.fq <<<<<\n",
            "Writing validated paired-end Read 1 reads to SRR17158946_1_val_1.fq\n",
            "Writing validated paired-end Read 2 reads to SRR17158946_2_val_2.fq\n",
            "\n",
            "Total number of sequences analysed: 26335934\n",
            "\n",
            "Number of sequence pairs removed because at least one read was shorter than the length cutoff (20 bp): 48230 (0.18%)\n",
            "\n",
            "Deleting both intermediate output files SRR17158946_1_trimmed.fq and SRR17158946_2_trimmed.fq\n",
            "\n",
            "====================================================================================================\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!trim_galore --paired --quality 20 --length 20 /content/drive/MyDrive/SRR17158978_1.fastq /content/drive/MyDrive/SRR17158978_2.fastq -o /content/drive/MyDrive/trimmed_output/\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "93vsivYJl0Zd",
        "outputId": "23ae6aa2-a2bc-4e1a-fa6c-cda77999c682"
      },
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Multicore support not enabled. Proceeding with single-core trimming.\n",
            "Path to Cutadapt set as: 'cutadapt' (default)\n",
            "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
            "Cutadapt version: 4.9\n",
            "single-core operation.\n",
            "Proceeding with 'gzip' for decompression\n",
            "To decrease CPU usage of decompression, please install 'igzip' and run again\n",
            "\n",
            "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
            "\n",
            "Output will be written into the directory: /content/drive/MyDrive/trimmed_output/\n",
            "\n",
            "\n",
            "AUTO-DETECTING ADAPTER TYPE\n",
            "===========================\n",
            "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> /content/drive/MyDrive/SRR17158978_1.fastq <<)\n",
            "\n",
            "Found perfect matches for the following adapter sequences:\n",
            "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
            "Illumina\t19674\tAGATCGGAAGAGC\t1000000\t1.97\n",
            "Nextera\t1\tCTGTCTCTTATA\t1000000\t0.00\n",
            "smallRNA\t0\tTGGAATTCTCGG\t1000000\t0.00\n",
            "Using Illumina adapter for trimming (count: 19674). Second best hit was Nextera (count: 1)\n",
            "\n",
            "Writing report to '/content/drive/MyDrive/trimmed_output/SRR17158978_1.fastq_trimming_report.txt'\n",
            "\n",
            "SUMMARISING RUN PARAMETERS\n",
            "==========================\n",
            "Input filename: /content/drive/MyDrive/SRR17158978_1.fastq\n",
            "Trimming mode: paired-end\n",
            "Trim Galore version: 0.6.10\n",
            "Cutadapt version: 4.9\n",
            "Number of cores used for trimming: 1\n",
            "Quality Phred score cutoff: 20\n",
            "Quality encoding type selected: ASCII+33\n",
            "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
            "Maximum trimming error rate: 0.1 (default)\n",
            "Minimum required adapter overlap (stringency): 1 bp\n",
            "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
            "\n",
            "Cutadapt seems to be fairly up-to-date (version 4.9). Setting -j 1\n",
            "Writing final adapter and quality trimmed output to SRR17158978_1_trimmed.fq\n",
            "\n",
            "\n",
            "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /content/drive/MyDrive/SRR17158978_1.fastq <<< \n",
            "10000000 sequences processed\n",
            "This is cutadapt 4.9 with Python 3.10.12\n",
            "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /content/drive/MyDrive/SRR17158978_1.fastq\n",
            "Processing single-end reads on 1 core ...\n",
            "Finished in 284.066 s (15.104 µs/read; 3.97 M reads/minute).\n",
            "\n",
            "=== Summary ===\n",
            "\n",
            "Total reads processed:              18,807,831\n",
            "Reads with adapters:                 6,733,332 (35.8%)\n",
            "Reads written (passing filters):    18,807,831 (100.0%)\n",
            "\n",
            "Total basepairs processed: 1,899,590,931 bp\n",
            "Quality-trimmed:                 415,739 bp (0.0%)\n",
            "Total written (filtered):  1,879,514,913 bp (98.9%)\n",
            "\n",
            "=== Adapter 1 ===\n",
            "\n",
            "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 6733332 times\n",
            "\n",
            "Minimum overlap: 1\n",
            "No. of allowed errors:\n",
            "1-9 bp: 0; 10-13 bp: 1\n",
            "\n",
            "Bases preceding removed adapters:\n",
            "  A: 23.8%\n",
            "  C: 33.9%\n",
            "  G: 25.1%\n",
            "  T: 17.1%\n",
            "  none/other: 0.0%\n",
            "\n",
            "Overview of removed sequences\n",
            "length\tcount\texpect\tmax.err\terror counts\n",
            "1\t4051905\t4701957.8\t0\t4051905\n",
            "2\t1424444\t1175489.4\t0\t1424444\n",
            "3\t400418\t293872.4\t0\t400418\n",
            "4\t125811\t73468.1\t0\t125811\n",
            "5\t74386\t18367.0\t0\t74386\n",
            "6\t54167\t4591.8\t0\t54167\n",
            "7\t49102\t1147.9\t0\t49102\n",
            "8\t46553\t287.0\t0\t46553\n",
            "9\t43005\t71.7\t0\t42804 201\n",
            "10\t39840\t17.9\t1\t39241 599\n",
            "11\t36547\t4.5\t1\t36004 543\n",
            "12\t33389\t1.1\t1\t33027 362\n",
            "13\t31851\t0.3\t1\t31513 338\n",
            "14\t30044\t0.3\t1\t29673 371\n",
            "15\t27570\t0.3\t1\t27237 333\n",
            "16\t25467\t0.3\t1\t25126 341\n",
            "17\t22773\t0.3\t1\t22441 332\n",
            "18\t20558\t0.3\t1\t20337 221\n",
            "19\t18023\t0.3\t1\t17831 192\n",
            "20\t17012\t0.3\t1\t16822 190\n",
            "21\t15173\t0.3\t1\t15007 166\n",
            "22\t13860\t0.3\t1\t13720 140\n",
            "23\t12892\t0.3\t1\t12725 167\n",
            "24\t11994\t0.3\t1\t11835 159\n",
            "25\t10566\t0.3\t1\t10440 126\n",
            "26\t9442\t0.3\t1\t9329 113\n",
            "27\t8697\t0.3\t1\t8601 96\n",
            "28\t7576\t0.3\t1\t7489 87\n",
            "29\t6883\t0.3\t1\t6784 99\n",
            "30\t6138\t0.3\t1\t6062 76\n",
            "31\t5286\t0.3\t1\t5223 63\n",
            "32\t4886\t0.3\t1\t4820 66\n",
            "33\t4449\t0.3\t1\t4377 72\n",
            "34\t4370\t0.3\t1\t4310 60\n",
            "35\t3583\t0.3\t1\t3536 47\n",
            "36\t3428\t0.3\t1\t3392 36\n",
            "37\t2942\t0.3\t1\t2907 35\n",
            "38\t2916\t0.3\t1\t2884 32\n",
            "39\t2475\t0.3\t1\t2438 37\n",
            "40\t2299\t0.3\t1\t2253 46\n",
            "41\t1970\t0.3\t1\t1941 29\n",
            "42\t1781\t0.3\t1\t1739 42\n",
            "43\t1612\t0.3\t1\t1560 52\n",
            "44\t1377\t0.3\t1\t1363 14\n",
            "45\t2237\t0.3\t1\t2211 26\n",
            "46\t270\t0.3\t1\t256 14\n",
            "47\t836\t0.3\t1\t803 33\n",
            "48\t1042\t0.3\t1\t1033 9\n",
            "49\t843\t0.3\t1\t835 8\n",
            "50\t667\t0.3\t1\t659 8\n",
            "51\t640\t0.3\t1\t624 16\n",
            "52\t574\t0.3\t1\t566 8\n",
            "53\t556\t0.3\t1\t546 10\n",
            "54\t476\t0.3\t1\t458 18\n",
            "55\t456\t0.3\t1\t447 9\n",
            "56\t331\t0.3\t1\t301 30\n",
            "57\t328\t0.3\t1\t313 15\n",
            "58\t258\t0.3\t1\t236 22\n",
            "59\t253\t0.3\t1\t235 18\n",
            "60\t226\t0.3\t1\t205 21\n",
            "61\t103\t0.3\t1\t86 17\n",
            "62\t146\t0.3\t1\t133 13\n",
            "63\t127\t0.3\t1\t110 17\n",
            "64\t97\t0.3\t1\t80 17\n",
            "65\t60\t0.3\t1\t43 17\n",
            "66\t84\t0.3\t1\t72 12\n",
            "67\t51\t0.3\t1\t35 16\n",
            "68\t62\t0.3\t1\t52 10\n",
            "69\t30\t0.3\t1\t24 6\n",
            "70\t33\t0.3\t1\t21 12\n",
            "71\t8\t0.3\t1\t2 6\n",
            "72\t18\t0.3\t1\t5 13\n",
            "73\t27\t0.3\t1\t13 14\n",
            "74\t28\t0.3\t1\t21 7\n",
            "75\t32\t0.3\t1\t18 14\n",
            "76\t31\t0.3\t1\t17 14\n",
            "77\t29\t0.3\t1\t19 10\n",
            "78\t17\t0.3\t1\t10 7\n",
            "79\t23\t0.3\t1\t13 10\n",
            "80\t30\t0.3\t1\t13 17\n",
            "81\t33\t0.3\t1\t16 17\n",
            "82\t25\t0.3\t1\t6 19\n",
            "83\t43\t0.3\t1\t36 7\n",
            "84\t23\t0.3\t1\t16 7\n",
            "85\t13\t0.3\t1\t10 3\n",
            "86\t18\t0.3\t1\t9 9\n",
            "87\t26\t0.3\t1\t16 10\n",
            "88\t22\t0.3\t1\t10 12\n",
            "89\t31\t0.3\t1\t25 6\n",
            "90\t13\t0.3\t1\t6 7\n",
            "91\t37\t0.3\t1\t30 7\n",
            "92\t20\t0.3\t1\t17 3\n",
            "93\t25\t0.3\t1\t21 4\n",
            "94\t9\t0.3\t1\t6 3\n",
            "95\t24\t0.3\t1\t8 16\n",
            "96\t25\t0.3\t1\t7 18\n",
            "97\t8\t0.3\t1\t5 3\n",
            "98\t23\t0.3\t1\t8 15\n",
            "99\t5\t0.3\t1\t2 3\n",
            "100\t7\t0.3\t1\t2 5\n",
            "101\t2413\t0.3\t1\t9 2404\n",
            "\n",
            "RUN STATISTICS FOR INPUT FILE: /content/drive/MyDrive/SRR17158978_1.fastq\n",
            "=============================================\n",
            "18807831 sequences processed in total\n",
            "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
            "\n",
            "Writing report to '/content/drive/MyDrive/trimmed_output/SRR17158978_2.fastq_trimming_report.txt'\n",
            "\n",
            "SUMMARISING RUN PARAMETERS\n",
            "==========================\n",
            "Input filename: /content/drive/MyDrive/SRR17158978_2.fastq\n",
            "Trimming mode: paired-end\n",
            "Trim Galore version: 0.6.10\n",
            "Cutadapt version: 4.9\n",
            "Number of cores used for trimming: 1\n",
            "Quality Phred score cutoff: 20\n",
            "Quality encoding type selected: ASCII+33\n",
            "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
            "Maximum trimming error rate: 0.1 (default)\n",
            "Minimum required adapter overlap (stringency): 1 bp\n",
            "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
            "\n",
            "Cutadapt seems to be fairly up-to-date (version 4.9). Setting -j -j 1\n",
            "Writing final adapter and quality trimmed output to SRR17158978_2_trimmed.fq\n",
            "\n",
            "\n",
            "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /content/drive/MyDrive/SRR17158978_2.fastq <<< \n",
            "10000000 sequences processed\n",
            "This is cutadapt 4.9 with Python 3.10.12\n",
            "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /content/drive/MyDrive/SRR17158978_2.fastq\n",
            "Processing single-end reads on 1 core ...\n",
            "Finished in 305.933 s (16.266 µs/read; 3.69 M reads/minute).\n",
            "\n",
            "=== Summary ===\n",
            "\n",
            "Total reads processed:              18,807,831\n",
            "Reads with adapters:                 7,388,575 (39.3%)\n",
            "Reads written (passing filters):    18,807,831 (100.0%)\n",
            "\n",
            "Total basepairs processed: 1,899,590,931 bp\n",
            "Quality-trimmed:               2,663,626 bp (0.1%)\n",
            "Total written (filtered):  1,876,457,884 bp (98.8%)\n",
            "\n",
            "=== Adapter 1 ===\n",
            "\n",
            "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 7388575 times\n",
            "\n",
            "Minimum overlap: 1\n",
            "No. of allowed errors:\n",
            "1-9 bp: 0; 10-13 bp: 1\n",
            "\n",
            "Bases preceding removed adapters:\n",
            "  A: 29.3%\n",
            "  C: 31.1%\n",
            "  G: 25.1%\n",
            "  T: 14.5%\n",
            "  none/other: 0.0%\n",
            "\n",
            "Overview of removed sequences\n",
            "length\tcount\texpect\tmax.err\terror counts\n",
            "1\t4491040\t4701957.8\t0\t4491040\n",
            "2\t1439101\t1175489.4\t0\t1439101\n",
            "3\t590804\t293872.4\t0\t590804\n",
            "4\t139474\t73468.1\t0\t139474\n",
            "5\t72825\t18367.0\t0\t72825\n",
            "6\t55502\t4591.8\t0\t55502\n",
            "7\t49431\t1147.9\t0\t49431\n",
            "8\t48550\t287.0\t0\t48550\n",
            "9\t41015\t71.7\t0\t40412 603\n",
            "10\t41697\t17.9\t1\t40711 986\n",
            "11\t35465\t4.5\t1\t34629 836\n",
            "12\t33595\t1.1\t1\t33227 368\n",
            "13\t31953\t0.3\t1\t31662 291\n",
            "14\t31548\t0.3\t1\t31267 281\n",
            "15\t26089\t0.3\t1\t25871 218\n",
            "16\t24979\t0.3\t1\t24765 214\n",
            "17\t26492\t0.3\t1\t26256 236\n",
            "18\t16120\t0.3\t1\t15987 133\n",
            "19\t19859\t0.3\t1\t19653 206\n",
            "20\t17041\t0.3\t1\t16896 145\n",
            "21\t13625\t0.3\t1\t13515 110\n",
            "22\t13869\t0.3\t1\t13757 112\n",
            "23\t12841\t0.3\t1\t12715 126\n",
            "24\t12173\t0.3\t1\t12049 124\n",
            "25\t10806\t0.3\t1\t10702 104\n",
            "26\t9275\t0.3\t1\t9187 88\n",
            "27\t8511\t0.3\t1\t8412 99\n",
            "28\t8050\t0.3\t1\t7984 66\n",
            "29\t6191\t0.3\t1\t6124 67\n",
            "30\t6370\t0.3\t1\t6299 71\n",
            "31\t5093\t0.3\t1\t5044 49\n",
            "32\t4756\t0.3\t1\t4722 34\n",
            "33\t4616\t0.3\t1\t4555 61\n",
            "34\t4278\t0.3\t1\t4220 58\n",
            "35\t3523\t0.3\t1\t3488 35\n",
            "36\t3767\t0.3\t1\t3724 43\n",
            "37\t2654\t0.3\t1\t2623 31\n",
            "38\t2885\t0.3\t1\t2855 30\n",
            "39\t2774\t0.3\t1\t2747 27\n",
            "40\t1926\t0.3\t1\t1907 19\n",
            "41\t1936\t0.3\t1\t1910 26\n",
            "42\t2207\t0.3\t1\t2171 36\n",
            "43\t1114\t0.3\t1\t1107 7\n",
            "44\t1605\t0.3\t1\t1582 23\n",
            "45\t1367\t0.3\t1\t1354 13\n",
            "46\t1511\t0.3\t1\t1492 19\n",
            "47\t1493\t0.3\t1\t1476 17\n",
            "48\t337\t0.3\t1\t324 13\n",
            "49\t1088\t0.3\t1\t1068 20\n",
            "50\t412\t0.3\t1\t406 6\n",
            "51\t337\t0.3\t1\t328 9\n",
            "52\t394\t0.3\t1\t385 9\n",
            "53\t564\t0.3\t1\t552 12\n",
            "54\t745\t0.3\t1\t736 9\n",
            "55\t124\t0.3\t1\t118 6\n",
            "56\t349\t0.3\t1\t341 8\n",
            "57\t719\t0.3\t1\t695 24\n",
            "58\t135\t0.3\t1\t126 9\n",
            "59\t56\t0.3\t1\t49 7\n",
            "60\t354\t0.3\t1\t350 4\n",
            "61\t71\t0.3\t1\t55 16\n",
            "62\t42\t0.3\t1\t40 2\n",
            "63\t223\t0.3\t1\t215 8\n",
            "64\t85\t0.3\t1\t79 6\n",
            "65\t27\t0.3\t1\t18 9\n",
            "66\t26\t0.3\t1\t25 1\n",
            "67\t15\t0.3\t1\t13 2\n",
            "68\t20\t0.3\t1\t14 6\n",
            "69\t19\t0.3\t1\t11 8\n",
            "70\t34\t0.3\t1\t23 11\n",
            "71\t18\t0.3\t1\t6 12\n",
            "72\t27\t0.3\t1\t21 6\n",
            "73\t22\t0.3\t1\t18 4\n",
            "74\t24\t0.3\t1\t23 1\n",
            "75\t15\t0.3\t1\t7 8\n",
            "76\t11\t0.3\t1\t6 5\n",
            "77\t18\t0.3\t1\t14 4\n",
            "78\t23\t0.3\t1\t10 13\n",
            "79\t18\t0.3\t1\t13 5\n",
            "80\t23\t0.3\t1\t11 12\n",
            "81\t30\t0.3\t1\t17 13\n",
            "82\t23\t0.3\t1\t5 18\n",
            "83\t39\t0.3\t1\t34 5\n",
            "84\t18\t0.3\t1\t17 1\n",
            "85\t13\t0.3\t1\t11 2\n",
            "86\t20\t0.3\t1\t10 10\n",
            "87\t22\t0.3\t1\t15 7\n",
            "88\t16\t0.3\t1\t8 8\n",
            "89\t30\t0.3\t1\t25 5\n",
            "90\t10\t0.3\t1\t7 3\n",
            "91\t29\t0.3\t1\t28 1\n",
            "92\t27\t0.3\t1\t16 11\n",
            "93\t24\t0.3\t1\t21 3\n",
            "94\t12\t0.3\t1\t6 6\n",
            "95\t12\t0.3\t1\t5 7\n",
            "96\t7\t0.3\t1\t6 1\n",
            "97\t10\t0.3\t1\t5 5\n",
            "98\t32\t0.3\t1\t7 25\n",
            "99\t10\t0.3\t1\t1 9\n",
            "100\t9\t0.3\t1\t2 7\n",
            "101\t11\t0.3\t1\t9 2\n",
            "\n",
            "RUN STATISTICS FOR INPUT FILE: /content/drive/MyDrive/SRR17158978_2.fastq\n",
            "=============================================\n",
            "18807831 sequences processed in total\n",
            "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
            "\n",
            "Validate paired-end files SRR17158978_1_trimmed.fq and SRR17158978_2_trimmed.fq\n",
            "file_1: SRR17158978_1_trimmed.fq, file_2: SRR17158978_2_trimmed.fq\n",
            "\n",
            "\n",
            ">>>>> Now validing the length of the 2 paired-end infiles: SRR17158978_1_trimmed.fq and SRR17158978_2_trimmed.fq <<<<<\n",
            "Writing validated paired-end Read 1 reads to SRR17158978_1_val_1.fq\n",
            "Writing validated paired-end Read 2 reads to SRR17158978_2_val_2.fq\n",
            "\n",
            "Total number of sequences analysed: 18807831\n",
            "\n",
            "Number of sequence pairs removed because at least one read was shorter than the length cutoff (20 bp): 18699 (0.10%)\n",
            "\n",
            "Deleting both intermediate output files SRR17158978_1_trimmed.fq and SRR17158978_2_trimmed.fq\n",
            "\n",
            "====================================================================================================\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!trim_galore --paired --quality 20 --length 20 /content/drive/MyDrive/SRR17158944_1.fastq /content/drive/MyDrive/SRR17158944_2.fastq -o /content/drive/MyDrive/trimmed_output/\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "PXk9HoyhuvDF",
        "outputId": "e27b2677-d233-4272-c664-bddb0d5decfc"
      },
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Multicore support not enabled. Proceeding with single-core trimming.\n",
            "Path to Cutadapt set as: 'cutadapt' (default)\n",
            "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
            "Cutadapt version: 4.9\n",
            "single-core operation.\n",
            "Proceeding with 'gzip' for decompression\n",
            "To decrease CPU usage of decompression, please install 'igzip' and run again\n",
            "\n",
            "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
            "\n",
            "Output will be written into the directory: /content/drive/MyDrive/trimmed_output/\n",
            "\n",
            "\n",
            "AUTO-DETECTING ADAPTER TYPE\n",
            "===========================\n",
            "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> /content/drive/MyDrive/SRR17158944_1.fastq <<)\n",
            "\n",
            "Found perfect matches for the following adapter sequences:\n",
            "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
            "Illumina\t16616\tAGATCGGAAGAGC\t1000000\t1.66\n",
            "Nextera\t0\tCTGTCTCTTATA\t1000000\t0.00\n",
            "smallRNA\t0\tTGGAATTCTCGG\t1000000\t0.00\n",
            "Using Illumina adapter for trimming (count: 16616). Second best hit was Nextera (count: 0)\n",
            "\n",
            "Writing report to '/content/drive/MyDrive/trimmed_output/SRR17158944_1.fastq_trimming_report.txt'\n",
            "\n",
            "SUMMARISING RUN PARAMETERS\n",
            "==========================\n",
            "Input filename: /content/drive/MyDrive/SRR17158944_1.fastq\n",
            "Trimming mode: paired-end\n",
            "Trim Galore version: 0.6.10\n",
            "Cutadapt version: 4.9\n",
            "Number of cores used for trimming: 1\n",
            "Quality Phred score cutoff: 20\n",
            "Quality encoding type selected: ASCII+33\n",
            "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
            "Maximum trimming error rate: 0.1 (default)\n",
            "Minimum required adapter overlap (stringency): 1 bp\n",
            "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
            "\n",
            "Cutadapt seems to be fairly up-to-date (version 4.9). Setting -j 1\n",
            "Writing final adapter and quality trimmed output to SRR17158944_1_trimmed.fq\n",
            "\n",
            "\n",
            "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /content/drive/MyDrive/SRR17158944_1.fastq <<< \n",
            "10000000 sequences processed\n",
            "20000000 sequences processed\n",
            "30000000 sequences processed\n",
            "This is cutadapt 4.9 with Python 3.10.12\n",
            "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /content/drive/MyDrive/SRR17158944_1.fastq\n",
            "Processing single-end reads on 1 core ...\n",
            "Finished in 550.526 s (13.865 µs/read; 4.33 M reads/minute).\n",
            "\n",
            "=== Summary ===\n",
            "\n",
            "Total reads processed:              39,706,568\n",
            "Reads with adapters:                13,859,071 (34.9%)\n",
            "Reads written (passing filters):    39,706,568 (100.0%)\n",
            "\n",
            "Total basepairs processed: 4,010,363,368 bp\n",
            "Quality-trimmed:                 883,370 bp (0.0%)\n",
            "Total written (filtered):  3,969,816,836 bp (99.0%)\n",
            "\n",
            "=== Adapter 1 ===\n",
            "\n",
            "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 13859071 times\n",
            "\n",
            "Minimum overlap: 1\n",
            "No. of allowed errors:\n",
            "1-9 bp: 0; 10-13 bp: 1\n",
            "\n",
            "Bases preceding removed adapters:\n",
            "  A: 23.6%\n",
            "  C: 35.1%\n",
            "  G: 24.3%\n",
            "  T: 17.0%\n",
            "  none/other: 0.0%\n",
            "\n",
            "Overview of removed sequences\n",
            "length\tcount\texpect\tmax.err\terror counts\n",
            "1\t8484264\t9926642.0\t0\t8484264\n",
            "2\t2998065\t2481660.5\t0\t2998065\n",
            "3\t817214\t620415.1\t0\t817214\n",
            "4\t252001\t155103.8\t0\t252001\n",
            "5\t123512\t38775.9\t0\t123512\n",
            "6\t83646\t9694.0\t0\t83646\n",
            "7\t76674\t2423.5\t0\t76674\n",
            "8\t71967\t605.9\t0\t71967\n",
            "9\t69631\t151.5\t0\t69245 386\n",
            "10\t66741\t37.9\t1\t63929 2812\n",
            "11\t61240\t9.5\t1\t58891 2349\n",
            "12\t55695\t2.4\t1\t53645 2050\n",
            "13\t53574\t0.6\t1\t51775 1799\n",
            "14\t50349\t0.6\t1\t48518 1831\n",
            "15\t47278\t0.6\t1\t45651 1627\n",
            "16\t44480\t0.6\t1\t42806 1674\n",
            "17\t40567\t0.6\t1\t39083 1484\n",
            "18\t36939\t0.6\t1\t35642 1297\n",
            "19\t34078\t0.6\t1\t32940 1138\n",
            "20\t31968\t0.6\t1\t30816 1152\n",
            "21\t29324\t0.6\t1\t28263 1061\n",
            "22\t27431\t0.6\t1\t26480 951\n",
            "23\t25479\t0.6\t1\t24537 942\n",
            "24\t23568\t0.6\t1\t22650 918\n",
            "25\t22347\t0.6\t1\t21545 802\n",
            "26\t19561\t0.6\t1\t18808 753\n",
            "27\t18218\t0.6\t1\t17540 678\n",
            "28\t17104\t0.6\t1\t16479 625\n",
            "29\t14955\t0.6\t1\t14370 585\n",
            "30\t14169\t0.6\t1\t13640 529\n",
            "31\t12268\t0.6\t1\t11858 410\n",
            "32\t11603\t0.6\t1\t11162 441\n",
            "33\t10516\t0.6\t1\t10121 395\n",
            "34\t10051\t0.6\t1\t9692 359\n",
            "35\t9399\t0.6\t1\t9018 381\n",
            "36\t8045\t0.6\t1\t7737 308\n",
            "37\t7937\t0.6\t1\t7610 327\n",
            "38\t6479\t0.6\t1\t6241 238\n",
            "39\t6354\t0.6\t1\t6112 242\n",
            "40\t5755\t0.6\t1\t5534 221\n",
            "41\t5121\t0.6\t1\t4925 196\n",
            "42\t4796\t0.6\t1\t4597 199\n",
            "43\t4605\t0.6\t1\t4430 175\n",
            "44\t3543\t0.6\t1\t3411 132\n",
            "45\t6410\t0.6\t1\t6157 253\n",
            "46\t805\t0.6\t1\t742 63\n",
            "47\t2440\t0.6\t1\t2341 99\n",
            "48\t4003\t0.6\t1\t3862 141\n",
            "49\t2392\t0.6\t1\t2291 101\n",
            "50\t1455\t0.6\t1\t1386 69\n",
            "51\t2640\t0.6\t1\t2540 100\n",
            "52\t1613\t0.6\t1\t1530 83\n",
            "53\t1474\t0.6\t1\t1415 59\n",
            "54\t1866\t0.6\t1\t1780 86\n",
            "55\t2317\t0.6\t1\t2208 109\n",
            "56\t1102\t0.6\t1\t1036 66\n",
            "57\t1224\t0.6\t1\t1168 56\n",
            "58\t655\t0.6\t1\t627 28\n",
            "59\t1601\t0.6\t1\t1504 97\n",
            "60\t437\t0.6\t1\t382 55\n",
            "61\t266\t0.6\t1\t189 77\n",
            "62\t1007\t0.6\t1\t943 64\n",
            "63\t537\t0.6\t1\t497 40\n",
            "64\t175\t0.6\t1\t144 31\n",
            "65\t327\t0.6\t1\t287 40\n",
            "66\t615\t0.6\t1\t575 40\n",
            "67\t162\t0.6\t1\t144 18\n",
            "68\t315\t0.6\t1\t293 22\n",
            "69\t264\t0.6\t1\t233 31\n",
            "70\t225\t0.6\t1\t180 45\n",
            "71\t21\t0.6\t1\t8 13\n",
            "72\t50\t0.6\t1\t17 33\n",
            "73\t104\t0.6\t1\t74 30\n",
            "74\t91\t0.6\t1\t60 31\n",
            "75\t111\t0.6\t1\t69 42\n",
            "76\t106\t0.6\t1\t54 52\n",
            "77\t77\t0.6\t1\t42 35\n",
            "78\t84\t0.6\t1\t71 13\n",
            "79\t58\t0.6\t1\t44 14\n",
            "80\t95\t0.6\t1\t54 41\n",
            "81\t55\t0.6\t1\t28 27\n",
            "82\t66\t0.6\t1\t35 31\n",
            "83\t150\t0.6\t1\t138 12\n",
            "84\t54\t0.6\t1\t45 9\n",
            "85\t54\t0.6\t1\t43 11\n",
            "86\t72\t0.6\t1\t57 15\n",
            "87\t57\t0.6\t1\t41 16\n",
            "88\t68\t0.6\t1\t48 20\n",
            "89\t88\t0.6\t1\t75 13\n",
            "90\t53\t0.6\t1\t36 17\n",
            "91\t122\t0.6\t1\t108 14\n",
            "92\t43\t0.6\t1\t36 7\n",
            "93\t87\t0.6\t1\t69 18\n",
            "94\t33\t0.6\t1\t27 6\n",
            "95\t50\t0.6\t1\t13 37\n",
            "96\t82\t0.6\t1\t39 43\n",
            "97\t38\t0.6\t1\t29 9\n",
            "98\t49\t0.6\t1\t29 20\n",
            "99\t14\t0.6\t1\t9 5\n",
            "100\t18\t0.6\t1\t11 7\n",
            "101\t2183\t0.6\t1\t14 2169\n",
            "\n",
            "RUN STATISTICS FOR INPUT FILE: /content/drive/MyDrive/SRR17158944_1.fastq\n",
            "=============================================\n",
            "39706568 sequences processed in total\n",
            "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
            "\n",
            "Writing report to '/content/drive/MyDrive/trimmed_output/SRR17158944_2.fastq_trimming_report.txt'\n",
            "\n",
            "SUMMARISING RUN PARAMETERS\n",
            "==========================\n",
            "Input filename: /content/drive/MyDrive/SRR17158944_2.fastq\n",
            "Trimming mode: paired-end\n",
            "Trim Galore version: 0.6.10\n",
            "Cutadapt version: 4.9\n",
            "Number of cores used for trimming: 1\n",
            "Quality Phred score cutoff: 20\n",
            "Quality encoding type selected: ASCII+33\n",
            "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
            "Maximum trimming error rate: 0.1 (default)\n",
            "Minimum required adapter overlap (stringency): 1 bp\n",
            "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
            "\n",
            "Cutadapt seems to be fairly up-to-date (version 4.9). Setting -j -j 1\n",
            "Writing final adapter and quality trimmed output to SRR17158944_2_trimmed.fq\n",
            "\n",
            "\n",
            "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /content/drive/MyDrive/SRR17158944_2.fastq <<< \n",
            "10000000 sequences processed\n",
            "20000000 sequences processed\n",
            "30000000 sequences processed\n",
            "This is cutadapt 4.9 with Python 3.10.12\n",
            "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /content/drive/MyDrive/SRR17158944_2.fastq\n",
            "Processing single-end reads on 1 core ...\n",
            "Finished in 634.292 s (15.974 µs/read; 3.76 M reads/minute).\n",
            "\n",
            "=== Summary ===\n",
            "\n",
            "Total reads processed:              39,706,568\n",
            "Reads with adapters:                15,267,790 (38.5%)\n",
            "Reads written (passing filters):    39,706,568 (100.0%)\n",
            "\n",
            "Total basepairs processed: 4,010,363,368 bp\n",
            "Quality-trimmed:              27,693,022 bp (0.7%)\n",
            "Total written (filtered):  3,941,474,213 bp (98.3%)\n",
            "\n",
            "=== Adapter 1 ===\n",
            "\n",
            "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 15267790 times\n",
            "\n",
            "Minimum overlap: 1\n",
            "No. of allowed errors:\n",
            "1-9 bp: 0; 10-13 bp: 1\n",
            "\n",
            "Bases preceding removed adapters:\n",
            "  A: 29.9%\n",
            "  C: 31.3%\n",
            "  G: 24.7%\n",
            "  T: 14.0%\n",
            "  none/other: 0.0%\n",
            "\n",
            "Overview of removed sequences\n",
            "length\tcount\texpect\tmax.err\terror counts\n",
            "1\t9521827\t9926642.0\t0\t9521827\n",
            "2\t3031384\t2481660.5\t0\t3031384\n",
            "3\t1162266\t620415.1\t0\t1162266\n",
            "4\t264982\t155103.8\t0\t264982\n",
            "5\t120568\t38775.9\t0\t120568\n",
            "6\t86427\t9694.0\t0\t86427\n",
            "7\t76981\t2423.5\t0\t76981\n",
            "8\t74534\t605.9\t0\t74534\n",
            "9\t67400\t151.5\t0\t66326 1074\n",
            "10\t67579\t37.9\t1\t65575 2004\n",
            "11\t59193\t9.5\t1\t57087 2106\n",
            "12\t54827\t2.4\t1\t53865 962\n",
            "13\t52533\t0.6\t1\t51815 718\n",
            "14\t51297\t0.6\t1\t50512 785\n",
            "15\t44611\t0.6\t1\t43953 658\n",
            "16\t42663\t0.6\t1\t42046 617\n",
            "17\t44425\t0.6\t1\t43784 641\n",
            "18\t30576\t0.6\t1\t30112 464\n",
            "19\t34888\t0.6\t1\t34344 544\n",
            "20\t32102\t0.6\t1\t31647 455\n",
            "21\t26639\t0.6\t1\t26273 366\n",
            "22\t26894\t0.6\t1\t26525 369\n",
            "23\t24698\t0.6\t1\t24302 396\n",
            "24\t23262\t0.6\t1\t22878 384\n",
            "25\t22444\t0.6\t1\t22104 340\n",
            "26\t18881\t0.6\t1\t18596 285\n",
            "27\t17595\t0.6\t1\t17258 337\n",
            "28\t17239\t0.6\t1\t17022 217\n",
            "29\t13816\t0.6\t1\t13625 191\n",
            "30\t14090\t0.6\t1\t13877 213\n",
            "31\t11827\t0.6\t1\t11638 189\n",
            "32\t11095\t0.6\t1\t10933 162\n",
            "33\t10733\t0.6\t1\t10556 177\n",
            "34\t9765\t0.6\t1\t9608 157\n",
            "35\t8854\t0.6\t1\t8712 142\n",
            "36\t8651\t0.6\t1\t8528 123\n",
            "37\t6869\t0.6\t1\t6749 120\n",
            "38\t6588\t0.6\t1\t6491 97\n",
            "39\t6740\t0.6\t1\t6616 124\n",
            "40\t5011\t0.6\t1\t4949 62\n",
            "41\t4999\t0.6\t1\t4902 97\n",
            "42\t5800\t0.6\t1\t5684 116\n",
            "43\t3097\t0.6\t1\t3036 61\n",
            "44\t4129\t0.6\t1\t4067 62\n",
            "45\t4190\t0.6\t1\t4121 69\n",
            "46\t3331\t0.6\t1\t3266 65\n",
            "47\t4730\t0.6\t1\t4652 78\n",
            "48\t1017\t0.6\t1\t982 35\n",
            "49\t3370\t0.6\t1\t3301 69\n",
            "50\t1266\t0.6\t1\t1231 35\n",
            "51\t1334\t0.6\t1\t1308 26\n",
            "52\t1114\t0.6\t1\t1091 23\n",
            "53\t1702\t0.6\t1\t1652 50\n",
            "54\t2340\t0.6\t1\t2286 54\n",
            "55\t518\t0.6\t1\t487 31\n",
            "56\t1223\t0.6\t1\t1177 46\n",
            "57\t3127\t0.6\t1\t3078 49\n",
            "58\t498\t0.6\t1\t478 20\n",
            "59\t335\t0.6\t1\t310 25\n",
            "60\t1627\t0.6\t1\t1594 33\n",
            "61\t305\t0.6\t1\t285 20\n",
            "62\t320\t0.6\t1\t294 26\n",
            "63\t1361\t0.6\t1\t1330 31\n",
            "64\t481\t0.6\t1\t459 22\n",
            "65\t108\t0.6\t1\t97 11\n",
            "66\t189\t0.6\t1\t171 18\n",
            "67\t157\t0.6\t1\t135 22\n",
            "68\t143\t0.6\t1\t124 19\n",
            "69\t140\t0.6\t1\t123 17\n",
            "70\t196\t0.6\t1\t176 20\n",
            "71\t85\t0.6\t1\t66 19\n",
            "72\t96\t0.6\t1\t80 16\n",
            "73\t85\t0.6\t1\t72 13\n",
            "74\t60\t0.6\t1\t50 10\n",
            "75\t50\t0.6\t1\t36 14\n",
            "76\t37\t0.6\t1\t20 17\n",
            "77\t67\t0.6\t1\t35 32\n",
            "78\t81\t0.6\t1\t62 19\n",
            "79\t59\t0.6\t1\t41 18\n",
            "80\t76\t0.6\t1\t49 27\n",
            "81\t37\t0.6\t1\t26 11\n",
            "82\t48\t0.6\t1\t36 12\n",
            "83\t142\t0.6\t1\t131 11\n",
            "84\t68\t0.6\t1\t43 25\n",
            "85\t62\t0.6\t1\t44 18\n",
            "86\t69\t0.6\t1\t52 17\n",
            "87\t54\t0.6\t1\t39 15\n",
            "88\t63\t0.6\t1\t48 15\n",
            "89\t84\t0.6\t1\t74 10\n",
            "90\t45\t0.6\t1\t38 7\n",
            "91\t119\t0.6\t1\t99 20\n",
            "92\t52\t0.6\t1\t35 17\n",
            "93\t80\t0.6\t1\t65 15\n",
            "94\t33\t0.6\t1\t20 13\n",
            "95\t18\t0.6\t1\t10 8\n",
            "96\t47\t0.6\t1\t30 17\n",
            "97\t43\t0.6\t1\t25 18\n",
            "98\t74\t0.6\t1\t24 50\n",
            "99\t21\t0.6\t1\t7 14\n",
            "100\t15\t0.6\t1\t5 10\n",
            "101\t19\t0.6\t1\t11 8\n",
            "\n",
            "RUN STATISTICS FOR INPUT FILE: /content/drive/MyDrive/SRR17158944_2.fastq\n",
            "=============================================\n",
            "39706568 sequences processed in total\n",
            "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
            "\n",
            "Validate paired-end files SRR17158944_1_trimmed.fq and SRR17158944_2_trimmed.fq\n",
            "file_1: SRR17158944_1_trimmed.fq, file_2: SRR17158944_2_trimmed.fq\n",
            "\n",
            "\n",
            ">>>>> Now validing the length of the 2 paired-end infiles: SRR17158944_1_trimmed.fq and SRR17158944_2_trimmed.fq <<<<<\n",
            "Writing validated paired-end Read 1 reads to SRR17158944_1_val_1.fq\n",
            "Writing validated paired-end Read 2 reads to SRR17158944_2_val_2.fq\n",
            "\n",
            "Total number of sequences analysed: 39706568\n",
            "\n",
            "Number of sequence pairs removed because at least one read was shorter than the length cutoff (20 bp): 231047 (0.58%)\n",
            "\n",
            "Deleting both intermediate output files SRR17158944_1_trimmed.fq and SRR17158944_2_trimmed.fq\n",
            "\n",
            "====================================================================================================\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!apt-get update\n",
        "!apt-get install -y hisat2"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "7wglQsEzAHXm",
        "outputId": "1941d072-0d50-4324-9fa3-bf6b114e964f"
      },
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Get:1 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease [3,626 B]\n",
            "Get:2 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  InRelease [1,581 B]\n",
            "Hit:3 http://archive.ubuntu.com/ubuntu jammy InRelease\n",
            "Get:4 http://security.ubuntu.com/ubuntu jammy-security InRelease [129 kB]\n",
            "Get:5 http://archive.ubuntu.com/ubuntu jammy-updates InRelease [128 kB]\n",
            "Get:6 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  Packages [1,071 kB]\n",
            "Ign:7 https://r2u.stat.illinois.edu/ubuntu jammy InRelease\n",
            "Hit:8 https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu jammy InRelease\n",
            "Get:9 https://r2u.stat.illinois.edu/ubuntu jammy Release [5,713 B]\n",
            "Hit:10 https://ppa.launchpadcontent.net/graphics-drivers/ppa/ubuntu jammy InRelease\n",
            "Get:11 https://r2u.stat.illinois.edu/ubuntu jammy Release.gpg [793 B]\n",
            "Get:12 http://archive.ubuntu.com/ubuntu jammy-backports InRelease [127 kB]\n",
            "Hit:13 https://ppa.launchpadcontent.net/ubuntugis/ppa/ubuntu jammy InRelease\n",
            "Get:14 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 Packages [2,654 kB]\n",
            "Get:15 https://r2u.stat.illinois.edu/ubuntu jammy/main amd64 Packages [2,602 kB]\n",
            "Get:16 http://archive.ubuntu.com/ubuntu jammy-updates/universe amd64 Packages [1,450 kB]\n",
            "Get:17 https://r2u.stat.illinois.edu/ubuntu jammy/main all Packages [8,428 kB]\n",
            "Fetched 16.6 MB in 6s (2,910 kB/s)\n",
            "Reading package lists... Done\n",
            "W: Skipping acquire of configured file 'main/source/Sources' as repository 'https://r2u.stat.illinois.edu/ubuntu jammy InRelease' does not seem to provide it (sources.list entry misspelt?)\n",
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "The following additional packages will be installed:\n",
            "  bcftools libhts3 libhtscodecs2 python3-hisat2 samtools\n",
            "Suggested packages:\n",
            "  python3-numpy python3-matplotlib texlive-latex-recommended cwltool\n",
            "The following NEW packages will be installed:\n",
            "  bcftools hisat2 libhts3 libhtscodecs2 python3-hisat2 samtools\n",
            "0 upgraded, 6 newly installed, 0 to remove and 50 not upgraded.\n",
            "Need to get 5,505 kB of archives.\n",
            "After this operation, 17.1 MB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libhtscodecs2 amd64 1.1.1-3 [53.2 kB]\n",
            "Get:2 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libhts3 amd64 1.13+ds-2build1 [390 kB]\n",
            "Get:3 http://archive.ubuntu.com/ubuntu jammy/universe amd64 bcftools amd64 1.13-1 [697 kB]\n",
            "Get:4 http://archive.ubuntu.com/ubuntu jammy/universe amd64 hisat2 amd64 2.2.1-3 [3,832 kB]\n",
            "Get:5 http://archive.ubuntu.com/ubuntu jammy/universe amd64 python3-hisat2 all 2.2.1-3 [12.7 kB]\n",
            "Get:6 http://archive.ubuntu.com/ubuntu jammy/universe amd64 samtools amd64 1.13-4 [520 kB]\n",
            "Fetched 5,505 kB in 3s (1,968 kB/s)\n",
            "Selecting previously unselected package libhtscodecs2:amd64.\n",
            "(Reading database ... 123657 files and directories currently installed.)\n",
            "Preparing to unpack .../0-libhtscodecs2_1.1.1-3_amd64.deb ...\n",
            "Unpacking libhtscodecs2:amd64 (1.1.1-3) ...\n",
            "Selecting previously unselected package libhts3:amd64.\n",
            "Preparing to unpack .../1-libhts3_1.13+ds-2build1_amd64.deb ...\n",
            "Unpacking libhts3:amd64 (1.13+ds-2build1) ...\n",
            "Selecting previously unselected package bcftools.\n",
            "Preparing to unpack .../2-bcftools_1.13-1_amd64.deb ...\n",
            "Unpacking bcftools (1.13-1) ...\n",
            "Selecting previously unselected package hisat2.\n",
            "Preparing to unpack .../3-hisat2_2.2.1-3_amd64.deb ...\n",
            "Unpacking hisat2 (2.2.1-3) ...\n",
            "Selecting previously unselected package python3-hisat2.\n",
            "Preparing to unpack .../4-python3-hisat2_2.2.1-3_all.deb ...\n",
            "Unpacking python3-hisat2 (2.2.1-3) ...\n",
            "Selecting previously unselected package samtools.\n",
            "Preparing to unpack .../5-samtools_1.13-4_amd64.deb ...\n",
            "Unpacking samtools (1.13-4) ...\n",
            "Setting up libhtscodecs2:amd64 (1.1.1-3) ...\n",
            "Setting up libhts3:amd64 (1.13+ds-2build1) ...\n",
            "Setting up bcftools (1.13-1) ...\n",
            "Setting up samtools (1.13-4) ...\n",
            "Setting up hisat2 (2.2.1-3) ...\n",
            "Setting up python3-hisat2 (2.2.1-3) ...\n",
            "Processing triggers for man-db (2.10.2-1) ...\n",
            "Processing triggers for libc-bin (2.35-0ubuntu3.4) ...\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_level_zero.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_5.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_opencl.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbb.so.12 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_0.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_loader.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc_proxy.so.2 is not a symbolic link\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!hisat2 -x /content/drive/MyDrive/reference_index/GRCm39_index \\\n",
        "       -1 /content/drive/MyDrive/trimmed_output/SRR17158944_1_val_1.fq\\\n",
        "       -2 /content/drive/MyDrive/trimmed_output/SRR17158944_2_val_2.fq \\\n",
        "       -S /content/drive/MyDrive/SRR17158944_aligned.sam"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "scDq6-sTAWC7",
        "outputId": "c115205c-5081-49c2-acab-b6e338cb7a92"
      },
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "39475521 reads; of these:\n",
            "  39475521 (100.00%) were paired; of these:\n",
            "    1827933 (4.63%) aligned concordantly 0 times\n",
            "    30964363 (78.44%) aligned concordantly exactly 1 time\n",
            "    6683225 (16.93%) aligned concordantly >1 times\n",
            "    ----\n",
            "    1827933 pairs aligned concordantly 0 times; of these:\n",
            "      174626 (9.55%) aligned discordantly 1 time\n",
            "    ----\n",
            "    1653307 pairs aligned 0 times concordantly or discordantly; of these:\n",
            "      3306614 mates make up the pairs; of these:\n",
            "        1695730 (51.28%) aligned 0 times\n",
            "        1197803 (36.22%) aligned exactly 1 time\n",
            "        413081 (12.49%) aligned >1 times\n",
            "97.85% overall alignment rate\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!samtools view -bS /content/drive/MyDrive/SRR17158944_aligned.sam | samtools sort -o /content/drive/MyDrive/SRR17158944_sorted.bam\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "cBACm1CVYGXr",
        "outputId": "61898b2b-d8d3-41c9-b543-7242a2abbf5b"
      },
      "execution_count": 10,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[bam_sort_core] merging from 45 files and 1 in-memory blocks...\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!samtools index /content/drive/MyDrive/SRR17158944_sorted.bam"
      ],
      "metadata": {
        "id": "opnM7uBwgiR2"
      },
      "execution_count": 11,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!samtools flagstat /content/drive/MyDrive/SRR17158944_sorted.bam"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "aPijhZttgy9Y",
        "outputId": "dfd00d54-7905-41a5-8b5b-0db6528541a4"
      },
      "execution_count": 12,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "115919732 + 0 in total (QC-passed reads + QC-failed reads)\n",
            "78951042 + 0 primary\n",
            "36968690 + 0 secondary\n",
            "0 + 0 supplementary\n",
            "0 + 0 duplicates\n",
            "0 + 0 primary duplicates\n",
            "114224002 + 0 mapped (98.54% : N/A)\n",
            "77255312 + 0 primary mapped (97.85% : N/A)\n",
            "78951042 + 0 paired in sequencing\n",
            "39475521 + 0 read1\n",
            "39475521 + 0 read2\n",
            "75295176 + 0 properly paired (95.37% : N/A)\n",
            "76362976 + 0 with itself and mate mapped\n",
            "892336 + 0 singletons (1.13% : N/A)\n",
            "991008 + 0 with mate mapped to a different chr\n",
            "740675 + 0 with mate mapped to a different chr (mapQ>=5)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!apt-get install -y subread"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "S8SA-63oheqJ",
        "outputId": "ec40e0b6-3fd1-4666-a3cb-c6c0702e73b9"
      },
      "execution_count": 13,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "subread is already the newest version (2.0.3+dfsg-1).\n",
            "0 upgraded, 0 newly installed, 0 to remove and 50 not upgraded.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!featureCounts -a /content/drive/MyDrive/gencode.vM35.basic.annotation.gtf -o /content/drive/MyDrive/SRR17158944_counts.txt -T 4 -p /content/drive/MyDrive/SRR17158944_sorted.bam\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "HtOrYKlCh3Fh",
        "outputId": "c6489320-35bc-496d-8b50-8bde29fb1002"
      },
      "execution_count": 14,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
            "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
            "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
            "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
            "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
            "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
            "\t  v2.0.3\n",
            "\n",
            "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
            "||  \u001b[0m                                                                          ||\n",
            "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "||                           \u001b[36mSRR17158944_sorted.bam\u001b[0m \u001b[0m                          ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "||             Output file : \u001b[36mSRR17158944_counts.txt\u001b[0m \u001b[0m                          ||\n",
            "||                 Summary : \u001b[36mSRR17158944_counts.txt.summary\u001b[0m \u001b[0m                  ||\n",
            "||              Paired-end : \u001b[36myes\u001b[0m \u001b[0m                                             ||\n",
            "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
            "||              Annotation : \u001b[36mgencode.vM35.basic.annotation.gtf (GTF)\u001b[0m \u001b[0m         ||\n",
            "||      Dir for temp files : \u001b[36m/content/drive/MyDrive\u001b[0m \u001b[0m                          ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "||                 Threads : \u001b[36m4\u001b[0m \u001b[0m                                               ||\n",
            "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
            "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
            "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
            "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "\\\\============================================================================//\n",
            "\n",
            "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
            "||  \u001b[0m                                                                          ||\n",
            "|| Load annotation file gencode.vM35.basic.annotation.gtf \u001b[0m... \u001b[0m                ||\n",
            "||    Features : \u001b[36m532131\u001b[0m \u001b[0m                                                      ||\n",
            "||    Meta-features : \u001b[36m57132\u001b[0m \u001b[0m                                                  ||\n",
            "||    Chromosomes/contigs : \u001b[36m22\u001b[0m \u001b[0m                                               ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "|| Process BAM file SRR17158944_sorted.bam... \u001b[0m                                ||\n",
            "||    Paired-end reads are included. \u001b[0m                                         ||\n",
            "||    The reads are assigned on the single-end mode. \u001b[0m                         ||\n",
            "||    Total alignments : \u001b[36m115919732\u001b[0m \u001b[0m                                           ||\n",
            "||    Successfully assigned alignments : \u001b[36m57870483 (49.9%)\u001b[0m \u001b[0m                    ||\n",
            "||    Running time : \u001b[36m2.17 minutes\u001b[0m \u001b[0m                                            ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "|| Write the final count table. \u001b[0m                                              ||\n",
            "|| Write the read assignment summary. \u001b[0m                                        ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "|| Summary of counting results can be found in file \"/content/drive/MyDrive/ \u001b[0m ||\n",
            "|| SRR17158944_counts.txt.summary\" \u001b[0m                                           ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "\\\\============================================================================//\n",
            "\n"
          ]
        }
      ]
    }
  ]
}